

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>filter_alignment.py – Filter sequence alignment by removing highly variable regions &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.6.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="filter-alignment"></span><div class="section" id="filter-alignment-py-filter-sequence-alignment-by-removing-highly-variable-regions">
<span id="index-0"></span><h1><em>filter_alignment.py</em> &#8211; Filter sequence alignment by removing highly variable regions<a class="headerlink" href="#filter-alignment-py-filter-sequence-alignment-by-removing-highly-variable-regions" title="Permalink to this headline">¶</a></h1>
<p><strong>Description:</strong></p>
<p>This script should be applied to generate a useful tree when aligning against a template alignment (e.g., with PyNAST). This script will remove positions which are gaps in every sequence (common for PyNAST, as typical sequences cover only 200-400 bases, and they are being aligned against the full 16S gene). Additionally, the user can supply a lanemask file, that defines which positions should included when building the tree, and which should be ignored. Typically, this will differentiate between non-conserved positions, which are uninformative for tree building, and conserved positions which are informative for tree building. FILTERING ALIGNMENTS WHICH WERE BUILD WITH PYNAST AGAINST THE GREENGENES CORE SET ALIGNMENT SHOULD BE CONSIDERED AN ESSENTIAL STEP.</p>
<p><strong>Usage:</strong> <tt class="file docutils literal"><span class="pre">filter_alignment.py</span> <span class="pre">[options]</span></tt></p>
<p><strong>Input Arguments:</strong></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><strong>[REQUIRED]</strong></p>
<dl class="docutils">
<dt>-i, <cite>-</cite>-input_fasta_file</dt>
<dd>The input directory</dd>
</dl>
<p><strong>[OPTIONAL]</strong></p>
<dl class="last docutils">
<dt>-o, <cite>-</cite>-output_dir</dt>
<dd>The output directory [default: .]</dd>
<dt>-m, <cite>-</cite>-lane_mask_fp</dt>
<dd>Path to lanemask file [default: /Users/caporaso/data/greengenes_core_sets/lanemask_in_1s_and_0s.txt]</dd>
<dt>-s, <cite>-</cite>-suppress_lane_mask_filter</dt>
<dd>Suppress lane mask filtering (necessary to turn off lane-mask-based filtering when a qiime_config default is  provided for &#8211;lane_mask_fp) [default: False]</dd>
<dt>-g, <cite>-</cite>-allowed_gap_frac</dt>
<dd>Gap filter threshold, filters positions which are gaps in &gt; allowed_gap_frac of the sequences [default: 0.999999]</dd>
<dt>-r, <cite>-</cite>-remove_outliers</dt>
<dd>Remove seqs very dissimilar to the alignment consensus (see &#8211;threshold).  [default: False]</dd>
<dt>-t, <cite>-</cite>-threshold</dt>
<dd>With -r, remove seqs whose dissimilarity to the consensus sequence is approximately &gt; x standard devaitions above the mean of the sequences [default: 3.0]</dd>
<dt>-e, <cite>-</cite>-entropy_threshold</dt>
<dd>Sets percent threshold for removing base positions with the highest entropy.  For example, if 0.10 were specified, the top 10% most entropic base positions would be filtered.  If this value is used, any lane mask supplied will be ignored.  Entropy filtered occurs after gap filtering.    [default: None]</dd>
</dl>
</div>
<p><strong>Output:</strong></p>
<p>The output of <a class="reference external" href="./filter_alignment.html">filter_alignment.py</a> consists of a single FASTA file, which ends with &#8220;pfiltered.fasta&#8221;, where the &#8220;p&#8221; stands for positional filtering of the columns.</p>
<p><strong>Examples:</strong></p>
<p>As a simple example of this script, the user can use the following command, which consists of an input FASTA file (i.e. resulting file from <a class="reference external" href="./align_seqs.html">align_seqs.py</a>), lanemask template file and the output directory &#8220;filtered_alignment/&#8221;:</p>
<div class="highlight-python"><pre>filter_alignment.py -i seqs_rep_set_aligned.fasta -m lanemask_in_1s_and_0s -o filtered_alignment/</pre>
</div>
<p>Alternatively, if the user would like to use a different gap fraction threshold (&#8220;-g&#8221;), they can use the following command:</p>
<div class="highlight-python"><pre>filter_alignment.py -i seqs_rep_set_aligned.fasta -m lanemask_in_1s_and_0s -o filtered_alignment/ -g 0.95</pre>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    

	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="../install/index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="../tutorials/index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.pycogent.org" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2012, QIIME Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>