<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>QIIME Overview Tutorial &mdash; Homepage</title>
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.1.0',
        COLLAPSE_MODINDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="qiime-overview-tutorial">
<span id="tutorial"></span><h1>QIIME Overview Tutorial<a class="headerlink" href="#qiime-overview-tutorial" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This tutorial explains how to use the <strong>QIIME</strong> (Quantitative Insights Into Microbial Ecology) Pipeline to process data from high-throughput 16S rRNA sequencing studies. The purpose of this pipeline is to provide a start-to-finish workflow, beginning with multiplexed sequence reads and finishing with taxonomic and phylogenetic profiles and comparisons of the samples in the study. With this information in hand, it is possible to determine biological and environmental factors that alter microbial community ecology in your experiment.</p>
<p>As an example, we will use data from a study of the response of mouse gut microbial communities to fasting (Crawford et al., 2009). To make this tutorial run quickly on a personal computer, we will use a subset of the data generated from 5 animals kept on the control ad libitum fed diet, and 4 animals fasted for 24 hours before sacrifice. At the end of our tutorial, we will be able to compare the community structure of control vs. fasted animals. In particular, we will be able to compare taxonomic profiles for each sample type, differences in diversity metrics within the samples and between the groups, and perform comparative clustering analysis to look for overall differences in the samples.</p>
<p>To process our data, we will perform the following steps, each of which is described in more detail in the Data Analysis Steps:</p>
<ul class="simple">
<li>Filter the sequence reads for quality and assign multiplexed reads to starting samples by nucleotide barcode.</li>
<li>Pick Operational Taxonomic Units (OTUs) based on sequence similarity within the reads, and pick a representative sequence from each OTU.</li>
<li>Assign the OTU to a taxonomic identity using reference databases.</li>
<li>Align the OTU sequences and create a phylogenetic tree.</li>
<li>Calculate diversity metrics for each sample and compare the types of communities, using the taxonomic and phylogenetic assignments.</li>
<li>Generate UPGMA and PCoA plots to visually depict the differences between the samples, and dynamically work with these graphs to generate publication quality figures.</li>
</ul>
</div>
<div class="section" id="essential-files">
<h2>Essential Files<a class="headerlink" href="#essential-files" title="Permalink to this headline">¶</a></h2>
<p>All the files you will need for this tutorial are here (<a class="reference external" href="http://bmf.colorado.edu/QIIME/qiime_tutorial-v1.1.0.zip">http://bmf.colorado.edu/QIIME/qiime_tutorial-v1.1.0.zip</a>). Descriptions of these files are below.</p>
<div class="section" id="sequences-fna">
<h3>Sequences (.fna)<a class="headerlink" href="#sequences-fna" title="Permalink to this headline">¶</a></h3>
<p>This is the 454-machine generated FASTA file. Using the Amplicon processing software on the 454 FLX standard, each region of the PTP plate will yield a fasta file of form <tt class="docutils literal"><span class="pre">1.TCA.454Reads.fna</span></tt>, where &#8220;1&#8221; is replaced with the appropriate region number. For the purposes of this tutorial, we will use the fasta file <tt class="docutils literal"><span class="pre">Fasting_Example.fna</span></tt>.</p>
</div>
<div class="section" id="quality-scores-qual">
<h3>Quality Scores (.qual)<a class="headerlink" href="#quality-scores-qual" title="Permalink to this headline">¶</a></h3>
<p>This is the 454-machine generated quality score file, which contains a score for each base in each sequence included in the FASTA file. Like the fasta file mentioned above, the Amplicon processing software will generate one of these files for each region of the PTP plate, named <tt class="docutils literal"><span class="pre">1.TCA.454Reads.qual</span></tt>, etc. For the purposes of this tutorial, we will use the quality scores file <tt class="docutils literal"><span class="pre">Fasting_Example.qual</span></tt>.</p>
</div>
<div class="section" id="mapping-file-tab-delimited-txt">
<h3>Mapping File (Tab-delimited .txt)<a class="headerlink" href="#mapping-file-tab-delimited-txt" title="Permalink to this headline">¶</a></h3>
<p>The mapping file is generated by the user. This file contains all of the information about the samples necessary to perform the data analysis. At a minimum, the mapping file should contain the name of each sample, the barcode sequence used for each sample, the linker/primer sequence used to amplify the sample, and a Description column. In general, you should also include in the mapping file any metadata that relates to the samples (for instance, health status or sampling site) and any additional information relating to specific samples that may be useful to have at hand when considering outliers (for example, what medications a patient was taking at time of sampling). Full format specifications can be found in the Documentation.</p>
<p>You are highly encouraged to validate your mapping file using <a class="reference external" href="../scripts/check_id_map.html">check_id_map.py</a> before attempting to analyze your data. This tool will check for errors, and make suggestions for other aspects of the file to be edited (errors and warnings are output to a log file, and suggested changes to invalid characters are output to a <tt class="docutils literal"><span class="pre">corrected_mapping.txt</span></tt> file). For the purposes of this tutorial, we will use the mapping file <tt class="docutils literal"><span class="pre">Fasting_Map.txt</span></tt>. The contents of the mapping file are shown here - as you can see, a nucleotide barcode sequence is provided for each of the 9 samples, as well as metadata related to treatment group and date of birth, and general run descriptions about the project. Fasting_Map.txt file contents:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>#SampleID  BarcodeSequence LinkerPrimerSequence    Treatment DOB   Description</li>
<li>#Example mapping file for the QIIME analysis package. These 9 samples are from a study of the effects of</li>
<li>#exercise and diet on mouse cardiac physiology (Crawford, et al, PNAS, 2009).</li>
<li>PC.354 AGCACGAGCCTA    YATGCTGCCTCCCGTAGGAGT   Control 20061218    Control_mouse__I.D._354</li>
<li>PC.355 AACTCGTCGATG    YATGCTGCCTCCCGTAGGAGT   Control 20061218    Control_mouse__I.D._355</li>
<li>PC.356 ACAGACCACTCA    YATGCTGCCTCCCGTAGGAGT   Control 20061126    Control_mouse__I.D._356</li>
<li>PC.481 ACCAGCGACTAG    YATGCTGCCTCCCGTAGGAGT   Control 20070314    Control_mouse__I.D._481</li>
<li>PC.593 AGCAGCACTTGT    YATGCTGCCTCCCGTAGGAGT   Control 20071210    Control_mouse__I.D._593</li>
<li>PC.607 AACTGTGCGTAC    YATGCTGCCTCCCGTAGGAGT   Fast    20071112    Fasting_mouse__I.D._607</li>
<li>PC.634 ACAGAGTCGGCT    YATGCTGCCTCCCGTAGGAGT   Fast    20080116    Fasting_mouse__I.D._634</li>
<li>PC.635 ACCGCAGAGTCA    YATGCTGCCTCCCGTAGGAGT   Fast    20080116    Fasting_mouse__I.D._635</li>
<li>PC.636 ACGGTGAGTGTC    YATGCTGCCTCCCGTAGGAGT   Fast    20080116    Fasting_mouse__I.D._636</li>
</ul>
</div>
</div>
<div class="section" id="flowgram-file-sff-optional">
<h3>Flowgram File (.sff) - (Optional)<a class="headerlink" href="#flowgram-file-sff-optional" title="Permalink to this headline">¶</a></h3>
<p>This is the 454-machine generated file which stores the sequencing trace data. This is the largest file returned from a 454 run. The sffinfo command in the 454 software package can be used to generate sequence and quality files from sff file(s) as follows</p>
<p>To generate a fasta file</p>
<div class="highlight-python"><pre>sffinfo -s NAME_OF_SFF_FILES &gt; OUTPUT_NAME.fna</pre>
</div>
<p>To generate a quality score file</p>
<div class="highlight-python"><pre>sffinfo -q NAME_OF_SFF_FILES &gt;OUTPUT_NAME.qual</pre>
</div>
</div>
</div>
<div class="section" id="data-analysis-steps">
<h2>Data Analysis Steps<a class="headerlink" href="#data-analysis-steps" title="Permalink to this headline">¶</a></h2>
<p>In this walkthrough, white text on a black background denote the command-line invocation of scripts. You can find full usage information for each script by passing the -h option (help) and/or by reading the full description in the <a class="reference external" href="../documentation/index.html">Documentation</a>. First, assemble the sequences (.fna), quality scores (.qual), and metadata mapping file into a directory. Execute all tutorial commands from within the <tt class="docutils literal"><span class="pre">qiime_tutorial</span></tt> directory, which can be downloaded from <a class="reference external" href="http://bmf.colorado.edu/QIIME/qiime_tutorial-v1.1.0.zip">here</a>.</p>
</div>
<div class="section" id="pre-processing-454-data">
<span id="preprocessing454"></span><h2>Pre-processing 454 Data<a class="headerlink" href="#pre-processing-454-data" title="Permalink to this headline">¶</a></h2>
<p>Filter the reads based on quality, and assign multiplexed reads to starting sample by nucleotide barcode.</p>
</div>
<div class="section" id="check-mapping-file">
<span id="checkmapping"></span><h2>Check Mapping File<a class="headerlink" href="#check-mapping-file" title="Permalink to this headline">¶</a></h2>
<p>Before beginning the pipeline, you should ensure that your mapping file is formatted correctly with the <a class="reference external" href="../scripts/check_id_map.html">check_id_map.py</a> script.</p>
<div class="highlight-python"><pre>check_id_map.py -m Fasting_Map.txt -o mapping_output/</pre>
</div>
<p>If verbose (-v) is enabled, this utility will print to STDOUT a message indicating whether or not problems were found in the mapping file. Errors and warnings will the output to a log file, which will be present in the specified (-o) output directory. Errors will cause fatal problems with subsequent scripts and must be corrected before moving forward. Warnings will not cause fatal problems, but it is encouraged that you fix these problems as they are often indicative of typos in your mapping file, invalid characters, or other unintended errors that will impact downstream analysis. A <tt class="docutils literal"><span class="pre">corrected_mapping.txt</span></tt> file will also be created in the output directory, which will have a copy of the mapping file with invalid characters replaced by underscores, or a message indicating that no invalid characters were found.</p>
</div>
<div class="section" id="assign-samples-to-multiplex-reads">
<span id="assignsamples"></span><h2>Assign Samples to Multiplex Reads<a class="headerlink" href="#assign-samples-to-multiplex-reads" title="Permalink to this headline">¶</a></h2>
<p>The next task is to assign the multiplex reads to samples based on their nucleotide barcode. Also, this step performs quality filtering based on the characteristics of each sequence, removing any low quality or ambiguous reads. The script for this step is <a class="reference external" href="../scripts/split_libraries.html">split_libraries.py</a>. A full description of parameters for this script are described in the <a class="reference external" href="../documentation/index.html">Documentation</a>. For this tutorial, we will use default parameters (minimum quality score = 25, minimum/maximum length = 200/1000, no ambiguous bases allowed and no mismatches allowed in the primer sequence).:</p>
<div class="highlight-python"><pre>split_libraries.py -m Fasting_Map.txt -f Fasting_Example.fna -q Fasting_Example.qual -o split_library_output</pre>
</div>
<p>This invocation will create three files in the new directory <tt class="docutils literal"><span class="pre">split_library_output/</span></tt>:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">split_library_log.txt</span></tt> : This file contains the summary of splitting, including the number of reads detected for each sample and a brief summary of any reads that were removed due to quality considerations.</li>
<li><tt class="docutils literal"><span class="pre">histograms.txt</span></tt> : This tab delimited file shows the number of reads at regular size intervals before and after splitting the library.</li>
<li><tt class="docutils literal"><span class="pre">seqs.fna</span></tt> : This is a fasta formatted file where each sequence is renamed according to the sample it came from. The header line also contains the name of the read in the input fasta file and information on any barcode errors that were corrected.</li>
</ul>
<p>A few lines from the <tt class="docutils literal"><span class="pre">seqs.fna</span></tt> file are shown below:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>&gt;PC.634_1 FLP3FBN01ELBSX orig_bc=ACAGAGTCGGCT new_bc=ACAGAGTCGGCT bc_diffs=0</li>
<li>CTGGGCCGTGTCTCAGTCCCAATGTGGCCGTTTACCCTCTCAGGCCGGCTACGCATCATCGCC....</li>
<li>&gt;PC.634_2 FLP3FBN01EG8AX orig_bc=ACAGAGTCGGCT new_bc=ACAGAGTCGGCT bc_diffs=0</li>
<li>TTGGACCGTGTCTCAGTTCCAATGTGGGGGCCTTCCTCTCAGAACCCCTATCCATCGAAGGCTT....</li>
<li>&gt;PC.354_3 FLP3FBN01EEWKD orig_bc=AGCACGAGCCTA new_bc=AGCACGAGCCTA bc_diffs=0</li>
<li>TTGGGCCGTGTCTCAGTCCCAATGTGGCCGATCAGTCTCTTAACTCGGCTATGCATCATTGCCTT....</li>
<li>&gt;PC.481_4 FLP3FBN01DEHK3 orig_bc=ACCAGCGACTAG new_bc=ACCAGCGACTAG bc_diffs=0</li>
<li>CTGGGCCGTGTCTCAGTCCCAATGTGGCCGTTCAACCTCTCAGTCCGGCTACTGATCGTCGACT....</li>
</ul>
</div>
</div>
<div class="section" id="workflow-scripts-and-the-parameters-file">
<h2>Workflow scripts and the parameters file<a class="headerlink" href="#workflow-scripts-and-the-parameters-file" title="Permalink to this headline">¶</a></h2>
<p>QIIME includes workflow scripts, which allow multiple tasks to be performed with one command.  Within the QIIME directory there is a file <tt class="docutils literal"><span class="pre">qiime_parameters.txt</span></tt>, where the user can set parameters for specific steps within a workflow script.  The user should make a copy of <tt class="docutils literal"><span class="pre">qiime_parameters.txt</span></tt> and place it into their working directory and give it a new filename (e.g. <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt>), but DO NOT EDIT the original file.  If you are using the tutorial dataset, the parameters file <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> is included, which has many parameters already set with appropriate values for the tutorial data. For more information on the <tt class="docutils literal"><span class="pre">qiime_parameters.txt</span></tt> file, please refer to <a class="reference external" href="./doc_qiime_parameters.html">here</a>. In this tutorial, we will utilize the workflow scripts when appropriate and within each section where the workflow is used, we will discuss which options in the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file associate to each step within the workflow. Users can run the workflow scripts in parallel by passing &#8220;-a&#8221; option to each of the scripts, however, this means that if you are running these scripts on a laptop, there must be more than one core in the machine (e.g. Intel duo or quad core).</p>
</div>
<div class="section" id="pick-operational-taxonomic-units-otus-through-making-otu-table">
<span id="pickotusandrepseqs"></span><h2>Pick Operational Taxonomic Units (OTUs) through making OTU table<a class="headerlink" href="#pick-operational-taxonomic-units-otus-through-making-otu-table" title="Permalink to this headline">¶</a></h2>
<p>Here we will be running the <a class="reference external" href="../scripts/pick_otus_through_otu_table.html">pick_otus_through_otu_table.py</a> workflow, which consists of the following steps:</p>
<ol class="arabic simple">
<li>Pick OTUs (for more information, refer to <a class="reference external" href="../scripts/pick_otus.html">pick_otus.py</a>)</li>
<li>Pick a representative sequence set (for more information, refer to <a class="reference external" href="../scripts/pick_rep_set.html">pick_rep_set.py</a>)</li>
<li>Align the representative sequence set (for more information, refer to <a class="reference external" href="../scripts/align_seqs.html">align_seqs.py</a>)</li>
<li>Assign taxonomy (for more information, refer to <a class="reference external" href="../scripts/assign_taxonomy.html">assign_taxonomy.py</a>)</li>
<li>Filter the alignment prior to tree building - remove positions which are all gaps, and specified as 0 in the lanemask (for more information, refer to <a class="reference external" href="../scripts/filter_alignment.html">filter_alignment.py</a>)</li>
<li>Build a phylogenetic tree  (for more information, refer to <a class="reference external" href="../scripts/make_phylogeny.html">make_phylogeny.py</a>)</li>
<li>Build an OTU table (for more information, refer to <a class="reference external" href="../scripts/make_otu_table.html">make_otu_table.py</a>)</li>
</ol>
<p>We will first go through each step and define the parameters in <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> and then at the end, we will run this workflow script.</p>
<p>Optionally, we can denoise the sequences based on clustering the flowgram sequences. For a single library/sff file we can simply use the workflow script <a class="reference external" href="../scripts/pick_otus_through_otu_table.html">pick_otus_through_otu_tables.py</a>, by providing the script with the sff file and the metadata mapping file. For multiple sff files refer to the special purpose tutorial <a class="reference external" href="denoising_454_data.html">Denoising of 454 Data Sets</a>.</p>
<div class="section" id="step-1-pick-otus-based-on-sequence-similarity-within-the-reads">
<span id="pickotusseqsim"></span><h3>Step 1. Pick OTUs based on Sequence Similarity within the Reads<a class="headerlink" href="#step-1-pick-otus-based-on-sequence-similarity-within-the-reads" title="Permalink to this headline">¶</a></h3>
<p>At this step, all of the sequences from all of the samples will be clustered into Operational Taxonomic Units (OTUs) based on their sequence similarity. OTUs in QIIME are clusters of sequences, frequently intended to represent some degree of taxonomic relatedness. For example, when sequences are clustered at 97% sequence similarity with uclust, each resulting cluster is typically thought of as representing a genus. This model and the current techniques for picking OTUs are known to be flawed, and determining exactly how OTUs should be defined, and what they represent, is an active area of research. Thus, OTU-picking will identify highly similar sequences across the samples and provide a platform for comparisons of community structure. The script <a class="reference external" href="../scripts/pick_otus.html">pick_otus.py</a> takes as input the fasta file output from <a class="reference internal" href="#assignsamples"><em>Assign Samples to Multiplex Reads</em></a> above, and returns a list of OTUs detected and the fasta header for sequences that belong in that OTU. To make the workflow invoke pick_otus.py using uclust to cluster and the default setting of 97% similarity determining an OTU, include the following settings in the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># OTU picker parameters</li>
<li>pick_otus:otu_picking_method  uclust</li>
<li>pick_otus:clustering_algorithm    furthest</li>
<li>pick_otus:max_cdhit_memory    400</li>
<li>pick_otus:refseqs_fp</li>
<li>pick_otus:blast_db</li>
<li>pick_otus:similarity  0.97</li>
<li>pick_otus:max_e_value 1e-10</li>
<li>pick_otus:prefix_prefilter_length</li>
<li>pick_otus:trie_prefilter</li>
<li>pick_otus:prefix_length</li>
<li>pick_otus:suffix_length</li>
<li>pick_otus:optimal_uclust</li>
<li>pick_otus:exact_uclust</li>
<li>pick_otus:user_sort</li>
<li>pick_otus:suppress_presort_by_abundance_uclust</li>
<li>pick_otus:suppress_new_clusters</li>
</ul>
</div>
<p>Note that tabs/space separate fields, e.g.: pick_otus:similarity 0.97.  Many of these parameters are blank, therefore default values are used; however, the user can supply variables when necessary.  Once this step in the workflow is run, in the newly created directory <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/</span></tt>, there will be two files. One is <tt class="docutils literal"><span class="pre">seqs.log</span></tt>, which contains information about the invocation of the script. The OTUs will be recorded in the tab-delimited file <tt class="docutils literal"><span class="pre">seqs_otus.txt</span></tt>. The OTUs are arbitrarily named by a number, which is recorded in the first column. The subsequent columns in each line identify the sequence or sequences that belong in that OTU.</p>
</div>
<div class="section" id="step-2-pick-representative-sequences-for-each-otu">
<span id="pickrepseqsforotu"></span><h3>Step 2. Pick Representative Sequences for each OTU<a class="headerlink" href="#step-2-pick-representative-sequences-for-each-otu" title="Permalink to this headline">¶</a></h3>
<p>Since each OTU may be made up of many sequences, we will pick a representative sequence for that OTU for downstream analysis. This representative sequence will be used for taxonomic identification of the OTU and phylogenetic alignment. The script <a class="reference external" href="../scripts/pick_rep_set.html">pick_rep_set.py</a> uses the OTU file created above and extracts a representative sequence from the fasta file by one of several methods. To use the default method, where the most abundant sequence in the OTU is used as the representative sequence, set the parameters in <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> as follows:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Representative set picker parameters</li>
<li>pick_rep_set:rep_set_picking_method   most_abundant</li>
<li>pick_rep_set:sort_by  otu</li>
</ul>
</div>
<p>In the <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/</span></tt> directory, the script has created two new files - the log file <tt class="docutils literal"><span class="pre">seqs_rep_set.log</span></tt> and the fasta file <tt class="docutils literal"><span class="pre">seqs_rep_set.fasta</span></tt> containing one representative sequence for each OTU. In this fasta file, the sequence has been renamed by the OTU, and the additional information on the header line reflects the sequence used as the representative:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>&gt;0 PC.636_424</li>
<li>CTGGGCCGTATCTCAGTCCCAATGTGGCCGGTCGACCTCTC....</li>
<li>&gt;1 PC.481_321</li>
<li>TTGGGCCGTGTCTCAGTCCCAATGTGGCCGTCCGCCCTCTC....</li>
</ul>
</div>
</div>
<div class="section" id="step-3-align-otu-sequences">
<span id="alignotuseq"></span><h3>Step 3. Align OTU Sequences<a class="headerlink" href="#step-3-align-otu-sequences" title="Permalink to this headline">¶</a></h3>
<p>Alignment of the sequences and phylogeny inference is necessary only if phylogenetic tools such as <a class="reference external" href="http://bmf2.colorado.edu/unifrac/index.psp">UniFrac</a> will be subsequently invoked. Alignments can either be generated de novo using programs such as MUSCLE, or through assignment to an existing alignment with tools like <a class="reference external" href="http://pynast.sourceforge.net/">PyNAST</a>. For small studies such as this tutorial, either method is possible. However, for studies involving many sequences (roughly, more than 1000), the de novo aligners are very slow and assignment with <a class="reference external" href="http://pynast.sourceforge.net/">PyNAST</a> is preferred. Either alignment approach is accomplished with the script <a class="reference external" href="../scripts/align_seqs.html">align_seqs.py</a>. Since this is one of the most computationally intensive bottlenecks in the pipeline, large studies would benefit greatly from parallelization of this task (described in detail in the <a class="reference external" href="../documentation/index.html">Documentation</a>):  When using <a class="reference external" href="http://pynast.sourceforge.net/">PyNAST</a> as an aligner, the user must supply a template alignment and if the user followed the instructions (<tt class="docutils literal"><span class="pre">4.Getting_started_with_QIIME.txt</span></tt>) in the Virtual Machine, then the greengenes files will be located in <tt class="docutils literal"><span class="pre">/home/qiime/</span></tt>.  For the tutorial, we will use <a class="reference external" href="http://pynast.sourceforge.net/">PyNAST</a> as the alignment method, UCLUST for the pairwise alignment method, a minimum length of 150 and a minimum percent identity of 75.0 in <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> as follows:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Multiple sequence alignment parameters</li>
<li>align_seqs:template_fp    /home/qiime/core_set_aligned.fasta.imputed</li>
<li>align_seqs:alignment_method   pynast</li>
<li>align_seqs:pairwise_alignment_method  uclust</li>
<li>align_seqs:blast_db</li>
<li>align_seqs:min_length 150</li>
<li>align_seqs:min_percent_id 75.0</li>
</ul>
</div>
<p>A log file and an alignment file are created in the directory <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/pynast_aligned_seqs/</span></tt>.</p>
</div>
<div class="section" id="step-4-assign-taxonomy">
<span id="assigntax"></span><h3>Step 4. Assign Taxonomy<a class="headerlink" href="#step-4-assign-taxonomy" title="Permalink to this headline">¶</a></h3>
<p>A primary goal of the QIIME pipeline is to assign high-throughput sequencing reads to taxonomic identities using established databases. This will give you information on the microbial lineages found in your samples. Using <a class="reference external" href="../scripts/assign_taxonomy.html">assign_taxonomy.py</a>, you can compare your OTUs against a reference database of your choosing. For our example, we will set the assignment_method to the RDP classification system and a confidence of 0.8 in <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt>. Note: the option &#8220;assign_taxonomy:e_value&#8221; is commented out, since it is not used for the rdp method and it will cause the parallel version of this workflow to fail.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Taxonomy assignment parameters</li>
<li>assign_taxonomy:id_to_taxonomy_fp</li>
<li>assign_taxonomy:reference_seqs_fp</li>
<li>assign_taxonomy:assignment_method rdp</li>
<li>assign_taxonomy:blast_db</li>
<li>assign_taxonomy:confidence    0.8</li>
<li>#assign_taxonomy:e_value   0.001</li>
</ul>
</div>
<p>In the directory <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy</span></tt>, there will be a log file and a text file. The text file contains a line for each OTU considered, with the RDP taxonomy assignment and a numerical confidence of that assignment (1 is the highest possible confidence). For some OTUs, the assignment will be as specific as a bacterial species, while others may be assignable to nothing more specific than the bacterial domain. Below are the first few lines of the text file and the user should note that the taxonomic assignment and confidence numbers from their run may not coincide with the output shown below, due to the RDP classification algorithm:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>41    PC.356_347  Root;Bacteria                                                                   0.980</li>
<li>63    PC.635_130  Root;Bacteria;Firmicutes;&#8221;Clostridia&#8221;;Clostridiales;&#8221;Lachnospiraceae&#8221;           0.960</li>
<li>353   PC.634_150  Root;Bacteria;Proteobacteria;Deltaproteobacteria                                0.880</li>
<li>18    PC.355_1011 Root;Bacteria;Bacteroidetes;Bacteroidetes;Bacteroidales;Rikenellaceae;Alistipes 0.990</li>
</ul>
</div>
</div>
<div class="section" id="step-5-filter-alignment">
<span id="filteraln"></span><h3>Step 5. Filter Alignment<a class="headerlink" href="#step-5-filter-alignment" title="Permalink to this headline">¶</a></h3>
<p>Before building the tree, one must filter the alignment to removed columns comprised of only gaps. Note that depending on where you obtained the lanemask file from, it will either be named lanemask_in_1s_and_0s.txt or lanemask_in_1s_and_0s.  If the user followed the instructions (<tt class="docutils literal"><span class="pre">4.Getting_started_with_QIIME.txt</span></tt>) in the Virtual Machine, then the greengenes files will be located in <tt class="docutils literal"><span class="pre">/home/qiime/</span></tt>. We will also set the allowed gap fraction as 0.999999, remove outliers to False and a threshold of 3.0 in <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> as follows:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Alignment filtering (prior to tree-building) parameters</li>
<li>filter_alignment:lane_mask_fp /home/qiime/lanemask_in_1s_and_0s.txt</li>
<li>filter_alignment:allowed_gap_frac  0.999999</li>
<li>filter_alignment:remove_outliers  False</li>
<li>filter_alignment:threshold    3.0</li>
</ul>
</div>
<p>A filtered alignment file is created in the directory <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/pynast_aligned_seqs/</span></tt>.</p>
</div>
<div class="section" id="step-6-make-phylogenetic-tree">
<span id="maketree"></span><h3>Step 6. Make Phylogenetic Tree<a class="headerlink" href="#step-6-make-phylogenetic-tree" title="Permalink to this headline">¶</a></h3>
<p>The filtered alignment file produced in the directory <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/pynast_aligned_seqs/</span></tt> can be used to build a phylogenetic tree using a tree-building program. As an example, we can set the tree_method to fasttree and the root_method to tree_method_default in <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Phylogenetic tree building parameters</li>
<li>make_phylogeny:tree_method    fasttree</li>
<li>make_phylogeny:root_method    tree_method_default</li>
</ul>
</div>
<p>The Newick format tree file is written to <tt class="docutils literal"><span class="pre">seqs_rep_set.tre</span></tt>, which is located in the <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/pynast_aligned_seqs/fasttree_phylogeny</span></tt> directory . This file can be viewed in a tree visualization software, and is necessary for <a class="reference external" href="http://bmf2.colorado.edu/unifrac/index.psp">UniFrac</a> diversity measurements (described below). For the following example, the FigTree program was used to visualize the phylogenetic tree obtained from <tt class="docutils literal"><span class="pre">seqs_rep_set.tre</span></tt>.</p>
<div align="center" class="align-center"><img alt="../_images/tree.png" class="align-center" src="../_images/tree.png" /></div>
</div>
<div class="section" id="step-7-make-otu-table">
<span id="makeotutable"></span><h3>Step 7. Make OTU Table<a class="headerlink" href="#step-7-make-otu-table" title="Permalink to this headline">¶</a></h3>
<p>Using these assignments and the OTU file created in <a class="reference internal" href="#pickotusseqsim"><em>Step 1. Pick OTUs based on Sequence Similarity within the Reads</em></a>, we can make a readable matrix of OTU by Sample with meaningful taxonomic identifiers for each OTU. Currently there are no parameters in <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> for the user to define when making an OTU table.</p>
<p>The result of this step is <tt class="docutils literal"><span class="pre">seqs_otu_table.txt</span></tt>, which is located in the <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/</span></tt> directory. The first few lines of <tt class="docutils literal"><span class="pre">seqs_otu_table.txt</span></tt> are shown below (OTUs 1-9), where the first column contains the OTU number, the last column contains the taxonomic assignment for the OTU, and 9 columns between are for each of our 9 samples. The value of each <em>ij</em> entry in the matrix is the number of times OTU <em>i</em> was found in the sequences for sample <em>j</em>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>#Full OTU Counts</li>
<li>#OTU ID    PC.354  PC.355  PC.356  PC.481  PC.593  PC.607  PC.634  PC.635  PC.636  Consensus Lineage</li>
<li>0  0   0   0   0   0   0   0   1   0   Root;Bacteria;Firmicutes;&#8221;Clostridia&#8221;;Clostridiales;&#8221;Lachnospiraceae&#8221;</li>
<li>1  0   0   0   0   0   1   0   0   0   Root;Bacteria;Firmicutes;&#8221;Clostridia&#8221;;Clostridiales;&#8221;Lachnospiraceae&#8221;</li>
<li>2  0   0   0   0   0   0   0   0   1   Root;Bacteria;Bacteroidetes;Bacteroidetes;Bacteroidales;Porphyromonadaceae;Parabacteroides</li>
<li>3  2   1   0   0   0   0   0   0   0   Root;Bacteria;Firmicutes;&#8221;Clostridia&#8221;;Clostridiales;&#8221;Lachnospiraceae&#8221;;&#8221;Lachnospiraceae Incertae Sedis&#8221;</li>
<li>4  1   0   0   0   0   0   0   0   0   Root;Bacteria;Firmicutes;&#8221;Clostridia&#8221;;Clostridiales;&#8221;Lachnospiraceae&#8221;</li>
<li>5  0   0   0   0   0   0   0   0   1   Root;Bacteria;Firmicutes;&#8221;Clostridia&#8221;;Clostridiales</li>
<li>6  0   0   0   0   0   0   0   1   0   Root;Bacteria;Actinobacteria;Actinobacteria</li>
<li>7  0   0   2   0   0   0   0   0   1   Root;Bacteria;Firmicutes;&#8221;Clostridia&#8221;;Clostridiales;&#8221;Ruminococcaceae&#8221;</li>
<li>8  1   1   0   2   4   0   0   0   0   Root;Bacteria;Firmicutes;&#8221;Bacilli&#8221;;&#8221;Lactobacillales&#8221;;Lactobacillaceae;Lactobacillus</li>
<li>9  0   0   2   0   0   0   0   0   0   Root;Bacteria;Firmicutes;&#8221;Clostridia&#8221;;Clostridiales;&#8221;Lachnospiraceae&#8221;</li>
</ul>
</div>
</div>
<div class="section" id="running-pick-otus-through-otu-table-py">
<h3>Running pick_otus_through_otu_table.py<a class="headerlink" href="#running-pick-otus-through-otu-table-py" title="Permalink to this headline">¶</a></h3>
<p>Now that we have set the parameters necessary for this workflow script, the user can run the following command, where we define the input sequence file &#8220;-i&#8221; (from <a class="reference external" href="../scripts/split_libraries.html">split_libraries.py</a>), the parameter file to use &#8220;-p&#8221; and the output directory &#8220;-o&#8221;:</p>
<div class="highlight-python"><pre>pick_otus_through_otu_table.py -i split_library_output/seqs.fna -p custom_parameters.txt -o wf_da</pre>
</div>
</div>
<div class="section" id="rarify-otu-table-to-remove-sample-heterogeneity-optional">
<span id="rareotutableremovehetero"></span><h3>Rarify OTU Table to Remove Sample Heterogeneity (Optional)<a class="headerlink" href="#rarify-otu-table-to-remove-sample-heterogeneity-optional" title="Permalink to this headline">¶</a></h3>
<p>To remove sample heterogeneity, we can perform rarefaction on our OTU table. Rarefaction is an ecological approach that allows users to standardize the data obtained from samples with different sequencing efforts, and to compare the OTU richness of the samples using this standardized platform. For instance, if one of your samples yielded 10,000 sequence counts, and another yielded only 1,000 counts, the species diversity within those samples may be much more influenced by sequencing effort than underlying biology. The approach of rarefaction is to randomly sample the same number of OTUs from each sample, and use this data to compare the communities at a given level of sampling effort.</p>
<p>To perform rarefaction, you need to set the boundaries for sampling and the step size between sampling intervals. You can find the number of sequences associated with each sample by looking in the <tt class="docutils literal"><span class="pre">split_library_log.txt</span></tt> file generated in <a class="reference internal" href="#assignsamples"><em>Assign Samples to Multiplex Reads</em></a> above. The line from our tutorial is pasted here:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>Sample ct min/max/mean: 146 / 150 / 148.11</li>
</ul>
</div>
<p>Since we are only removing sample heterogeneity from the OTU table, we will use <a class="reference external" href="../scripts/single_rarefaction.html">single_rarefaction.py</a>, which only requires the depth of sampling. Rarefaction is most useful when most samples have the specified number of sequences, so your upper bound of rarefaction should be close to the minimum number of sequences found in a sample. For this case, we will set the depth to 146.</p>
<div class="highlight-python"><pre>single_rarefaction.py -i wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/seqs_otu_table.txt -d 146 -o wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/rarified_otu_table.txt</pre>
</div>
<p>As a result, a newly created file <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/rarified_otu_table.txt</span></tt> has been written. For all subsequent analyses, one can use this text file in place of <tt class="docutils literal"><span class="pre">seqs_otu_table.txt</span></tt>, since this file is also an OTU table, except that it has been rarified.</p>
</div>
<div class="section" id="make-otu-heatmap">
<span id="makeheatmap"></span><h3>Make OTU Heatmap<a class="headerlink" href="#make-otu-heatmap" title="Permalink to this headline">¶</a></h3>
<p>The QIIME pipeline includes a very useful utility to generate images of the OTU table. The script is <a class="reference external" href="../scripts/make_otu_heatmap_html.html">make_otu_heatmap_html.py</a></p>
<div class="highlight-python"><pre>make_otu_heatmap_html.py -i wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/seqs_otu_table.txt -o wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/OTU_Heatmap/</pre>
</div>
<p>An html file is created in the directory &#8220;wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/Fasting_OTU_Heatmap/&#8221;. You can open this file with any web browser, and will be prompted to enter a value for &#8220;Filter by Counts per OTU&#8221;. Only OTUs with total counts at or above this threshold will be displayed. The OTU heatmap displays raw OTU counts per sample, where the counts are colored based on the contribution of each OTU to the total OTU count present in that sample (blue: contributes low percentage of OTUs to sample; red: contributes high percentage of OTUs). Click the &#8220;Sample ID&#8221; button, and a graphic will be generated like the figure below. For each sample, you will see in a heatmap the number of times each OTU was found in that sample. You can mouse over any individual count to get more information on the OTU (including taxonomic assignment). Within the mouseover, there is a link for the terminal lineage assignment, so you can easily search Google for more information about that assignment.</p>
<div align="center" class="align-center"><img alt="../_images/heatmap.png" class="align-center" src="../_images/heatmap.png" /></div>
<p>Alternatively, you can click on one of the counts in the heatmap and a new pop-up window will appear. The pop-up window uses a Google Visualization API called Magic-Table. Depending on which table count you clicked on, the pop-up window will put the clicked-on count in the middle of the pop-up heatmap as shown below. For the following example, the table count with the red arrow mouseover is the same one being focused on using the Magic-Table.</p>
<div align="center" class="align-center"><img alt="../_images/fisheyeheatmap.png" class="align-center" src="../_images/fisheyeheatmap.png" /></div>
<p>On the original heatmap webpage, if you select the &#8220;Taxonomy&#8221; button instead, you will generate a heatmap keyed by taxon assignment, which allows you to conveniently look for organisms and lineages of interest in your study. Again, mousing over an individual count will show additional information for that OTU and sample.</p>
<div align="center" class="align-center"><img alt="../_images/taxheatmap.png" class="align-center" src="../_images/taxheatmap.png" /></div>
</div>
<div class="section" id="make-otu-network">
<span id="makeotunetwork"></span><h3>Make OTU Network<a class="headerlink" href="#make-otu-network" title="Permalink to this headline">¶</a></h3>
<p>An alternative to viewing the OTU table as a heatmap is to create an OTU network, using the following command.:</p>
<div class="highlight-python"><pre>make_otu_network.py -m Fasting_Map.txt -i wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/seqs_otu_table.txt -o wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/OTU_Network</pre>
</div>
<p>To visualize the network, we use the <a class="reference external" href="http://www.cytoscape.org/">Cytoscape</a> program (which you can run by calling cytoscape from the command line -- you may need to call this beginning either with a capital or lowercase 'C' depending on your version of Cytoscape), where each red circle represents a sample and each white square represents an OTU. The lines represent the OTUs present in a particular sample (blue for controls and green for fasting). For more information about opening the files in <a class="reference external" href="http://www.cytoscape.org/">Cytoscape</a> please refer <a class="reference external" href="../scripts/cytoscape_usage.html">here</a>.</p>
<div align="center" class="align-center"><img alt="../_images/network.png" class="align-center" src="../_images/network.png" /></div>
<p>You can group OTUs by different taxonomic levels (division, class, family) with the script <a class="reference external" href="../scripts/summarize_taxa.html">summarize_taxa.py</a>. The input is the OTU table created above and the taxonomic level you need to group the OTUs. For the RDP taxonomy, the following taxonomic levels correspond to: 2 = Domain (Bacteria), 3 = Phylum (Actinobacteria), 4 = Class, and so on.</p>
<div class="highlight-python"><pre>summarize_taxa.py -i wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/seqs_otu_table.txt -o wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/otu_table_Level3.txt -L 3 -r 0</pre>
</div>
<p>The script will generate a new OTU table <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/otu_table_Level3.txt</span></tt>, where the value of each <em>ij</em> entry in the matrix is the count of the number of times all OTUs belonging to the taxon <em>i</em> (for example, Phylum Actinobacteria) were found in the sequences for sample <em>j</em>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>#Full OTU Counts</li>
<li>Taxon              PC.354 PC.355   PC.356  PC.481  PC.593  PC.607  PC.634  PC.635  PC.636</li>
<li>Root;Bacteria;Actinobacteria   0.0 0.0 0.0 1.0 0.0 2.0 3.0 1.0     1.0</li>
<li>Root;Bacteria;Bacteroidetes    7.0 38.0    15.0    19.0    30.0    40.0    86.0    54.0    90.0</li>
<li>Root;Bacteria;Deferribacteres  0.0 0.0 0.0 0.0 0.0 3.0 5.0 2.0 7.0</li>
<li>Root;Bacteria;Firmicutes   136.0   102.0   115.0   117.0   65.0    66.0    37.0    63.0    34.0</li>
<li>Root;Bacteria;Other        5.0 6.0 18.0    9.0 49.0    35.0    14.0    27.0    14.0</li>
<li>Root;Bacteria;Proteobacteria   0.0 0.0 0.0 0.0 5.0 3.0 2.0 0.0 1.0</li>
<li>Root;Bacteria;TM7      0.0 0.0 0.0 0.0 0.0 0.0 2.0 0.0 0.0</li>
<li>Root;Bacteria;Verrucomicrobia  0.0 0.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0</li>
<li>Root;Other         0.0 0.0 2.0 0.0 0.0 0.0 0.0 1.0 0.0</li>
</ul>
</div>
</div>
<div class="section" id="make-pie-charts">
<span id="makepiecharts"></span><h3>Make Pie Charts<a class="headerlink" href="#make-pie-charts" title="Permalink to this headline">¶</a></h3>
<p>To visualize the summarized taxa, you can use the <a class="reference external" href="../scripts/make_pie_charts.html">make_pie_charts.py</a> script, which shows which taxons are present in all samples or within each sample (-s).  To use this script, we need to set the taxonomy level label &#8220;-l&#8221;, an output directory &#8220;-o&#8221;, and the background color &#8220;-k&#8221; as white:</p>
<div class="highlight-python"><pre>make_pie_charts.py -i wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/otu_table_Level3.txt -l Phylum -o wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/Pie_Charts -k white -s</pre>
</div>
<p>To view the resulting pie charts, open the html file located in the <tt class="docutils literal"><span class="pre">wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/Pie_Charts/</span></tt> folder. The following pie chart shows the taxa assignments for all samples.</p>
<div align="center" class="align-center"><img alt="../_images/piechart1.png" class="align-center" src="../_images/piechart1.png" /></div>
<p>The following pie chart shows the taxa assignments for one of the samples (PC.354).</p>
<div align="center" class="align-center"><img alt="../_images/piechart2.png" class="align-center" src="../_images/piechart2.png" /></div>
</div>
</div>
<div class="section" id="compute-alpha-diversity-within-the-samples-and-generate-rarefaction-curves">
<span id="compalphadivrarecurves"></span><h2>Compute Alpha Diversity within the Samples and Generate Rarefaction Curves<a class="headerlink" href="#compute-alpha-diversity-within-the-samples-and-generate-rarefaction-curves" title="Permalink to this headline">¶</a></h2>
<p>Community ecologists typically describe the microbial diversity within their study. This diversity can be assessed within a sample (alpha diversity) or between a collection of samples (beta diversity). Here, we will determine the level of alpha diversity in our samples using a series of scripts from the QIIME pipeline.  To perform this analysis, we will use the <tt class="docutils literal"><span class="pre">alpha_rarefaction.py</span></tt> workflow script.  We will first set the parameters in <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt>, then at the end, we will run the script.  This script performs the following steps:</p>
<ol class="arabic simple">
<li>Generate rarefied OTU tables (for more information, refer to <a class="reference external" href="../scripts/multiple_rarefactions.html">multiple_rarefactions.py</a>)</li>
<li>Compute alpha diversity metrics for each rarefied OTU table (for more information, refer to <a class="reference external" href="../scripts/alpha_diversity.html">alpha_diversity.py</a>)</li>
<li>Collate alpha diversity results (for more information, refer to <a class="reference external" href="../scripts/collate_alpha.html">collate_alpha.py</a>)</li>
<li>Generate alpha rarefaction plots (for more information, refer to <a class="reference external" href="../scripts/make_rarefaction_plots.html">make_rarefaction_plots.py</a>)</li>
</ol>
<div class="section" id="step-1-rarify-otu-table">
<span id="rareotutable"></span><h3>Step 1. Rarify OTU Table<a class="headerlink" href="#step-1-rarify-otu-table" title="Permalink to this headline">¶</a></h3>
<p>For this highly artificial example, all of the samples had sequence counts between 146 and 150, as discussed in <a class="reference internal" href="#rareotutableremovehetero"><em>Rarify OTU Table to Remove Sample Heterogeneity (Optional)</em></a>. In real datasets, the range will generally be much larger. In practice, rarefaction is most useful when most samples have the specified number of sequences, so your upper bound of rarefaction should be close to the minimum number of sequences found in a sample.  For the this workflow script the min/max values are defined by the workflow script.  If the user would like to define their own values, they should perform each step individually.  In <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt>, the user can define the number of iterations at each sequence/sample level, where we will use &#8220;num-rep 5&#8221; and whether to include lineages, which we set to False:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Rarefaction parameters</li>
<li>multiple_rarefactions:num-reps    5</li>
<li>multiple_rarefactions:depth</li>
<li>multiple_rarefactions:lineages_included   False</li>
</ul>
</div>
<p>The directory <tt class="docutils literal"><span class="pre">wf_arare/rarefaction/</span></tt> will contain many text files named <tt class="docutils literal"><span class="pre">rarefaction_##_#.txt</span></tt>; the first set of numbers represents the number of sequences sampled, and the last number represents the iteration number. If you opened one of these files, you would find an OTU table where for each sample the sum of the counts equals the number of samples taken.</p>
</div>
<div class="section" id="step-2-compute-alpha-diversity">
<span id="computealphadiv"></span><h3>Step 2. Compute Alpha Diversity<a class="headerlink" href="#step-2-compute-alpha-diversity" title="Permalink to this headline">¶</a></h3>
<p>The rarefaction tables are the basis for calculating diversity metrics, which reflect the diversity within the sample based on taxon counts of phylogeny. The QIIME pipeline allows users to conveniently calculate more than two dozen different diversity metrics. The full list of available metrics is available <a class="reference external" href="../scripts/alpha_diversity_metrics.html">here</a>. Every metric has different strengths and limitations - technical discussion of each metric is readily available online and in ecology textbooks, but it is beyond the scope of this document. Here, we will calculate three metrics:</p>
<ol class="arabic simple">
<li>Chao1 metric estimates the species richness.</li>
<li>The Observed Species metric is simply the count of unique OTUs found in the sample.</li>
<li>Phylogenetic Distance (PD_whole_tree) is the only phylogenetic metric used in this script and requires a phylogenetic tree as an input.</li>
</ol>
<p>In the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file, the user can define a comma-delimited list of alpha diversity metrics to use, as follows:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Alpha diversity parameters</li>
<li>alpha_diversity:metrics   chao1,observed_species,PD_whole_tree</li>
</ul>
</div>
<p>The result of this step produces several text files, located in the <tt class="docutils literal"><span class="pre">wf_arare/alpha_div/</span></tt> directory.</p>
</div>
<div class="section" id="step-3-collate-rarified-otu-tables">
<span id="collateotutable"></span><h3>Step 3. Collate Rarified OTU Tables<a class="headerlink" href="#step-3-collate-rarified-otu-tables" title="Permalink to this headline">¶</a></h3>
<p>The output directory <tt class="docutils literal"><span class="pre">wf_arare/alpha_div/</span></tt> will contain one text file <tt class="docutils literal"><span class="pre">alpha_rarefaction_##_#</span></tt> for every file input from <tt class="docutils literal"><span class="pre">wf_arare/rarefaction/</span></tt>, where the numbers represent the number of samples and iterations as before. The content of this tab delimited file is the calculated metrics for each sample. To collapse the individual files into a single combined table, the workflow uses the script <a class="reference external" href="../scripts/collate_alpha.html">collate_alpha.py</a>. The user can define an &#8220;example_path&#8221; in the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file, however, for the tutorial, we will leave this blank.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Collate alpha</li>
<li>collate_alpha:example_path</li>
</ul>
</div>
<p>In the newly created directory <tt class="docutils literal"><span class="pre">wf_arare/alpha_div_collated/</span></tt>, there will be one matrix for every diversity metric used in the <a class="reference external" href="../scripts/alpha_diversity.html">alpha_diversity.py</a> script. This matrix will contain the metric for every sample, arranged in ascending order from lowest number of sequences per sample to highest. A portion of the <tt class="docutils literal"><span class="pre">observed_species.txt</span></tt> file are shown below:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>Sequences per sample   iteration   PC.354  PC.355  PC.356  PC.481  PC.593</li>
<li>alpha_rarefaction_21_0.txt 21          0       14.0    16.0    18.0    18.0    13.0</li>
<li>alpha_rarefaction_21_1.txt 21          1       15.0    17.0    18.0    20.0    12.0</li>
<li>alpha_rarefaction_21_2.txt 21          2       15.0    16.0    21.0    19.0    13.0</li>
<li>alpha_rarefaction_21_3.txt 21          3       10.0    19.0    18.0    21.0    13.0</li>
<li>alpha_rarefaction_21_4.txt 21          4       14.0    18.0    16.0    15.0    12.0</li>
<li>...</li>
</ul>
</div>
</div>
<div class="section" id="step-4-generate-rarefaction-curves">
<span id="generaterarecurves"></span><h3>Step 4. Generate Rarefaction Curves<a class="headerlink" href="#step-4-generate-rarefaction-curves" title="Permalink to this headline">¶</a></h3>
<p>The script <a class="reference external" href="../scripts/make_rarefaction_plots.html">make_rarefaction_plots.py</a> takes a mapping file and any number of rarefaction files generated by <a class="reference external" href="../scripts/collate_alpha.html">collate_alpha.py</a> and uses matplotlib to create rarefaction curves. Each curve represents a sample and can be colored by the sample metadata supplied in the mapping file. In the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file, the user can set the image format (i.e. png), resolution (i.e. 75), and background_color (i.e. white) as follows:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Make rarefaction plots parameters</li>
<li>make_rarefaction_plots:imagetype  png</li>
<li>make_rarefaction_plots:resolution 75</li>
<li>make_rarefaction_plots:background_color   white</li>
<li>make_rarefaction_plots:prefs_path</li>
</ul>
</div>
<p>This step generates a <tt class="docutils literal"><span class="pre">wf_arare/alpha_rarefaction_plots/average_tables/</span></tt> folder, which contains the rarefaction averages for each diversity metric, so the user can plot the rarefaction curves in another application, like MS Excel.   The <tt class="docutils literal"><span class="pre">wf_arare/alpha_rarefaction_plots/average_plots/</span></tt> folder contains the average plots for each metric and category and the <tt class="docutils literal"><span class="pre">wf_arare/alpha_rarefaction_plots/html_plots/</span></tt> folder contains all the images used in the html page generated. To view the rarefaction plots the user can open the file <tt class="docutils literal"><span class="pre">wf_arare/alpha_rarefaction_plots/rarefaction_plots.html</span></tt> in a browser. Once the browser window is open, the user can select the metric and category for whichever rarefaction plots they would like to display.  The user can also turn on/off lines in the plot by (un)checking the box next to each label in the legend.  The user can click on the triangle next to each label in the legend to see all the samples that contribute to that category. Below each plot, the user will see the average data over all metrics for the specified category.</p>
<div align="center" class="align-center"><img alt="../_images/rarecurve.png" class="align-center" src="../_images/rarecurve.png" /></div>
</div>
<div class="section" id="running-alpha-rarefaction-py">
<h3>Running alpha_rarefaction.py<a class="headerlink" href="#running-alpha-rarefaction-py" title="Permalink to this headline">¶</a></h3>
<p>Now that we have set the parameters, necessary for this workflow script, the user can run the following command, where we define the input OTU table &#8220;-i&#8221; and tree file &#8220;-t&#8221; (from <a class="reference external" href="../scripts/pick_otus_through_otu_table.html">pick_otus_through_otu_table.py</a>), the parameter file to use &#8220;-p&#8221;, and the output directory &#8220;-o&#8221;:</p>
<div class="highlight-python"><pre>alpha_rarefaction.py -i wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/seqs_otu_table.txt -m Fasting_Map.txt -o wf_arare/ -p custom_parameters.txt -t wf_da/uclust_picked_otus/rep_set/pynast_aligned_seqs/fasttree_phylogeny/seqs_rep_set.tre</pre>
</div>
</div>
</div>
<div class="section" id="compute-beta-diversity-and-generate-3d-principal-coordinate-analysis-pcoa-plots">
<span id="compbetadivgenpcoa"></span><h2>Compute Beta Diversity and Generate 3D Principal Coordinate Analysis (PCoA) Plots<a class="headerlink" href="#compute-beta-diversity-and-generate-3d-principal-coordinate-analysis-pcoa-plots" title="Permalink to this headline">¶</a></h2>
<p>Here we will be running the <a class="reference external" href="../scripts/beta_diversity_through_3d_plots.html">beta_diversity_through_3d_plots.py</a> workflow, which consists of the following steps:</p>
<ol class="arabic simple">
<li>Compute Beta Diversity (for more information, refer to <a class="reference external" href="../scripts/beta_diversity.html">beta_diversity.py</a>)</li>
<li>Generate Principal Coordinates (for more information, refer to <a class="reference external" href="../scripts/principal_coordinates.html">principal_coordinates.py</a>)</li>
<li>Make preferences file (for more information, refer to <a class="reference external" href="../scripts/make_prefs_file.html">make_prefs_file.py</a>)</li>
<li>Generate 3D PCoA plots (for more information, refer to <a class="reference external" href="../scripts/make_3d_plots.html">make_3d_plots.py</a>)</li>
</ol>
<div class="section" id="step-1-compute-beta-diversity">
<span id="compbetadiv"></span><h3>Step 1. Compute Beta Diversity<a class="headerlink" href="#step-1-compute-beta-diversity" title="Permalink to this headline">¶</a></h3>
<p>Beta-diversity metrics assess the differences between microbial communities. In general, these metrics are calculated to study diversity along an environmental gradient (pH or temperature) or different disease states (lean vs. obese). The basic output of this comparison is a square matrix where a &#8220;distance&#8221; is calculated between every pair of samples reflecting the similarity between the samples. The data in this distance matrix can be visualized with clustering analyses, namely Principal Coordinate Analysis (PCoA) and UPGMA clustering. Like alpha diversity, there are many possible metrics which can be calculated with the QIIME pipeline - the full list of options can be found <a class="reference external" href="../scripts/beta_diversity_metrics.html">here</a>. For our example, we will calculate weighted and unweighted unifrac, which are phylogenetic measures used extensively in recent microbial community sequencing projects, by defining the metric parameter in the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file, as follows:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Beta diversity parameters</li>
<li>beta_diversity:metrics    weighted_unifrac,unweighted_unifrac</li>
</ul>
</div>
<p>The resulting distance matrices ( <tt class="docutils literal"><span class="pre">wf_bdiv/unweighted_unifrac_seqs_otu_table.txt</span></tt> and <tt class="docutils literal"><span class="pre">wf_bdiv/weighted_unifrac_seqs_otu_table.txt</span></tt>) are the basis for two methods of visualization and sample comparison: PCoA and UPGMA.</p>
</div>
<div class="section" id="step-2-generate-principal-coordinates">
<h3>Step 2. Generate Principal Coordinates<a class="headerlink" href="#step-2-generate-principal-coordinates" title="Permalink to this headline">¶</a></h3>
<p>Principal Coordinate Analysis (PCoA) is a technique that helps to extract and visualize a few highly informative gradients of variation from complex, multidimensional data. This is a complex transformation that maps the distance matrix to a new set of orthogonal axes such that a maximum amount of variation is explained by the first principal coordinate, the second largest amount of variation is explained by the second principal coordinate, etc. The principal coordinates can be plotted in two or three dimensions to provide an intuitive visualization of the data structure and look at differences between the samples, and look for similarities by sample category. The transformation is accomplished with the script <a class="reference external" href="../scripts/principal_coordinates.html">principal_coordinates.py</a>.  Since this script only takes an input/output file, there are no parameters for the user to set in  <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt>.</p>
<p>The files <tt class="docutils literal"><span class="pre">wf_bdiv/unweighted_unifrac_pc.txt</span></tt> and <tt class="docutils literal"><span class="pre">wf_bdiv/weighted_unifrac_pc.txt</span></tt> lists every sample in the first column, and the subsequent columns contain the value for the sample against the noted principal coordinate. At the bottom of each Principal Coordinate column, you will find the eigenvalue and percent of variation explained by the coordinate. To determine which axes are useful for your project, you can generate a &#8220;scree plot&#8221; by plotting the eigenvalues of each principal component in descending order.</p>
</div>
<div class="section" id="step-3-make-preferences-file">
<span id="gen2d3dpcoa"></span><h3>Step 3. Make Preferences File<a class="headerlink" href="#step-3-make-preferences-file" title="Permalink to this headline">¶</a></h3>
<p>In order to generate the PCoA plots, we want to generate a preferences file, which defines the colors for each of the samples or for a particular category within a mapping column.  For more information on making a preferences file, please refer to <a class="reference external" href="../scripts/make_prefs_file.html">make_prefs_file.py</a>.  In the  <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file, the user can set the background color to be used for the 3D PCoA plot (either black or white), the mapping header categories to plot (uses ALL if left blank) and the monte carlo distance to use (this is for <a class="reference external" href="../scripts/make_distance_histograms.html">make_distance_histograms.py</a>, which we will do in a few steps).</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Make prefs file parameters</li>
<li>make_prefs_file:background_color  black</li>
<li>make_prefs_file:mapping_headers_to_use    Treatment,DOB</li>
<li>make_prefs_file:monte_carlo_dists 10</li>
</ul>
</div>
</div>
<div class="section" id="step-4-generate-3d-pcoa-plots">
<h3>Step 4. Generate 3D PCoA Plots<a class="headerlink" href="#step-4-generate-3d-pcoa-plots" title="Permalink to this headline">¶</a></h3>
<p>To plot the coordinates, you can use the QIIME scripts <a class="reference external" href="../scripts/make_2d_plots.html">make_2d_plots.py</a> and <a class="reference external" href="../scripts/make_3d_plots.html">make_3d_plots.py</a>. The two dimensional plot will be rendered as a html file which can be opened with a standard web browser, while the three dimensional plot will be a kinemage file which requires additional software to render and manipulate. The usage for both scripts use the same convention, detailed in <a class="reference external" href="../scripts/make_3d_plots.html">make_3d_plots.py</a>.  Since the coloring was set for the preferences file parameters, we only need to set the custom_axes in the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt>, although we can leave it blank, as follows:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Make 3D plot parameters</li>
<li>make_3d_plots:custom_axes</li>
</ul>
</div>
<p>The html files are created in <tt class="docutils literal"><span class="pre">wf_bdiv/unweighted_unifrac_3d...</span></tt> and <tt class="docutils literal"><span class="pre">wf_bdiv/weighted_unifrac_3d...</span></tt> directories.  In the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt>, we specified that the samples should be colored by the value of the &#8220;Treatment&#8221; and &#8220;DOB&#8221; columns under the make_prefs_file parameters. For the &#8220;Treatment&#8221; column, all samples with the same &#8220;Treatment&#8221; will get the same color. For our tutorial, the five control samples are all blue and the four control samples are all green. This lets you easily visualize &#8220;clustering&#8221; by metadata category. The 3d visualization software allows you to rotate the axes to see the data from different perspectives. By default, the script will plot the first three dimensions in your file. Other combinations can be viewed using the &#8220;Views:Choose viewing axes&#8221; option in the KiNG viewer (may require the installation of kinemage software). The first 10 components can be viewed using &#8220;Views:Paralleled coordinates&#8221; option or typing &#8220;/&#8221;.</p>
<div align="center" class="align-center"><img alt="../_images/pcoa2.png" class="align-center" src="../_images/pcoa2.png" /></div>
</div>
<div class="section" id="running-beta-diversity-through-3d-plots-py">
<h3>Running beta_diversity_through_3d_plots.py<a class="headerlink" href="#running-beta-diversity-through-3d-plots-py" title="Permalink to this headline">¶</a></h3>
<p>Now that we have set the parameters, necessary for this workflow script, the user can run the following command, where we define the input OTU table &#8220;-i&#8221; and tree file &#8220;-t&#8221; (from <a class="reference external" href="../scripts/pick_otus_through_otu_table.html">pick_otus_through_otu_table.py</a>), the parameter file to use &#8220;-p&#8221;, the user-defined mapping file &#8220;-m&#8221; and the output directory &#8220;-o&#8221;:</p>
<div class="highlight-python"><pre>beta_diversity_through_3d_plots.py -i wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/seqs_otu_table.txt -m Fasting_Map.txt -o wf_bdiv/ -p custom_parameters.txt -t wf_da/uclust_picked_otus/rep_set/pynast_aligned_seqs/fasttree_phylogeny/seqs_rep_set.tre</pre>
</div>
</div>
<div class="section" id="generate-2d-pcoa-plots">
<h3>Generate 2D PCoA Plots<a class="headerlink" href="#generate-2d-pcoa-plots" title="Permalink to this headline">¶</a></h3>
<p>To plot the coordinates for the unweighted unifrac principal coordinates in 2D, you can use the QIIME script <a class="reference external" href="../scripts/make_2d_plots.html">make_2d_plots.py</a>.  Here we will use the same preferences file generated from the <a class="reference external" href="../scripts/beta_diversity_through_3d_plots.html">beta_diversity_through_3d_plots.py</a>, set the background color &#8220;-k&#8221; to white and output the results to <cite>wf_bdiv/unweighted_unifrac_2d</cite>:</p>
<div class="highlight-python"><pre>make_2d_plots.py -i wf_bdiv/unweighted_unifrac_pc.txt -m Fasting_Map.txt -o wf_bdiv/unweighted_unifrac_2d -k white -p wf_bdiv/prefs.txt</pre>
</div>
<p>The html file created in directory <tt class="docutils literal"><span class="pre">wf_bdiv/unweighted_unifrac_2d</span></tt> shows a plot for each combination of the first three principal coordinates. Since we specified Treatment and DOB to use for coloring the samples, each sample colored according to the category it corresponds. You can get the name for each sample by holding your mouse over the data point.</p>
<div align="center" class="align-center"><img alt="../_images/pcoa1.png" class="align-center" src="../_images/pcoa1.png" /></div>
</div>
<div class="section" id="generate-distance-histograms">
<span id="gendisthist"></span><h3>Generate Distance Histograms<a class="headerlink" href="#generate-distance-histograms" title="Permalink to this headline">¶</a></h3>
<p>Distance Histograms are a way to compare different categories and see which tend to have larger/smaller distances than others. For example, in the hand study, you may want to compare the distances between hands to the distances between individuals. Here we will use the distance matrix and prefs file generated by <a class="reference external" href="../scripts/beta_diversity_through_3d_plots.html">beta_diversity_through_3d_plots.py</a>, the mapping file, an output directory <tt class="docutils literal"><span class="pre">wf_bdiv/Distance_Histograms</span></tt> and write the output as html, as follows:</p>
<div class="highlight-python"><pre>make_distance_histograms.py -d wf_bdiv/unweighted_unifrac_seqs_otu_table.txt -m Fasting_Map.txt -o wf_bdiv/Distance_Histograms -p wf_bdiv/prefs.txt --html_output</pre>
</div>
<p>For each of these groups of distances a histogram is made. The output is a HTML file (<tt class="docutils literal"><span class="pre">wf_bdiv/Distance_Histograms/QIIME_Distance_Histograms.html</span></tt>) where you can look at all the distance histograms individually, and compare them between each other. Within the webpage, the user can mouseover and/or select the checkboxes in the right panel to turn on/off the different distances within/between categories. For this example, we are comparing the distances between the samples in the Control versus themselves, along with samples from Fasting versus the Control.</p>
<div align="center" class="align-center"><img alt="../_images/hist.png" class="align-center" src="../_images/hist.png" /></div>
</div>
</div>
<div class="section" id="upgma-clustering-and-jackknifing-support">
<span id="hiarchclustjack"></span><span id="genpcoa"></span><h2>UPGMA Clustering and Jackknifing Support<a class="headerlink" href="#upgma-clustering-and-jackknifing-support" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt>The steps performed by this workflow are:</dt>
<dd><ol class="first last arabic simple">
<li>Compute beta diversity distance matrix from OTU table (and tree, if applicable) (for more information, refer to <a class="reference external" href="../scripts/beta_diversity.html">beta_diversity.py</a>)</li>
<li>Build UPGMA tree from full distance matrix; (for more information, refer to <a class="reference external" href="../scripts/upgma_cluster.html">upgma_cluster.py</a>)</li>
<li>Build rarefied OTU tables (for more information, refer to <a class="reference external" href="../scripts/multiple_rarefactions.html">multiple_rarefactions.py</a>)</li>
<li>Compute distance matrices for rarefied OTU tables (for more information, refer to <a class="reference external" href="../scripts/beta_diversity.html">beta_diversity.py</a>)</li>
<li>Build UPGMA trees from rarefied OTU table distance matrices (for more information, refer to <a class="reference external" href="../scripts/upgma_cluster.html">upgma_cluster.py</a>)</li>
<li>Compare rarefied OTU table distance matrix UPGMA trees to tree full UPGMA tree and write support file and newick tree with support values as node labels (for more information, refer to <a class="reference external" href="../scripts/tree_compare.html">tree_compare.py</a>)</li>
</ol>
</dd>
</dl>
<div class="section" id="steps-1-and-2-upgma-clustering">
<span id="hiarchclust"></span><h3>Steps 1 and 2. UPGMA Clustering<a class="headerlink" href="#steps-1-and-2-upgma-clustering" title="Permalink to this headline">¶</a></h3>
<p>Unweighted Pair Group Method with Arithmetic mean (UPGMA) is type of UPGMA clustering method using average linkage and can be used to visualize the distance matrix produced by <a class="reference external" href="../scripts/beta_diversity.html">beta_diversity.py</a>.</p>
<p>The output is a file that can be opened with tree viewing software, such as FigTree.</p>
<div align="center" class="align-center"><img alt="../_images/hiarchclust.png" class="align-center" src="../_images/hiarchclust.png" /></div>
<p>This tree shows the relationship among the 9 samples, and reveals that the 4 samples from the guts of fasting mice cluster together (PC.6xx, fasting data is in <tt class="docutils literal"><span class="pre">Fasting_Map.txt</span></tt>).</p>
</div>
<div class="section" id="steps-3-4-and-5-perform-jackknifing-support">
<span id="jacksupport"></span><h3>Steps 3, 4 and 5. Perform Jackknifing Support<a class="headerlink" href="#steps-3-4-and-5-perform-jackknifing-support" title="Permalink to this headline">¶</a></h3>
<p>To measure the robustness of this result to sequencing effort, we perform a jackknifing analysis, wherein a smaller number of sequences are chosen at random from each sample, and the resulting UPGMA tree from this subset of data is compared with the tree representing the entire available data set. This process is repeated with many random subsets of data, and the tree nodes which prove more consistent across jackknifed datasets are deemed more robust.</p>
<p>First the jackknifed OTU tables must be generated, by subsampling the full available data set. In this tutorial, each sample contains between 146 and 150 sequences, as shown in the <tt class="docutils literal"><span class="pre">split_library_log.txt</span></tt> file:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>Sample ct min/max/mean: 146 / 150 / 148.11</li>
</ul>
</div>
<p>To ensure that a random subset of sequences is selected from each sample, we chose to select 110 sequences from each sample (75% of the smallest sample, though this value is only a guideline), which is designated by the &#8220;-e&#8221; option when running the workflow script (see below). In the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file, we set the number of jackknife replicates as follows:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Multiple Rarefactions</li>
<li>multiple_rarefactions_even_depth:num-reps 20</li>
</ul>
</div>
<p>This generates 20 subsets of the available data, each subset a simulation of a smaller sequencing effort (110 sequences in each sample, as defined below).</p>
<p>We then calculate the distance matrix for each jackknifed dataset, using <a class="reference external" href="../scripts/beta_diversity.html">beta_diversity.py</a> as before, but now in batch mode, which results in 20 distance matrix files written to the <tt class="docutils literal"><span class="pre">wf_jack/unweighted_unifrac/rare_dm/</span></tt> and <tt class="docutils literal"><span class="pre">wf_jack/weighted_unifrac/rare_dm/</span></tt> directories. Each of those is then used as the basis for UPGMA clustering, using <a class="reference external" href="../scripts/upgma_cluster.html">upgma_cluster.py</a> in batch mode and written to the <tt class="docutils literal"><span class="pre">wf_jack/unweighted_unifrac/rare_upgma/</span></tt> and <tt class="docutils literal"><span class="pre">wf_jack/weighted_unifrac/rare_upgma/</span></tt> directories.</p>
</div>
<div class="section" id="step-6-compare-jackknifed-trees-to-cluster-tree">
<span id="compjackclustertree"></span><h3>Step 6. Compare Jackknifed Trees to Cluster Tree<a class="headerlink" href="#step-6-compare-jackknifed-trees-to-cluster-tree" title="Permalink to this headline">¶</a></h3>
<p>UPGMA clustering of the 20 distance matrix files results in 20 UPGMA samples clusters, each based on a random sub-sample of the available sequence data. These are then compared to the UPGMA result using all available data.</p>
<p>This compares the UPGMA clustering based on all available data with the jackknifed UPGMA results.  Three files are written to <tt class="docutils literal"><span class="pre">wf_jack/unweighted_unifrac/upgma_cmp/</span></tt> and <tt class="docutils literal"><span class="pre">wf_jack/weighted_unifrac/upgma_cmp/</span></tt>:</p>
<blockquote>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">master_tree.tre</span></tt>, which is virtually identical to <tt class="docutils literal"><span class="pre">jackknife_named_nodes.tre</span></tt> but each internal node of the UPGMA clustering is assigned a unique name</li>
<li><tt class="docutils literal"><span class="pre">jackknife_named_nodes.tre</span></tt></li>
<li><tt class="docutils literal"><span class="pre">jackknife_support.txt</span></tt> explains how frequently a given internal node had the same set of descendant samples in the jackknifed UPGMA clusters as it does in the UPGMA cluster using the full available data.  A value of 0.5 indicates that half of the jackknifed data sets support that node, while 1.0 indicates perfect support.</li>
</ul>
</blockquote>
</div>
<div class="section" id="running-jackknifed-upgma-py">
<h3>Running jackknifed_upgma.py<a class="headerlink" href="#running-jackknifed-upgma-py" title="Permalink to this headline">¶</a></h3>
<p>Now that we have set the parameter, necessary for this workflow script, the user can run the following command, where we define the input OTU table &#8220;-i&#8221; and tree file &#8220;-t&#8221; (from <a class="reference external" href="../scripts/pick_otus_through_otu_table.html">pick_otus_through_otu_table.py</a>), the parameter file to use &#8220;-p&#8221;, the output directory &#8220;-o&#8221; and the number of sequences per sample &#8220;-e&#8221; (i.e. 100):</p>
<div class="highlight-python"><pre>jackknifed_upgma.py -i wf_da/uclust_picked_otus/rep_set/rdp_assigned_taxonomy/otu_table/seqs_otu_table.txt -o wf_jack -p custom_parameters.txt -e 110 -t wf_da/uclust_picked_otus/rep_set/pynast_aligned_seqs/fasttree_phylogeny/seqs_rep_set.tre</pre>
</div>
</div>
<div class="section" id="generate-bootstrapped-tree">
<span id="genboottree"></span><h3>Generate Bootstrapped Tree<a class="headerlink" href="#generate-bootstrapped-tree" title="Permalink to this headline">¶</a></h3>
<p>As an example, we can visualize the bootstrapped tree using unweighted unifrac using <a class="reference external" href="../scripts/make_bootstrapped_tree.html">make_bootstrapped_tree.py</a>, as follows:</p>
<div class="highlight-python"><pre>make_bootstrapped_tree.py -m wf_jack/unweighted_unifrac/upgma_cmp/master_tree.tre -s wf_jack/unweighted_unifrac/upgma_cmp/jackknife_support.txt -o wf_jack/unweighted_unifrac/upgma_cmp/jackknife_named_nodes.pdf</pre>
</div>
<p>The resulting pdf shows the tree with internal nodes colored, red for 75-100% support, yellow for 50-75%, green for 25-50%, and blue for &lt; 25% support. Although UPGMA shows that PC.354 and PC.593 cluster together and PC.481 with PC.6xx cluster together, we can not have high confidence in that result. However, there is excellent jackknife support for all fasted samples (PC.6xx) which are clustering together, separate from the non-fasted (PC.35x) samples.</p>
<div align="center" class="align-center"><img alt="../_images/boottree.png" class="align-center" src="../_images/boottree.png" /></div>
</div>
</div>
<div class="section" id="running-workflow-scripts-in-parallel">
<h2>Running Workflow Scripts in Parallel<a class="headerlink" href="#running-workflow-scripts-in-parallel" title="Permalink to this headline">¶</a></h2>
<p>Users can run the workflow scripts in parallel by passing &#8220;-a&#8221; option to each of the scripts.  In the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file, the users can customize the number of jobs to start (i.e. jobs_to_start), whether to keep the temporary files generated (retain_temp_files), and the number of seconds to sleep (seconds_to_sleep).  If running on a dual-core computer, you can set the number of jobs to start as 2, as follows:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li># Parallel options</li>
<li>parallel:jobs_to_start    2</li>
<li>parallel:retain_temp_files    False</li>
<li>parallel:seconds_to_sleep 1</li>
</ul>
</div>
</div>
<div class="section" id="running-the-qiime-tutorial-shell-scripts">
<h2>Running the QIIME Tutorial Shell Scripts<a class="headerlink" href="#running-the-qiime-tutorial-shell-scripts" title="Permalink to this headline">¶</a></h2>
<p>Now that we have gone through the whole tutorial and customized the <tt class="docutils literal"><span class="pre">custom_parameters.txt</span></tt> file, we can run the shell scripts via the Terminal, which contain all the commands that you ran in this tutorial.  To run the shell scripts, you may need to allow all users to execute them, using the following commands:</p>
<div class="highlight-python"><pre>chmod a+x ./qiime_tutorial_commands_serial.sh
chmod a+x ./qiime_tutorial_commands_parallel.sh</pre>
</div>
<p>To run the QIIME tutorial in serial:</p>
<div class="highlight-python"><pre>./qiime_tutorial_commands_serial.sh</pre>
</div>
<p>To run the QIIME tutorial in parallel:</p>
<div class="highlight-python"><pre>./qiime_tutorial_commands_parallel.sh</pre>
</div>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<p>Crawford, P. A., Crowley, J. R., Sambandam, N., Muegge, B. D., Costello, E. K., Hamady, M., et al. (2009). Regulation of myocardial ketone body metabolism by the gut microbiota during nutrient deprivation. Proc Natl Acad Sci U S A, 106(27), 11276-11281.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
            <h3><a href="../index.html">Table Of Contents</a></h3>
            <ul>
<li><a class="reference external" href="">QIIME Overview Tutorial</a><ul>
<li><a class="reference external" href="#introduction">Introduction</a></li>
<li><a class="reference external" href="#essential-files">Essential Files</a><ul>
<li><a class="reference external" href="#sequences-fna">Sequences (.fna)</a></li>
<li><a class="reference external" href="#quality-scores-qual">Quality Scores (.qual)</a></li>
<li><a class="reference external" href="#mapping-file-tab-delimited-txt">Mapping File (Tab-delimited .txt)</a></li>
<li><a class="reference external" href="#flowgram-file-sff-optional">Flowgram File (.sff) - (Optional)</a></li>
</ul>
</li>
<li><a class="reference external" href="#data-analysis-steps">Data Analysis Steps</a></li>
<li><a class="reference external" href="#pre-processing-454-data">Pre-processing 454 Data</a></li>
<li><a class="reference external" href="#check-mapping-file">Check Mapping File</a></li>
<li><a class="reference external" href="#assign-samples-to-multiplex-reads">Assign Samples to Multiplex Reads</a></li>
<li><a class="reference external" href="#workflow-scripts-and-the-parameters-file">Workflow scripts and the parameters file</a></li>
<li><a class="reference external" href="#pick-operational-taxonomic-units-otus-through-making-otu-table">Pick Operational Taxonomic Units (OTUs) through making OTU table</a><ul>
<li><a class="reference external" href="#step-1-pick-otus-based-on-sequence-similarity-within-the-reads">Step 1. Pick OTUs based on Sequence Similarity within the Reads</a></li>
<li><a class="reference external" href="#step-2-pick-representative-sequences-for-each-otu">Step 2. Pick Representative Sequences for each OTU</a></li>
<li><a class="reference external" href="#step-3-align-otu-sequences">Step 3. Align OTU Sequences</a></li>
<li><a class="reference external" href="#step-4-assign-taxonomy">Step 4. Assign Taxonomy</a></li>
<li><a class="reference external" href="#step-5-filter-alignment">Step 5. Filter Alignment</a></li>
<li><a class="reference external" href="#step-6-make-phylogenetic-tree">Step 6. Make Phylogenetic Tree</a></li>
<li><a class="reference external" href="#step-7-make-otu-table">Step 7. Make OTU Table</a></li>
<li><a class="reference external" href="#running-pick-otus-through-otu-table-py">Running pick_otus_through_otu_table.py</a></li>
<li><a class="reference external" href="#rarify-otu-table-to-remove-sample-heterogeneity-optional">Rarify OTU Table to Remove Sample Heterogeneity (Optional)</a></li>
<li><a class="reference external" href="#make-otu-heatmap">Make OTU Heatmap</a></li>
<li><a class="reference external" href="#make-otu-network">Make OTU Network</a></li>
<li><a class="reference external" href="#make-pie-charts">Make Pie Charts</a></li>
</ul>
</li>
<li><a class="reference external" href="#compute-alpha-diversity-within-the-samples-and-generate-rarefaction-curves">Compute Alpha Diversity within the Samples and Generate Rarefaction Curves</a><ul>
<li><a class="reference external" href="#step-1-rarify-otu-table">Step 1. Rarify OTU Table</a></li>
<li><a class="reference external" href="#step-2-compute-alpha-diversity">Step 2. Compute Alpha Diversity</a></li>
<li><a class="reference external" href="#step-3-collate-rarified-otu-tables">Step 3. Collate Rarified OTU Tables</a></li>
<li><a class="reference external" href="#step-4-generate-rarefaction-curves">Step 4. Generate Rarefaction Curves</a></li>
<li><a class="reference external" href="#running-alpha-rarefaction-py">Running alpha_rarefaction.py</a></li>
</ul>
</li>
<li><a class="reference external" href="#compute-beta-diversity-and-generate-3d-principal-coordinate-analysis-pcoa-plots">Compute Beta Diversity and Generate 3D Principal Coordinate Analysis (PCoA) Plots</a><ul>
<li><a class="reference external" href="#step-1-compute-beta-diversity">Step 1. Compute Beta Diversity</a></li>
<li><a class="reference external" href="#step-2-generate-principal-coordinates">Step 2. Generate Principal Coordinates</a></li>
<li><a class="reference external" href="#step-3-make-preferences-file">Step 3. Make Preferences File</a></li>
<li><a class="reference external" href="#step-4-generate-3d-pcoa-plots">Step 4. Generate 3D PCoA Plots</a></li>
<li><a class="reference external" href="#running-beta-diversity-through-3d-plots-py">Running beta_diversity_through_3d_plots.py</a></li>
<li><a class="reference external" href="#generate-2d-pcoa-plots">Generate 2D PCoA Plots</a></li>
<li><a class="reference external" href="#generate-distance-histograms">Generate Distance Histograms</a></li>
</ul>
</li>
<li><a class="reference external" href="#upgma-clustering-and-jackknifing-support">UPGMA Clustering and Jackknifing Support</a><ul>
<li><a class="reference external" href="#steps-1-and-2-upgma-clustering">Steps 1 and 2. UPGMA Clustering</a></li>
<li><a class="reference external" href="#steps-3-4-and-5-perform-jackknifing-support">Steps 3, 4 and 5. Perform Jackknifing Support</a></li>
<li><a class="reference external" href="#step-6-compare-jackknifed-trees-to-cluster-tree">Step 6. Compare Jackknifed Trees to Cluster Tree</a></li>
<li><a class="reference external" href="#running-jackknifed-upgma-py">Running jackknifed_upgma.py</a></li>
<li><a class="reference external" href="#generate-bootstrapped-tree">Generate Bootstrapped Tree</a></li>
</ul>
</li>
<li><a class="reference external" href="#running-workflow-scripts-in-parallel">Running Workflow Scripts in Parallel</a></li>
<li><a class="reference external" href="#running-the-qiime-tutorial-shell-scripts">Running the QIIME Tutorial Shell Scripts</a></li>
<li><a class="reference external" href="#references">References</a></li>
</ul>
</li>
</ul>


	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
		
	 <li><a href="../index.html">Home</a><br /></li>
	 <li><a href="../install/index.html">Install</a><br /></li>
	 <li><a href="../documentation/index.html">Documentation</a><br /></li>
	 <li><a href="index.html">Tutorials</a><br /></li>
	 <li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	 <li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


          <div id="searchbox" style="display: none">
            <h3>Quick search</h3>
              <form class="search" action="../search.html" method="get">
                <input type="text" name="q" size="18" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
              </form>
              <p class="searchtip" style="font-size: 90%">
              Enter search terms or a module, class or function name.
              </p>
          </div>
          <script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
      &copy; Copyright 2010, QIIME Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 0.6.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>