
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Index &mdash; Homepage</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.9.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="shortcut icon" href="_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="index.html" />
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="index.html"><img src="_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="#" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            

<h1 id="index">Index</h1>

<div class="genindex-jumpbox">
 <a href="#A"><strong>A</strong></a>
 | <a href="#B"><strong>B</strong></a>
 | <a href="#C"><strong>C</strong></a>
 | <a href="#D"><strong>D</strong></a>
 | <a href="#E"><strong>E</strong></a>
 | <a href="#F"><strong>F</strong></a>
 | <a href="#G"><strong>G</strong></a>
 | <a href="#I"><strong>I</strong></a>
 | <a href="#J"><strong>J</strong></a>
 | <a href="#L"><strong>L</strong></a>
 | <a href="#M"><strong>M</strong></a>
 | <a href="#N"><strong>N</strong></a>
 | <a href="#O"><strong>O</strong></a>
 | <a href="#P"><strong>P</strong></a>
 | <a href="#Q"><strong>Q</strong></a>
 | <a href="#R"><strong>R</strong></a>
 | <a href="#S"><strong>S</strong></a>
 | <a href="#T"><strong>T</strong></a>
 | <a href="#U"><strong>U</strong></a>
 | <a href="#V"><strong>V</strong></a>
 
</div>
<h2 id="A">A</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/add_alpha_to_mapping_file.html#index-0">add_alpha_to_mapping_file.py</a>
  </dt>

      
  <dt><a href="scripts/add_qiime_labels.html#index-0">add_qiime_labels.py</a>
  </dt>

      
  <dt><a href="scripts/adjust_seq_orientation.html#index-0">adjust_seq_orientation.py</a>
  </dt>

      
  <dt><a href="scripts/align_seqs.html#index-0">align_seqs.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/alpha_diversity.html#index-0">alpha_diversity.py</a>
  </dt>

      
  <dt><a href="scripts/alpha_rarefaction.html#index-0">alpha_rarefaction.py</a>
  </dt>

      
  <dt><a href="scripts/ampliconnoise.html#index-0">ampliconnoise.py</a>
  </dt>

      
  <dt><a href="scripts/assign_taxonomy.html#index-0">assign_taxonomy.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="B">B</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/beta_diversity.html#index-0">beta_diversity.py</a>
  </dt>

      
  <dt><a href="scripts/beta_diversity_through_plots.html#index-0">beta_diversity_through_plots.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/beta_significance.html#index-0">beta_significance.py</a>
  </dt>

      
  <dt><a href="scripts/blast_wrapper.html#index-0">blast_wrapper.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="C">C</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/categorized_dist_scatterplot.html#index-0">categorized_dist_scatterplot.py</a>
  </dt>

      
  <dt><a href="scripts/clean_raxml_parsimony_tree.html#index-0">clean_raxml_parsimony_tree.py</a>
  </dt>

      
  <dt><a href="scripts/cluster_quality.html#index-0">cluster_quality.py</a>
  </dt>

      
  <dt><a href="scripts/collapse_samples.html#index-0">collapse_samples.py</a>
  </dt>

      
  <dt><a href="scripts/collate_alpha.html#index-0">collate_alpha.py</a>
  </dt>

      
  <dt><a href="scripts/compare_alpha_diversity.html#index-0">compare_alpha_diversity.py</a>
  </dt>

      
  <dt><a href="scripts/compare_categories.html#index-0">compare_categories.py</a>
  </dt>

      
  <dt><a href="scripts/compare_distance_matrices.html#index-0">compare_distance_matrices.py</a>
  </dt>

      
  <dt><a href="scripts/compare_taxa_summaries.html#index-0">compare_taxa_summaries.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/compare_trajectories.html#index-0">compare_trajectories.py</a>
  </dt>

      
  <dt><a href="scripts/compute_core_microbiome.html#index-0">compute_core_microbiome.py</a>
  </dt>

      
  <dt><a href="scripts/compute_taxonomy_ratios.html#index-0">compute_taxonomy_ratios.py</a>
  </dt>

      
  <dt><a href="scripts/conditional_uncovered_probability.html#index-0">conditional_uncovered_probability.py</a>
  </dt>

      
  <dt><a href="scripts/consensus_tree.html#index-0">consensus_tree.py</a>
  </dt>

      
  <dt><a href="scripts/convert_fastaqual_fastq.html#index-0">convert_fastaqual_fastq.py</a>
  </dt>

      
  <dt><a href="scripts/core_diversity_analyses.html#index-0">core_diversity_analyses.py</a>
  </dt>

      
  <dt><a href="scripts/count_seqs.html#index-0">count_seqs.py</a>
  </dt>

      
  <dt><a href="scripts/cytoscape_usage.html#index-0">cytoscape</a>, <a href="tutorials/making_cytoscape_networks.html#index-0">[1]</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="D">D</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/demultiplex_fasta.html#index-0">demultiplex_fasta.py</a>
  </dt>

      
  <dt><a href="scripts/denoise_wrapper.html#index-0">denoise_wrapper.py</a>
  </dt>

      
  <dt><a href="scripts/denoiser.html#index-0">denoiser.py</a>
  </dt>

      
  <dt><a href="scripts/denoiser_preprocess.html#index-0">denoiser_preprocess.py</a>
  </dt>

      
  <dt><a href="scripts/denoiser_worker.html#index-0">denoiser_worker.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/detrend.html#index-0">detrend.py</a>
  </dt>

      
  <dt><a href="scripts/differential_abundance.html#index-0">differential_abundance.py</a>
  </dt>

      
  <dt><a href="scripts/dissimilarity_mtx_stats.html#index-0">dissimilarity_mtx_stats.py</a>
  </dt>

      
  <dt><a href="scripts/distance_matrix_from_mapping.html#index-0">distance_matrix_from_mapping.py</a>
  </dt>

      
  <dt><a href="documentation/index.html#index-0">Documentation</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="E">E</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/estimate_observation_richness.html#index-0">estimate_observation_richness.py</a>
  </dt>

      
  <dt><a href="scripts/exclude_seqs_by_blast.html#index-0">exclude_seqs_by_blast.py</a>
  </dt>

      
  <dt><a href="scripts/extract_barcodes.html#index-0">extract_barcodes.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/extract_reads_from_interleaved_file.html#index-0">extract_reads_from_interleaved_file.py</a>
  </dt>

      
  <dt><a href="scripts/extract_seqs_by_sample_id.html#index-0">extract_seqs_by_sample_id.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="F">F</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/filter_alignment.html#index-0">filter_alignment.py</a>
  </dt>

      
  <dt><a href="scripts/filter_distance_matrix.html#index-0">filter_distance_matrix.py</a>
  </dt>

      
  <dt><a href="scripts/filter_fasta.html#index-0">filter_fasta.py</a>
  </dt>

      
  <dt><a href="scripts/filter_otus_by_sample.html#index-0">filter_otus_by_sample.py</a>
  </dt>

      
  <dt><a href="scripts/filter_otus_from_otu_table.html#index-0">filter_otus_from_otu_table.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/filter_samples_from_otu_table.html#index-0">filter_samples_from_otu_table.py</a>
  </dt>

      
  <dt><a href="scripts/filter_taxa_from_otu_table.html#index-0">filter_taxa_from_otu_table.py</a>
  </dt>

      
  <dt><a href="scripts/filter_tree.html#index-0">filter_tree.py</a>
  </dt>

      
  <dt><a href="scripts/fix_arb_fasta.html#index-0">fix_arb_fasta.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="G">G</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/group_significance.html#index-0">group_significance.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="I">I</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/identify_chimeric_seqs.html#index-0">identify_chimeric_seqs.py</a>
  </dt>

      
  <dt><a href="scripts/identify_missing_files.html#index-0">identify_missing_files.py</a>
  </dt>

      
  <dt><a href="scripts/identify_paired_differences.html#index-0">identify_paired_differences.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/inflate_denoiser_output.html#index-0">inflate_denoiser_output.py</a>
  </dt>

      
  <dt><a href="install/install.html#index-0">Installing QIIME</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="J">J</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/jackknifed_beta_diversity.html#index-0">jackknifed_beta_diversity.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/join_paired_ends.html#index-0">join_paired_ends.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="L">L</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/load_remote_mapping_file.html#index-0">load_remote_mapping_file.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="M">M</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/make_2d_plots.html#index-0">make_2d_plots.py</a>
  </dt>

      
  <dt><a href="scripts/make_bipartite_network.html#index-0">make_bipartite_network.py</a>
  </dt>

      
  <dt><a href="scripts/make_bootstrapped_tree.html#index-0">make_bootstrapped_tree.py</a>
  </dt>

      
  <dt><a href="scripts/make_distance_boxplots.html#index-0">make_distance_boxplots.py</a>
  </dt>

      
  <dt><a href="scripts/make_distance_comparison_plots.html#index-0">make_distance_comparison_plots.py</a>
  </dt>

      
  <dt><a href="scripts/make_fastq.html#index-0">make_fastq.py</a>
  </dt>

      
  <dt><a href="scripts/make_library_id_lists.html#index-0">make_library_id_lists.py</a>
  </dt>

      
  <dt><a href="scripts/make_otu_heatmap.html#index-0">make_otu_heatmap.py</a>
  </dt>

      
  <dt><a href="scripts/make_otu_network.html#index-0">make_otu_network.py</a>
  </dt>

      
  <dt><a href="scripts/make_otu_table.html#index-0">make_otu_table.py</a>
  </dt>

      
  <dt><a href="scripts/make_per_library_sff.html#index-0">make_per_library_sff.py</a>
  </dt>

      
  <dt><a href="scripts/make_phylogeny.html#index-0">make_phylogeny.py</a>
  </dt>

      
  <dt><a href="scripts/make_prefs_file.html#index-0">make_prefs_file.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/make_qiime_py_file.html#index-0">make_qiime_py_file.py</a>
  </dt>

      
  <dt><a href="scripts/make_rarefaction_plots.html#index-0">make_rarefaction_plots.py</a>
  </dt>

      
  <dt><a href="scripts/make_tep.html#index-0">make_tep.py</a>
  </dt>

      
  <dt><a href="scripts/map_reads_to_reference.html#index-0">map_reads_to_reference.py</a>
  </dt>

      
  <dt><a href="scripts/merge_mapping_files.html#index-0">merge_mapping_files.py</a>
  </dt>

      
  <dt><a href="scripts/merge_otu_maps.html#index-0">merge_otu_maps.py</a>
  </dt>

      
  <dt><a href="scripts/merge_otu_tables.html#index-0">merge_otu_tables.py</a>
  </dt>

      
  <dt><a href="scripts/multiple_extract_barcodes.html#index-0">multiple_extract_barcodes.py</a>
  </dt>

      
  <dt><a href="scripts/multiple_join_paired_ends.html#index-0">multiple_join_paired_ends.py</a>
  </dt>

      
  <dt><a href="scripts/multiple_rarefactions.html#index-0">multiple_rarefactions.py</a>
  </dt>

      
  <dt><a href="scripts/multiple_rarefactions_even_depth.html#index-0">multiple_rarefactions_even_depth.py</a>
  </dt>

      
  <dt><a href="scripts/multiple_split_libraries_fastq.html#index-0">multiple_split_libraries_fastq.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="N">N</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/neighbor_joining.html#index-0">neighbor_joining.py</a>
  </dt>

      
  <dt><a href="scripts/nmds.html#index-0">nmds.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/normalize_table.html#index-0">normalize_table.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="O">O</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/observation_metadata_correlation.html#index-0">observation_metadata_correlation.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="P">P</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/parallel_align_seqs_pynast.html#index-0">parallel_align_seqs_pynast.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_alpha_diversity.html#index-0">parallel_alpha_diversity.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_assign_taxonomy_blast.html#index-0">parallel_assign_taxonomy_blast.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_assign_taxonomy_rdp.html#index-0">parallel_assign_taxonomy_rdp.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_assign_taxonomy_uclust.html#index-0">parallel_assign_taxonomy_uclust.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_beta_diversity.html#index-0">parallel_beta_diversity.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_blast.html#index-0">parallel_blast.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_identify_chimeric_seqs.html#index-0">parallel_identify_chimeric_seqs.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_map_reads_to_reference.html#index-0">parallel_map_reads_to_reference.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_merge_otu_tables.html#index-0">parallel_merge_otu_tables.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_multiple_rarefactions.html#index-0">parallel_multiple_rarefactions.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_pick_otus_blast.html#index-0">parallel_pick_otus_blast.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_pick_otus_sortmerna.html#index-0">parallel_pick_otus_sortmerna.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_pick_otus_trie.html#index-0">parallel_pick_otus_trie.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_pick_otus_uclust_ref.html#index-0">parallel_pick_otus_uclust_ref.py</a>
  </dt>

      
  <dt><a href="scripts/parallel_pick_otus_usearch61_ref.html#index-0">parallel_pick_otus_usearch61_ref.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/pick_closed_reference_otus.html#index-0">pick_closed_reference_otus.py</a>
  </dt>

      
  <dt><a href="scripts/pick_de_novo_otus.html#index-0">pick_de_novo_otus.py</a>
  </dt>

      
  <dt><a href="scripts/pick_open_reference_otus.html#index-0">pick_open_reference_otus.py</a>
  </dt>

      
  <dt><a href="scripts/pick_otus.html#index-0">pick_otus.py</a>
  </dt>

      
  <dt><a href="scripts/pick_rep_set.html#index-0">pick_rep_set.py</a>
  </dt>

      
  <dt><a href="scripts/plot_rank_abundance_graph.html#index-0">plot_rank_abundance_graph.py</a>
  </dt>

      
  <dt><a href="scripts/plot_semivariogram.html#index-0">plot_semivariogram.py</a>
  </dt>

      
  <dt><a href="scripts/plot_taxa_summary.html#index-0">plot_taxa_summary.py</a>
  </dt>

      
  <dt><a href="scripts/poller.html#index-0">poller.py</a>
  </dt>

      
  <dt><a href="scripts/principal_coordinates.html#index-0">principal_coordinates.py</a>
  </dt>

      
  <dt><a href="scripts/print_metadata_stats.html#index-0">print_metadata_stats.py</a>
  </dt>

      
  <dt><a href="scripts/print_qiime_config.html#index-0">print_qiime_config.py</a>
  </dt>

      
  <dt><a href="scripts/process_iseq.html#index-0">process_iseq.py</a>
  </dt>

      
  <dt><a href="scripts/process_qseq.html#index-0">process_qseq.py</a>
  </dt>

      
  <dt><a href="scripts/process_sff.html#index-0">process_sff.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="Q">Q</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/quality_scores_plot.html#index-0">quality_scores_plot.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="R">R</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/relatedness.html#index-0">relatedness.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="S">S</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/shared_phylotypes.html#index-0">shared_phylotypes.py</a>
  </dt>

      
  <dt><a href="scripts/simsam.html#index-0">simsam.py</a>
  </dt>

      
  <dt><a href="scripts/single_rarefaction.html#index-0">single_rarefaction.py</a>
  </dt>

      
  <dt><a href="scripts/sort_otu_table.html#index-0">sort_otu_table.py</a>
  </dt>

      
  <dt><a href="scripts/split_libraries.html#index-0">split_libraries.py</a>
  </dt>

      
  <dt><a href="scripts/split_libraries_fastq.html#index-0">split_libraries_fastq.py</a>
  </dt>

      
  <dt><a href="scripts/split_libraries_lea_seq.html#index-0">split_libraries_lea_seq.py</a>
  </dt>

      
  <dt><a href="scripts/split_otu_table.html#index-0">split_otu_table.py</a>
  </dt>

      
  <dt><a href="scripts/split_otu_table_by_taxonomy.html#index-0">split_otu_table_by_taxonomy.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/split_sequence_file_on_sample_ids.html#index-0">split_sequence_file_on_sample_ids.py</a>
  </dt>

      
  <dt><a href="scripts/start_parallel_jobs.html#index-0">start_parallel_jobs.py</a>
  </dt>

      
  <dt><a href="scripts/start_parallel_jobs_sc.html#index-0">start_parallel_jobs_sc.py</a>
  </dt>

      
  <dt><a href="scripts/start_parallel_jobs_slurm.html#index-0">start_parallel_jobs_slurm.py</a>
  </dt>

      
  <dt><a href="scripts/start_parallel_jobs_torque.html#index-0">start_parallel_jobs_torque.py</a>
  </dt>

      
  <dt><a href="scripts/subsample_fasta.html#index-0">subsample_fasta.py</a>
  </dt>

      
  <dt><a href="scripts/summarize_taxa.html#index-0">summarize_taxa.py</a>
  </dt>

      
  <dt><a href="scripts/summarize_taxa_through_plots.html#index-0">summarize_taxa_through_plots.py</a>
  </dt>

      
  <dt><a href="scripts/supervised_learning.html#index-0">supervised_learning.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="T">T</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/transform_coordinate_matrices.html#index-0">transform_coordinate_matrices.py</a>
  </dt>

      
  <dt><a href="scripts/tree_compare.html#index-0">tree_compare.py</a>
  </dt>

      
  <dt><a href="scripts/trflp_file_to_otu_table.html#index-0">trflp_file_to_otu_table.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/trim_sff_primers.html#index-0">trim_sff_primers.py</a>
  </dt>

      
  <dt><a href="scripts/truncate_fasta_qual_files.html#index-0">truncate_fasta_qual_files.py</a>
  </dt>

      
  <dt><a href="scripts/truncate_reverse_primer.html#index-0">truncate_reverse_primer.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="U">U</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/unweight_fasta.html#index-0">unweight_fasta.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/upgma_cluster.html#index-0">upgma_cluster.py</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="V">V</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/validate_demultiplexed_fasta.html#index-0">validate_demultiplexed_fasta.py</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="scripts/validate_mapping_file.html#index-0">validate_mapping_file.py</a>
  </dt>

  </dl></td>
</tr></table>



          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    

    <h3><a href="index.html">Site index</a></h3>
    <ul><li><ul>
    <li><a href="http://www.qiime.org">Home</a><br /></li>
    <li><a href="install/index.html">Install</a><br /></li>
    <li><a href="tutorials/index.html">Tutorials</a><br /></li>
    <li><a href="scripts/index.html">Scripts</a><br /></li>
    <li><a href="http://help.qiime.org">Help</a><br /></li>
    <li><a href="http://qiime.org/home_static/dataFiles.html">Resources</a><br /></li>
    <li><a href="documentation/index.html">File Formats</a><br /></li>
    <li><a href="http://workshops.qiime.org">Workshops</a><br /></li>
    <li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
    <li><a href="developer/index.html">Developer</a><br /></li>
    </ul></li></ul>



   

<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="#" title="General Index"
             >index</a></li>
        <li><a href="index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.scikit-bio.org" target="_blank"><img src="_static/powered_by_skbio.png" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2010-2014, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>