<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Performing Procrustes Analysis &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.7.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="performing-procrustes-analysis">
<span id="procrustes-analysis"></span><h1>Performing Procrustes Analysis<a class="headerlink" href="#performing-procrustes-analysis" title="Permalink to this headline">¶</a></h1>
<p>This tutorial covers how to perform Procrustes Analysis using QIIME to compare UniFrac PCoA plots generated by two different processing pipelines. Procrustes analysis takes as input two coordinate matrices with corresponding points (in QIIME, these are generated by running <tt class="docutils literal"><span class="pre">principal_coordinates.py</span></tt> on a distance matrix generated by <tt class="docutils literal"><span class="pre">beta_diversity.py</span></tt>), and transforming the second coordinate set by rotating, scaling, and translating it to minimize the distances between corresponding points in the two shapes. This is done with <tt class="docutils literal"><span class="pre">transform_coordinate_matrices.py</span></tt>. The results can then be visualized using QIIME by running <tt class="docutils literal"><span class="pre">compare_3d_plots.py</span></tt>. Both sets of coordinates will be plotted in the resulting figure, with bars connecting the corresponding points from each data set.</p>
<p>For example, Figure 1 is a plot that was generated by comparing two different sets of reads, one set generated on 454 and the other generated on Illumina, from the same collection of samples. In this scenario, Procrustes analysis allows us to determine whether we would derive the same beta diversity conclusions, regardless of which technology was used to sequence the amplicons. This tutorial illustrates the steps used to generate this plot, beginning with closed reference OTU picking on both the 454 and the Illumina data. The data sets used in this tutorial are a subset of the data that was used in Supplementary Figure 1 of <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/21624126">Moving Pictures of the Human Microbiome</a>.</p>
<div class="figure">
<a class="reference internal image-reference" href="../_images/procrustes_tutorial_figure1.png"><img alt="../_images/procrustes_tutorial_figure1.png" src="../_images/procrustes_tutorial_figure1.png" /></a>
<p class="caption">Figure 1: Procrustes comparison of unweighted UniFrac PCoA plots derived from 454 and Illumina sequencing of the same samples from three human-associated habitats (Skin: reds/blues; Feces: greens; Tongue: oranges) reads from a paired-end Illumina run.</p>
</div>
<p>p-values are generated using a Monte Carlo simulation: sample identifiers are shuffled in one of the PC matrices, and the M2 value is re-computed <tt class="docutils literal"><span class="pre">--random_trials</span></tt> times. The proportion of M2 values that are equal to or lower than the actual M2 value is the Monte Carlo p-value.</p>
<div class="section" id="initial-setup">
<h2>Initial setup<a class="headerlink" href="#initial-setup" title="Permalink to this headline">¶</a></h2>
<p>This tutorial works through several steps in QIIME, culminating in a Procrustes analysis of a subset of the data presented in <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/21624126">Moving Pictures of the Human Microbiome</a>. This analysis compares beta diversity PCoA plots generated from 454 and Illumina sequences to test the hypothesis that the same beta diversity conclusions are consistent regardless of sequencing technology. For the sake of runtime, this tutorial is based on approximately 60 samples from each sequencing technologies, and a small subset of the total number of reads for each of these samples.</p>
<p>Do some initial set up for the tutorial. This version of the tutorial is based on running the QIIME 1.4.0 EC2 instance. You can always find <a class="reference external" href="http://qiime.org/home_static/dataFiles.html">the latest QIIME AMIs here</a>.</p>
<div class="highlight-python"><pre>cd
wget https://s3.amazonaws.com/s3-qiime_tutorial_files/moving_pictures_procrustes_demo.tgz
tar -xzf moving_pictures_procrustes_demo.tgz
screen
cd moving_pictures_procrustes_demo</pre>
</div>
</div>
<div class="section" id="defining-reference-filepaths-with-environment-variables">
<h2>Defining reference filepaths with environment variables<a class="headerlink" href="#defining-reference-filepaths-with-environment-variables" title="Permalink to this headline">¶</a></h2>
<p>Through-out this tutorial we make use of a reference sequence collection, tree, and taxonomy derived from the Greengenes database. As these files may be store in different locations on your system, we&#8217;ll define them as environment variables using the paths as they would be if you&#8217;re running in a QIIME virtual machine (e.g., on AWS or with the Virtual Box). We&#8217;ll then reference the environment variables through-out this tutorial when they are used. If you&#8217;re not working on either of these systems, you&#8217;ll have to modify these paths. Run the following:</p>
<div class="highlight-python"><pre>export QIIME_DIR=$HOME/qiime_software
export reference_seqs $QIIME_DIR/gg_otus-4feb2011-release/rep_set/gg_97_otus_4feb2011.fasta
export reference_tree $QIIME_DIR/gg_otus-4feb2011-release/trees/gg_97_otus_4feb2011.tre
export reference_tax $QIIME_DIR/gg_otus-4feb2011-release/taxonomies/greengenes_tax.txt</pre>
</div>
<p>Pick OTUs on Illumina data and generate an OTU table (including taxonomic assignment of samples):</p>
<div class="highlight-python"><pre>pick_closed_reference_otus.py -i ./subsampled_illumina_seqs.fna -o ./illumina_ucrC/ -r $reference_seqs -t $reference_tax -aO8 -p ./otu_params.txt</pre>
</div>
<p>Determine the number of sequences per sample and related statistics. You&#8217;ll want to choose an even sampling depth for the beta diversity analysis from these data. In this tutorial we choose the smallest number of sequences per sample (239).</p>
<div class="highlight-python"><pre>print_biom_table_summary.py -i ./illumina_ucrC/uclust_ref_picked_otus/otu_table.biom</pre>
</div>
<p>Compute UniFrac distances between samples, run principal coordinates analysis, and build 3D PCoA plots:</p>
<div class="highlight-python"><pre>beta_diversity_through_plots.py -i ./illumina_ucrC/uclust_ref_picked_otus/otu_table.biom -e 239 -o ./illumina_ucrC/bdiv_even239/ -t $reference_tree -m ./illumina_map.txt -aO8 -p ./bdiv_params.txt --suppress_2d_plots</pre>
</div>
<p>Repeat the above steps on the 454 data:</p>
<div class="highlight-python"><pre>pick_closed_reference_otus.py -i ./subsampled_454_seqs.fna -o ./454_ucrC/ -r $reference_seqs -t $reference_tax -aO8 -p ./otu_params.txt
print_biom_table_summary.py -i ./454_ucrC/uclust_ref_picked_otus/otu_table.biom
beta_diversity_through_plots.py -i ./454_ucrC/uclust_ref_picked_otus/otu_table.biom -e 135 -o ./454_ucrC/bdiv_even135/ -t $reference_tree -m ./454_map.txt -aO8 -p ./bdiv_params.txt --suppress_2d_plots</pre>
</div>
<p>Perform Procrustes analysis:</p>
<div class="highlight-python"><pre>transform_coordinate_matrices.py -i ./illumina_ucrC/bdiv_even239/unweighted_unifrac_pc.txt,./454_ucrC/bdiv_even135/unweighted_unifrac_pc.txt -s ./procrustes_sid_map.txt -r 100 -o ./454_v_illumina/</pre>
</div>
<p>Generate Procrustes plot, including an explicit time axis:</p>
<div class="highlight-python"><pre>compare_3d_plots.py -i ./454_v_illumina/pc1_transformed.txt,./454_v_illumina/pc2_transformed.txt -o ./454_v_illumina/plots/ -m ./procrustes_metadata_map.txt --custom_axes days_since_epoch</pre>
</div>
<p>There will now be several results of interest. For the Procrustes analysis you can find the statistical results in <tt class="docutils literal"><span class="pre">./454_v_illumina/unweighted_unifrac_pc_unweighted_unifrac_pc_procrustes_results.txt</span></tt> and you can view the Procrustes plot by opening <tt class="docutils literal"><span class="pre">./454_v_illumina/plots/pc1_transformed_3D_PCoA_plots.html</span></tt> in a web browser.</p>
</div>
<div class="section" id="comparing-data-sets-with-different-sample-ids">
<h2>Comparing data sets with different sample ids<a class="headerlink" href="#comparing-data-sets-with-different-sample-ids" title="Permalink to this headline">¶</a></h2>
<p>In the cases described here, we always have the same samples in our two principal coordinate matrices. If that is not the case for your study, you&#8217;ll need to pass a sample id mapping file (<strong>which is different from a QIIME metadata mapping file</strong>). For a description of this file format, see <a class="reference internal" href="../documentation/file_formats.html#sample-id-map"><em>Sample id map</em></a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Performing Procrustes Analysis</a><ul>
<li><a class="reference internal" href="#initial-setup">Initial setup</a></li>
<li><a class="reference internal" href="#defining-reference-filepaths-with-environment-variables">Defining reference filepaths with environment variables</a></li>
<li><a class="reference internal" href="#comparing-data-sets-with-different-sample-ids">Comparing data sets with different sample ids</a></li>
</ul>
</li>
</ul>


	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="../install/index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.pycogent.org" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2012, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>