<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Denoising and chimera detection usage differences in QIIME &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />
    <link rel="up" title="QIIME Tutorials" href="index.html" />
    <link rel="next" title="Taxonomic classifications of single- and paired-end sequences with RTAX" href="rtax.html" />
    <link rel="prev" title="Using usearch quality filtering with QIIME" href="usearch_quality_filter.html" />
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="rtax.html" title="Taxonomic classifications of single- and paired-end sequences with RTAX"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="usearch_quality_filter.html" title="Using usearch quality filtering with QIIME"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="denoising-and-chimera-detection-usage-differences-in-qiime">
<span id="denoising-and-chimera-detection-usage-comparison-rst"></span><h1>Denoising and chimera detection usage differences in QIIME<a class="headerlink" href="#denoising-and-chimera-detection-usage-differences-in-qiime" title="Permalink to this headline">¶</a></h1>
<p>This tutorial covers some of the main differences in the utilization of the various denoising and chimera detection software implemented in QIIME.</p>
<p>The <a class="reference external" href="tutorial.html">overview tutorial</a> describes the steps one would use to process 454 data without denoising or chimera detection.  The data processing can be roughly summarized as the following:</p>
<ol class="arabic simple">
<li>SFF (raw 454 data) -&gt; 2.  fasta/qual files -&gt; 3.  demultiplexing/quality filtering -&gt; 4.  OTU picking -&gt; 5.  representative sequences -&gt; 6.  taxonomic assignments/tree building -&gt; 7.  OTU table and downstream processing</li>
</ol>
<p>Differences from the default pipeline listed above will be used to describe how each denoising/chimera detection software integrates into the QIIME software package.</p>
<div class="section" id="ampliconnoise">
<h2>Ampliconnoise<a class="headerlink" href="#ampliconnoise" title="Permalink to this headline">¶</a></h2>
<p>Ampliconnoise uses flowgram files generated from SFF files to denoise 454 data and optionally detect chimeras.  See script details here:  <a class="reference external" href="../scripts/ampliconnoise.html">ampliconnoise.py</a></p>
<p>Ampliconnoise effectively replaces the demultiplexing/quality filtering step above, making the pipeline this:</p>
<ol class="arabic simple">
<li>SFF (raw 454 data) -&gt; 2.  flowgram (sff.txt) -&gt; 3.  ampliconnoise.py (plus suggested step of reverse primer removal) -&gt; 4.  OTU picking -&gt; 5.  representative sequences -&gt; 6.  taxonomic assignments/tree building -&gt; 7.  OTU table and downstream processing</li>
</ol>
<p>Barcodes and forward primers are removed by ampliconnoise.py, however, reverse primers at the end of the sequence may be retained, so it is strongly recommended that <a class="reference external" href="../scripts/truncate_reverse_primer.html">truncate_reverse_primer.py</a> be run immediately after ampliconnoise.py so the reverse primer and subsequent sequence does not interfere with downstream steps.</p>
</div>
<div class="section" id="denoiser">
<h2>Denoiser<a class="headerlink" href="#denoiser" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="denoising_454_data.html">Denoiser</a> also utilizes flowgram files to detect and correct sequencing errors (but not chimeras).  However, it utilizes the output of split_libraries.py to limit the sequences tested to those present in the output fasta file generated by split_libraries.py.  Reverse primer removal with <a class="reference external" href="../scripts/truncate_reverse_primer.html">truncate_reverse_primer.py</a> is also strongly encouraged.  The steps involved are:</p>
<ol class="arabic simple">
<li>SFF (raw 454 data) -&gt; 2.  fasta/qual/flowgram (sff.txt) files -&gt; 3.  split_libraries.py -&gt; 4.  denoise_wrapper.py (plus suggested step of reverse primer removal) -&gt; 4.  OTU picking -&gt; 5.  representative sequences -&gt; 6.  taxonomic assignments/tree building -&gt; 7.  OTU table and downstream processing</li>
</ol>
</div>
<div class="section" id="usearch">
<h2>USEARCH<a class="headerlink" href="#usearch" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="usearch_quality_filter.html">Usearch</a> uses cluster abundance for <cite>de novo</cite> chimera detection, a reference sequence set for reference based chimera detection, and a cluster size filtering step (which is similar to filtering singletons as a rough but fast way to remove noise from data), and clusters sequences into OTUs.  Usearch is used after demultiplexing sequences, so the steps for processing data are:</p>
<ol class="arabic simple">
<li>SFF (raw 454 data) -&gt; 2.  fasta/qual files -&gt; 3.  demultiplexing/quality filtering -&gt; 4.  OTU picking/chimera detection/low abundance cluster filtering with usearch implementation in pick_otus.py -&gt; 5.  representative sequences -&gt; 6.  taxonomic assignments/tree building -&gt; 7.  OTU table and downstream processing</li>
</ol>
</div>
<div class="section" id="chimeraslayer">
<h2>ChimeraSlayer<a class="headerlink" href="#chimeraslayer" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="chimera_checking.html">ChimeraSlayer</a> utilizes a reference dataset to detect potential chimeras in a representative sequence set.  The processing pipeline is:</p>
<ol class="arabic simple">
<li>SFF (raw 454 data) -&gt; 2.  fasta/qual files -&gt; 3.  demultiplexing/quality filtering -&gt; 4.  OTU picking -&gt; 5.  representative sequences -&gt; 6.  Chimera detection with identify_chimeric_seqs.py -&gt; 7.  Filter chimeras as described <a class="reference external" href="chimera_checking.html">here</a>.  -&gt; 8.  taxonomic assignments/tree building -&gt; 9.  OTU table and downstream processing</li>
</ol>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Denoising and chimera detection usage differences in QIIME</a><ul>
<li><a class="reference internal" href="#ampliconnoise">Ampliconnoise</a></li>
<li><a class="reference internal" href="#denoiser">Denoiser</a></li>
<li><a class="reference internal" href="#usearch">USEARCH</a></li>
<li><a class="reference internal" href="#chimeraslayer">ChimeraSlayer</a></li>
</ul>
</li>
</ul>


    <h3><a href="../index.html">Site index</a></h3>
    <ul><li><ul>
    <li><a href="http://www.qiime.org">Home</a><br /></li>
    <li><a href="../install/index.html">Install</a><br /></li>
    <li><a href="index.html">Tutorials</a><br /></li>
    <li><a href="../scripts/index.html">Scripts</a><br /></li>
    <li><a href="http://help.qiime.org">Help</a><br /></li>
    <li><a href="http://qiime.org/home_static/dataFiles.html">Resources</a><br /></li>
    <li><a href="../documentation/index.html">File Formats</a><br /></li>
    <li><a href="http://workshops.qiime.org">Workshops</a><br /></li>
    <li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
    <li><a href="../developer/index.html">Developer</a><br /></li>
    </ul></li></ul>


  <h4>Previous topic</h4>
  <p class="topless"><a href="usearch_quality_filter.html"
                        title="previous chapter">Using usearch quality filtering with QIIME</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="rtax.html"
                        title="next chapter">Taxonomic classifications of single- and paired-end sequences with RTAX</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="rtax.html" title="Taxonomic classifications of single- and paired-end sequences with RTAX"
             >next</a> |</li>
        <li class="right" >
          <a href="usearch_quality_filter.html" title="Using usearch quality filtering with QIIME"
             >previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" >QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.scikit-bio.org" target="_blank"><img src="../_static/powered_by_skbio.png" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2010-2014, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>