<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Chimera checking sequences with QIIME &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />
    <link rel="up" title="QIIME Tutorials" href="index.html" />
    <link rel="next" title="Using usearch quality filtering with QIIME" href="usearch_quality_filter.html" />
    <link rel="prev" title="Denoising of 454 Data Sets" href="denoising_454_data.html" />
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="usearch_quality_filter.html" title="Using usearch quality filtering with QIIME"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="denoising_454_data.html" title="Denoising of 454 Data Sets"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="chimera-checking-sequences-with-qiime">
<span id="chimera-checking"></span><h1>Chimera checking sequences with QIIME<a class="headerlink" href="#chimera-checking-sequences-with-qiime" title="Permalink to this headline">¶</a></h1>
<div class="section" id="chimeraslayer">
<h2>ChimeraSlayer<a class="headerlink" href="#chimeraslayer" title="Permalink to this headline">¶</a></h2>
<p>This document describes how to chimera check sequences with QIIME via ChimeraSlayer, and subsequently exclude chimeric sequences when building your tree and OTU table so they will not be included in downstream analyses. (Note that because we recommend applying chimera checking after OTU and representative sequence picking, for the purposes of this document <cite>chimeric sequence</cite> should be considered synonymous to <cite>chimeric OTU</cite>.)</p>
<p>ChimeraSlayer requires aligned sequences, you should apply chimera checking after you&#8217;ve aligned your sequences with PyNAST, but before you apply the lanemask filtering to your alignment.</p>
<p>Applying ChimeraSlayer is done as follows:</p>
<div class="highlight-python"><div class="highlight"><pre>identify_chimeric_seqs.py -m ChimeraSlayer -i rep_set_aligned.fasta -a reference_set_aligned.fasta -o chimeric_seqs.txt
</pre></div>
</div>
<p>Before building your phylogenetic tree, you should now remove chimeric sequences from your alignment using your chimeric sequence list (<tt class="docutils literal"><span class="pre">chimeric_seqs.txt</span></tt> in this example) and the alignment to as follows:</p>
<div class="highlight-python"><div class="highlight"><pre>filter_fasta.py -f rep_set_aligned.fasta -o non_chimeric_rep_set_aligned.fasta -s chimeric_seqs.txt -n
</pre></div>
</div>
<p>Don&#8217;t forget to pass the <tt class="docutils literal"><span class="pre">-n</span></tt> parameter to <tt class="docutils literal"><span class="pre">filter_fasta.py</span></tt> &#8211; this tells <tt class="docutils literal"><span class="pre">filter_fasta.py</span></tt> to discard the sequences you&#8217;ve passed via <tt class="docutils literal"><span class="pre">-s</span></tt>, rather than to keep only those sequences. Then pass the resulting subalignment (<tt class="docutils literal"><span class="pre">non_chimeric_rep_set_aligned.fasta</span></tt>) in the downstream steps <tt class="docutils literal"><span class="pre">filter_alignment.py</span></tt> prior to tree-building.</p>
<p>You&#8217;ll also need to exclude the chimeric sequences at the <tt class="docutils literal"><span class="pre">make_otu_table.py</span></tt> step. To do that, you pass the chimeric sequence list to <tt class="docutils literal"><span class="pre">make_otu_table.py</span></tt> via the <tt class="docutils literal"><span class="pre">-e</span></tt> parameter:</p>
<div class="highlight-python"><div class="highlight"><pre>make_otu_table.py -i otu_map.txt -o otu_table.biom -e chimeric_seqs.txt -t taxonomy.txt
</pre></div>
</div>
<p>This will cause the chimeric sequences to be excluded from all downstream analysis that you&#8217;ll be performing, as the OTU table is always the input provided (for example to beta diversity, alpha rarefaction, etc.).</p>
</div>
<div class="section" id="usearch-6-1">
<h2>USEARCH 6.1<a class="headerlink" href="#usearch-6-1" title="Permalink to this headline">¶</a></h2>
<p>USEARCH 6.1 performs reference based chimera detection, like ChimeraSlayer, but also can perform de novo chimera detection based upon abundances of input sequences.</p>
<p>Important notes about USEARCH 6.1 usage:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li><ol class="first arabic">
<li>Input sequences should be demultiplexed sequences (e.g. output of <tt class="docutils literal"><span class="pre">split_libraries.py</span></tt>) that are not already clustered.</li>
</ol>
</li>
<li><ol class="first arabic" start="2">
<li>The reference database should not be aligned.</li>
</ol>
</li>
<li><ol class="first arabic" start="3">
<li>The reference sequences need to be in the same orientation as the query sequences.  Use <tt class="docutils literal"><span class="pre">adjust_seq_orientation.py</span></tt> to reverse complement your reads if needed.</li>
</ol>
</li>
<li><ol class="first arabic" start="4">
<li>Chimera checking should be done first, followed by filtering chimeras out of the input reads, and these filtered sequences can then be clustered with <tt class="docutils literal"><span class="pre">pick_otus.py</span></tt>.</li>
</ol>
</li>
</ul>
</div>
<p>An example step by step process for removing chimeras with USEARCH 6.1 starts as follows, using the seqs.fna file (output of <tt class="docutils literal"><span class="pre">split_libraries.py</span></tt>) as the input sequence file:</p>
<div class="highlight-python"><div class="highlight"><pre>identify_chimeric_seqs.py -i seqs.fna -m usearch61 -o usearch_checked_chimeras/ -r gg_97_otus_4feb2011.fasta
</pre></div>
</div>
<p>Next, filter the input seqs.fna file by passing the chimeras.txt file created in the previous step and specifying that these should be removed with the -n option:</p>
<div class="highlight-python"><div class="highlight"><pre>filter_fasta.py -f seqs.fna -o seqs_chimeras_filtered.fna -s usearch_checked_chimeras/chimeras.txt -n
</pre></div>
</div>
<p>Finally, use the <tt class="docutils literal"><span class="pre">pick_otus.py</span></tt> to cluster these sequences using usearch61:</p>
<div class="highlight-python"><div class="highlight"><pre>pick_otus.py -m usearch61 -i seqs_chimeras_filtered.fna -o usearch61_picked_otus/
</pre></div>
</div>
<p>The remainder of the processing is the same as described in the main <a class="reference external" href="./tutorial.html#step-2-pick-representative-sequences-for-each-otu">tutorial</a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Chimera checking sequences with QIIME</a><ul>
<li><a class="reference internal" href="#chimeraslayer">ChimeraSlayer</a></li>
<li><a class="reference internal" href="#usearch-6-1">USEARCH 6.1</a></li>
</ul>
</li>
</ul>


    <h3><a href="../index.html">Site index</a></h3>
    <ul><li><ul>
    <li><a href="http://www.qiime.org">Home</a><br /></li>
    <li><a href="../install/index.html">Install</a><br /></li>
    <li><a href="index.html">Tutorials</a><br /></li>
    <li><a href="../scripts/index.html">Scripts</a><br /></li>
    <li><a href="http://help.qiime.org">Help</a><br /></li>
    <li><a href="http://qiime.org/home_static/dataFiles.html">Resources</a><br /></li>
    <li><a href="../documentation/index.html">File Formats</a><br /></li>
    <li><a href="http://workshops.qiime.org">Workshops</a><br /></li>
    <li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
    <li><a href="../developer/index.html">Developer</a><br /></li>
    </ul></li></ul>


  <h4>Previous topic</h4>
  <p class="topless"><a href="denoising_454_data.html"
                        title="previous chapter">Denoising of 454 Data Sets</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="usearch_quality_filter.html"
                        title="next chapter">Using usearch quality filtering with QIIME</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="usearch_quality_filter.html" title="Using usearch quality filtering with QIIME"
             >next</a> |</li>
        <li class="right" >
          <a href="denoising_454_data.html" title="Denoising of 454 Data Sets"
             >previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" >QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.scikit-bio.org" target="_blank"><img src="../_static/powered_by_skbio.png" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2010-2014, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>