<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pick_open_reference_otus.py – &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.7.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="pick-open-reference-otus"></span><div class="section" id="pick-open-reference-otus-py">
<span id="index-0"></span><h1><em>pick_open_reference_otus.py</em> &#8211;<a class="headerlink" href="#pick-open-reference-otus-py" title="Permalink to this headline">¶</a></h1>
<p><strong>Description:</strong></p>
<p><strong>Usage:</strong> <tt class="file docutils literal"><span class="pre">pick_open_reference_otus.py</span> <span class="pre">[options]</span></tt></p>
<p><strong>Input Arguments:</strong></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><strong>[REQUIRED]</strong></p>
<dl class="docutils">
<dt>-i, <cite>-</cite>-input_fps</dt>
<dd>The input sequences filepath or comma-separated list of filepaths</dd>
<dt>-r, <cite>-</cite>-reference_fp</dt>
<dd>The reference sequences</dd>
<dt>-o, <cite>-</cite>-output_dir</dt>
<dd>The output directory</dd>
</dl>
<p><strong>[OPTIONAL]</strong></p>
<dl class="last docutils">
<dt>-m, <cite>-</cite>-otu_picking_method</dt>
<dd>The OTU picking method to use for reference and de novo steps. Passing usearch61, for example, means that usearch61 will be used for the de novo steps and usearch61_ref will be used for reference steps. [default: uclust]</dd>
<dt>-p, <cite>-</cite>-parameter_fp</dt>
<dd>Path to the parameter file, which specifies changes to the default behavior. See <a class="reference external" href="http://www.qiime.org/documentation/file_formats.html#qiime-parameters">http://www.qiime.org/documentation/file_formats.html#qiime-parameters</a> . [if omitted, default values will be used]</dd>
<dt><cite>-</cite>-prefilter_refseqs_fp</dt>
<dd>The reference sequences to use for the prefilter, if different from the reference sequecnces to use for the OTU picking [default: same as passed for &#8211;reference_fp]</dd>
<dt>-n, <cite>-</cite>-new_ref_set_id</dt>
<dd>Unique identifier for OTUs that get created in this ref set (this is useful to support combining of reference sets) [default:New]</dd>
<dt>-f, <cite>-</cite>-force</dt>
<dd>Force overwrite of existing output directory (note: existing files in output_dir will not be removed) [default: None]</dd>
<dt>-a, <cite>-</cite>-parallel</dt>
<dd>Run in parallel where available [default: False]</dd>
<dt>-O, <cite>-</cite>-jobs_to_start</dt>
<dd>Number of jobs to start. NOTE: you must also pass -a to run in parallel, this defines the number of jobs to be started if and only if -a is passed [default: 2]</dd>
<dt>-s, <cite>-</cite>-percent_subsample</dt>
<dd>Percent of failure sequences to include in the subsample to cluster de novo (larger numbers should give more comprehensive ,results but will be slower) [default:0.001]</dd>
<dt><cite>-</cite>-prefilter_percent_id</dt>
<dd>Sequences are pre-clustered at this percent id against the reference and any reads which fail to hit are discarded (a quality filter); pass 0.0 to disable [default:0.60]</dd>
<dt><cite>-</cite>-step1_otu_map_fp</dt>
<dd>Reference OTU picking OTU map  (to avoid rebuilding if one has already been built)</dd>
<dt><cite>-</cite>-step1_failures_fasta_fp</dt>
<dd>Reference OTU picking failures fasta filepath  (to avoid rebuilding if one has already been built)</dd>
<dt><cite>-</cite>-suppress_step4</dt>
<dd>Suppress the final de novo OTU picking step  (may be necessary for extremely large data sets) [default: False]</dd>
<dt><cite>-</cite>-min_otu_size</dt>
<dd>The minimum otu size (in number of sequences) to retain the otu [default: 2]</dd>
<dt><cite>-</cite>-suppress_taxonomy_assignment</dt>
<dd>Skip the taxonomy assignment step, resulting in an OTU table without taxonomy [default: False]</dd>
<dt><cite>-</cite>-suppress_align_and_tree</dt>
<dd>Skip the sequence alignment and tree-building steps [default: False]</dd>
</dl>
</div>
<p><strong>Output:</strong></p>
<p>Run the subsampled open-reference OTU picking workflow on seqs1.fna using refseqs.fna as the reference collection. ALWAYS SPECIFY ABSOLUTE FILE PATHS (absolute path represented here as $PWD, but will generally look something like /home/ubuntu/my_analysis/</p>
<div class="highlight-python"><pre>pick_open_reference_otus.py -i $PWD/seqs1.fna -r $PWD/refseqs.fna -o $PWD/ucrss/ -s 0.1 -p $PWD/ucrss_params.txt</pre>
</div>
<p>Run the subsampled open-reference OTU picking workflow on seqs1.fna using refseqs.fna as the reference collection and using usearch61 and usearch61_ref as the OTU picking methods. ALWAYS SPECIFY ABSOLUTE FILE PATHS (absolute path represented here as $PWD, but will generally look something like /home/ubuntu/my_analysis/</p>
<div class="highlight-python"><pre>pick_open_reference_otus.py -i $PWD/seqs1.fna -r $PWD/refseqs.fna -o $PWD/ucrss_usearch/ -s 0.1 -p $PWD/ucrss_params.txt -m usearch61</pre>
</div>
<p>Run the subsampled open-reference OTU picking workflow in iterative mode on seqs1.fna and seqs2.fna using refseqs.fna as the initial reference collection. ALWAYS SPECIFY ABSOLUTE FILE PATHS (absolute path represented here as $PWD, but will generally look something like /home/ubuntu/my_analysis/</p>
<div class="highlight-python"><pre>pick_open_reference_otus.py -i $PWD/seqs1.fna,$PWD/seqs2.fna -r $PWD/refseqs.fna -o $PWD/ucrss_iter/ -s 0.1 -p $PWD/ucrss_params.txt</pre>
</div>
<p>Run the subsampled open-reference OTU picking workflow in iterative mode on seqs1.fna and seqs2.fna using refseqs.fna as the initial reference collection. This is useful if you&#8217;re working with marker genes that do not result in useful alignment (e.g., fungal ITS). ALWAYS SPECIFY ABSOLUTE FILE PATHS (absolute path represented here as $PWD, but will generally look something like /home/ubuntu/my_analysis/</p>
<div class="highlight-python"><pre>pick_open_reference_otus.py -i $PWD/seqs1.fna,$PWD/seqs2.fna -r $PWD/refseqs.fna -o $PWD/ucrss_iter_no_tree/ -s 0.1 -p $PWD/ucrss_params.txt --suppress_align_and_tree</pre>
</div>
<p>Run the subsampled open-reference OTU picking workflow in iterative mode on seqs1.fna and seqs2.fna using refseqs.fna as the initial reference collection, suppressing assignment of taxonomy. This is useful if you&#8217;re working with a reference collection without associated taxonomy. ALWAYS SPECIFY ABSOLUTE FILE PATHS (absolute path represented here as $PWD, but will generally look something like /home/ubuntu/my_analysis/</p>
<div class="highlight-python"><pre>pick_open_reference_otus.py -i $PWD/seqs1.fna,$PWD/seqs2.fna -r $PWD/refseqs.fna -o $PWD/ucrss_iter_no_tax/ -s 0.1 -p $PWD/ucrss_params.txt --suppress_taxonomy_assignment</pre>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    

	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="../install/index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="../tutorials/index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.pycogent.org" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2012, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>