<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>SRA Submission &mdash; Homepage</title>
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.2.0',
        COLLAPSE_MODINDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="doc-sra-submission"></span><div class="section" id="sra-submission">
<span id="index-104"></span><h1>SRA Submission<a class="headerlink" href="#sra-submission" title="Permalink to this headline">¶</a></h1>
<p>QIIME&#8217;s SRA submission code is changing rapidly in the <tt class="docutils literal"><span class="pre">1.2.0-dev</span></tt> versions of QIIME. We&#8217;re making every effort to keep the documentation up-to-date with the code, but if you are interested in using this code you&#8217;ll most likely want to get on the <tt class="docutils literal"><span class="pre">qiime-sra-users</span></tt> mailing list to stay current on the state of the code. Head <a class="reference external" href="http://groups.google.com/group/qiime-sra-users?hl=en">here</a> for information on joining that public group.</p>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This tutorial illustrates how to apply QIIME to submit barcoded 16S community sequencing data to the SRA.</p>
<p><strong>To submit your own data to SRA, see</strong> <a class="reference external" href="../documentation/prepare_and_submit_sra.html">this document</a>. This tutorial focuses on a specific example data set.</p>
<p>To run this tutorial, you&#8217;ll need to have QIIME installed, as well as the 454 off-instrument tools (<tt class="docutils literal"><span class="pre">sffinfo</span></tt>, <tt class="docutils literal"><span class="pre">sfffile</span></tt>). Future versions of QIIME will likely support similar workflows and tutorials for SRA submission of meta-genomics data, and data generated on the Illumina platforms.</p>
<p>This tutorial uses as an example the <a class="reference external" href="http://www.pnas.org/content/105/46/17994.long">Fierer et al. 2008 hand dataset</a>. This represents the first barcoded SRA submission, and the finished product can be found via SRA accession #: <a class="reference external" href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=sra&amp;term=SRS001216">SRS001216</a>.</p>
<p>To get started, download the data from: <a class="reference external" href="http://bmf.colorado.edu/QIIME/knight_handstudy_demo-v1.2.0-dev.zip">SRA Tutorial Data</a></p>
</div>
<div class="section" id="philosophy-of-qiime-s-sra-submission-code">
<h2>Philosophy of QIIME&#8217;s SRA submission code<a class="headerlink" href="#philosophy-of-qiime-s-sra-submission-code" title="Permalink to this headline">¶</a></h2>
<p>This document and code does not seek to capture the full complexity of what is possible in the SRA. Instead, the goal is to provide a simple pathway for submission to the SRA of the most common types of data. For example, more normalization could be provided through requiring additional tables, but this additional complexity is not justified when the appropriate rows can be copied/pasted/updated in Excel in a few seconds. More flexibility and better normalization is certainly possible and may be supported in future: the focus here is on getting something that works now.</p>
<p>Originally, the plan was to do this as a series of standalone scripts, but this ended up with an unacceptable level of reimplementation of things that are already in QIIME. We have therefore included the code in QIIME.</p>
<p>In addition to QIIME&#8217;s standard dependencies, you will need several of QIIME&#8217;s optional dependencies to complete this tutorial:</p>
<blockquote>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">uclust</span></tt></li>
<li>454 off-instrument tools (<tt class="docutils literal"><span class="pre">sffinfo</span></tt>, <tt class="docutils literal"><span class="pre">sfffile</span></tt>). These must be obtained from Roche 454.</li>
</ul>
</blockquote>
<p>You should refer to the <a class="reference external" href="../install/index.html">QIIME install pages</a> for all information related to getting up and running with QIIME.</p>
</div>
<div class="section" id="overview-of-the-submission-process">
<h2>Overview of the Submission Process<a class="headerlink" href="#overview-of-the-submission-process" title="Permalink to this headline">¶</a></h2>
<p>An SRA submission can consist of metadata only, or of metadata together with sequence data. The SRA has recently added support for submission of barcoded pyrosequencing runs. This document describes how to prepare such submissions in a two-stage process:</p>
<blockquote>
<ol class="arabic simple">
<li>Submission of the study and sample metadata, including clinical metadata.</li>
<li>Submission of the experiment and run data, and associated technical metadata.</li>
</ol>
</blockquote>
<p>Updates are handled by regenerating the submission xml files (and optionally the .sff files), at which point they will be reloaded into the SRA. The current pipeline does not provide a mechanism for doing partial updates (e.g. adding an experiment or a run to an existing submission) &#8211; all the metadata and/or data for the submission must be replaced with clean files. It may be possible in the future to allow incremental submissions but this is not presently supported.</p>
</div>
<div class="section" id="preparing-a-submission-using-the-qiime-workflow-scripts">
<h2>Preparing a submission using the QIIME workflow scripts<a class="headerlink" href="#preparing-a-submission-using-the-qiime-workflow-scripts" title="Permalink to this headline">¶</a></h2>
<p>A submission consists of a <tt class="docutils literal"><span class="pre">submission.xml</span></tt>, metadata file, which references other xml metadata files and optionally tar files of sequence data files.</p>
<div class="section" id="set-up-your-data-for-the-tutorial-run">
<h3>Set up your data for the tutorial run<a class="headerlink" href="#set-up-your-data-for-the-tutorial-run" title="Permalink to this headline">¶</a></h3>
<p>Download and unpack the tutorial data, and change to the resulting directory:</p>
<div class="highlight-python"><pre>wget http://bmf.colorado.edu/QIIME/knight_handstudy_demo-v1.2.0-dev.zip
unzip knight_handstudy_demo-v1.2.0-dev.zip
cd knight_handstudy_demo-v1.2.0-dev</pre>
</div>
</div>
<div class="section" id="generate-the-experiment-and-run-metadata-submissions">
<h3>Generate the experiment and run metadata submissions<a class="headerlink" href="#generate-the-experiment-and-run-metadata-submissions" title="Permalink to this headline">¶</a></h3>
<blockquote>
<dl class="docutils">
<dt><strong>Input:</strong></dt>
<dd><ol class="first last loweralpha simple">
<li><tt class="docutils literal"><span class="pre">experiment.txt</span></tt> - tabular metadata about the contents of each combination of library and sff file.</li>
<li><tt class="docutils literal"><span class="pre">sff_files</span></tt> - a directory of multiple sff files containing the actual sequence data.</li>
<li><tt class="docutils literal"><span class="pre">submission.txt</span></tt> - a two-column file of tabular metadata about the submission.</li>
<li><tt class="docutils literal"><span class="pre">sra_parameters.txt</span></tt> - a qiime parameters file, defining what parameters should be passed to the individual component scripts</li>
<li><tt class="docutils literal"><span class="pre">greengenes_unaligned.fasta-OTUs_at_0.05.fasta</span></tt> - a FASTA file of known 16S sequences to screen for non-16S contaminants (here, we use a subset of the unaligned greengenes database, filtered at 95% sequence identity)  [<strong>optional: only required to perform human screen</strong>]</li>
<li><tt class="docutils literal"><span class="pre">study.txt</span></tt> - tabular metadata about the study (this is used to accession the study) [<strong>optional</strong>].</li>
<li><tt class="docutils literal"><span class="pre">sample.txt</span></tt> - tabular metadata about each sample (this is used to accession samples) [<strong>optional</strong>].</li>
</ol>
</dd>
<dt><strong>Output:</strong></dt>
<dd><ol class="first last loweralpha simple">
<li><tt class="docutils literal"><span class="pre">experiment.xml</span></tt> - xml-format metadata about the set of experiments described in <tt class="docutils literal"><span class="pre">library.txt</span></tt></li>
<li><tt class="docutils literal"><span class="pre">run.xml</span></tt> - xml-format metadata about each run, i.e. the association between a specific member of a pool and a specific xml file.</li>
<li><tt class="docutils literal"><span class="pre">data.tgz</span></tt> - a gzipped tar archive containing individual sff files for each SRA RUN (see the Questions above if you are unclear on the distinction between the SRA RUN concept and the concept of an instrument run).</li>
<li><tt class="docutils literal"><span class="pre">study.xml</span></tt> - xml-format metadata about the study.</li>
<li><tt class="docutils literal"><span class="pre">sample.xml</span></tt> - xml-format metadata about each sample</li>
<li><tt class="docutils literal"><span class="pre">submission.xml</span></tt> - xml-format metadata about the study and sample submission</li>
</ol>
</dd>
</dl>
</blockquote>
<p>The workflow script <a class="reference external" href="../scripts/process_sra_submission.html">process_sra_submission.py</a> may be used to create a submission of experiment and run metadata in one step.</p>
<p>Run the following command:</p>
<div class="highlight-python"><pre>process_sra_submission.py -s sff_files -e experiment.txt -r greengenes_unaligned.fasta-OTUs_at_0.05.fasta -u submission.txt -p sra_parameters.txt -o sra_out -a samples.txt -t study.txt</pre>
</div>
<p>This produces a tar archive of per-sample SFF files, <tt class="docutils literal"><span class="pre">experiment.xml</span></tt>, <tt class="docutils literal"><span class="pre">run.xml</span></tt>, and <tt class="docutils literal"><span class="pre">submission.xml</span></tt> from the input files. The list of commands that were actually run is available in the log file in the top-level <tt class="docutils literal"><span class="pre">sra_out/</span></tt> directory.</p>
<p>Users who wish to bypass the human screening step (which occupies at least 2/3 of the total runtime of <tt class="docutils literal"><span class="pre">process_sra_submission.py</span></tt>) can achieve this by not passing <tt class="docutils literal"><span class="pre">-r</span> <span class="pre">greengenes_unaligned.fasta-OTUs_at_0.05.fasta</span></tt>. This is useful, for example, when analyzing soil data which is unlikely to have human contaminants in high abundance.</p>
<p>The -a and -t parameters are optional. These (and the other xml files) can be generated directly using the <tt class="docutils literal"><span class="pre">make_sra_submission.py</span></tt> script.</p>
</div>
<div class="section" id="generate-the-metadata-xml-submission-files-without-processing-the-sequence-data">
<h3>Generate the metadata xml submission files without processing the sequence data<a class="headerlink" href="#generate-the-metadata-xml-submission-files-without-processing-the-sequence-data" title="Permalink to this headline">¶</a></h3>
<blockquote>
<dl class="docutils">
<dt>Input:</dt>
<dd><ol class="first last loweralpha simple">
<li><tt class="docutils literal"><span class="pre">study.txt</span></tt> - tabular metadata about the study (this is used to accession the study).</li>
<li><tt class="docutils literal"><span class="pre">sample.txt</span></tt> - tabular metadata about each sample (this is used to accession samples).</li>
<li><tt class="docutils literal"><span class="pre">submission.txt</span></tt> - a two-column file of tabular metadata about the submission.</li>
<li><tt class="docutils literal"><span class="pre">experiment.txt</span></tt> - tabular metadata about the contents of each combination of library and sff file.</li>
</ol>
</dd>
<dt>Output:</dt>
<dd><ol class="first last loweralpha simple">
<li><tt class="docutils literal"><span class="pre">study.xml</span></tt> - xml-format metadata about the study.</li>
<li><tt class="docutils literal"><span class="pre">sample.xml</span></tt> - xml-format metadata about each sample</li>
<li><tt class="docutils literal"><span class="pre">submission.xml</span></tt> - xml-format metadata about the study and sample submission</li>
<li><tt class="docutils literal"><span class="pre">experiment.xml</span></tt> - xml-format metadata about the contents of each combination of library and sff file.</li>
</ol>
</dd>
</dl>
</blockquote>
<p>Run the following command:</p>
<div class="highlight-python"><pre>make_sra_submission.py -a sample.txt -t study.txt -u submission.txt -e experiment.txt</pre>
</div>
<p>This produces <tt class="docutils literal"><span class="pre">sample.xml</span></tt>, <tt class="docutils literal"><span class="pre">study.xml</span></tt>, <tt class="docutils literal"><span class="pre">experiment.xml</span></tt>, and <tt class="docutils literal"><span class="pre">submission.xml</span></tt> from the corresponding tab-delimited text files. Any or all of the -a, -t, -u, -e options to generate the corresponding xml files.</p>
</div>
</div>
<div class="section" id="notes-regarding-individual-steps-of-the-sra-submission-process">
<h2>Notes regarding individual steps of the SRA submission process<a class="headerlink" href="#notes-regarding-individual-steps-of-the-sra-submission-process" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="../scripts/process_sra_submission.html">process_sra_submission.py</a> workflow script combines many separate QIIME commands. This section provides a discussion of the key components of the SRA submission workflow.</p>
<div class="section" id="print-the-commands-to-be-run-by-the-workflow-without-actually-running-them">
<h3>Print the commands to be run by the workflow without actually running them<a class="headerlink" href="#print-the-commands-to-be-run-by-the-workflow-without-actually-running-them" title="Permalink to this headline">¶</a></h3>
<p>It is sometimes useful to get the individual commands that will be run, but not actually run them. This is useful, for example, if you want to tweak one or more of the commands and then run them all via a bash script. To get the commands, but not run them, you can append <tt class="docutils literal"><span class="pre">-w</span></tt> to the <tt class="docutils literal"><span class="pre">process_sra_submission.py</span></tt> call:</p>
<div class="highlight-python"><pre>process_sra_submission.py -s sff_files -e experiment.txt -r greengenes_unaligned.fasta-OTUs_at_0.05.fasta -u submission.txt -p sra_parameters.txt -o sra_out -w</pre>
</div>
</div>
<div class="section" id="get-fasta-and-qual-from-sff-files">
<h3>Get fasta and qual from sff files<a class="headerlink" href="#get-fasta-and-qual-from-sff-files" title="Permalink to this headline">¶</a></h3>
<p>This step converts the sff files into text formats that are more usable.</p>
<p><strong>Output:</strong> Makes .fna and .qual files for each sff file.</p>
</div>
<div class="section" id="produce-valid-mapping-file-for-library-demultiplexing">
<h3>Produce valid mapping file for library demultiplexing<a class="headerlink" href="#produce-valid-mapping-file-for-library-demultiplexing" title="Permalink to this headline">¶</a></h3>
<p>This step converts the input experiment file into separate mapping files for each combination of STUDY and RUN_PREFIX (separating by run prefix is necessary when the same barcodes are used in different runs). This allows demultiplexing of the separate studies, which will then be sent in as separate submissions, and of the different barcoded plates, which will be demultiplexed separately.</p>
<p>Note: the LINKER field is no longer required in the spreadsheet.</p>
<p><strong>Output:</strong> Produces valid mapping files per 454 plate: <tt class="docutils literal"><span class="pre">fierer_hand_study_E86FECS.map</span></tt> and <tt class="docutils literal"><span class="pre">fierer_hand_study_FA6P1OK.map</span></tt></p>
</div>
<div class="section" id="demultiplex-libraries">
<h3>Demultiplex libraries<a class="headerlink" href="#demultiplex-libraries" title="Permalink to this headline">¶</a></h3>
<p>This step assigns each sequence to a library, dropping low-quality sequences and producing a log explaining why specific sequences were dropped.</p>
<p>NOTE: The SRA requests that you deposit ALL your sequence data, including bad reads, unless there is an IRB reason not to do so (i.e. human contamination). Therefore the quality and length filtering should be turned off. We do this by setting high values for the quality and length filtering, that in practice are not exceeded. For details on the quality and length filtering options run:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">split_libraries</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<p><strong>Output:</strong> Produces two files: <tt class="docutils literal"><span class="pre">seqs.fna</span></tt> with valid sequences assigned to samples via barcodes, and <tt class="docutils literal"><span class="pre">split_libraries_log.txt</span></tt> with info about which sequences failed QC. The parameters used are essentially turning off the default quality filters. You can turn off the quality filtering steps if you want to make sure that all the sequences appear in the output. You should do this by editing the appropriate values in your <tt class="docutils literal"><span class="pre">sra_parameters.txt</span></tt> file.</p>
<p>Briefly, the relevant settings in <tt class="docutils literal"><span class="pre">sra_parameters.txt</span></tt> require an average qual score of at least 5; a minimum sequence length of 30 (basically just the primer_barcode); a maximum sequence length of 1000; max homopolymer run of 1000; up to 100 errors in the primer; etc. In this run, we specify that we are using 12-base barcodes, (turning off the Golay error-correction which would be specified with <tt class="docutils literal"><span class="pre">split_libraries:barcode-type</span> <span class="pre">golay_12</span></tt>). These parameters are specified in <tt class="docutils literal"><span class="pre">sra_parameters.txt</span></tt> as:</p>
<div class="highlight-python"><pre>split_libraries:min-qual-score  5
split_libraries:min-seq-length  30
split_libraries:max-seq-length  1000
split_libraries:barcode-type    12
split_libraries:max-homopolymer 1000
split_libraries:max-primer-mismatch     100
split_libraries:max-ambig       1000</pre>
</div>
</div>
<div class="section" id="optional-human-screening-apply-uclust-ref-against-greengenes-to-eliminate-sequences-that-aren-t-really-16s-rrna">
<h3>Optional human screening: apply <tt class="docutils literal"><span class="pre">uclust_ref</span></tt> against greengenes to eliminate sequences that aren&#8217;t really 16S rRNA<a class="headerlink" href="#optional-human-screening-apply-uclust-ref-against-greengenes-to-eliminate-sequences-that-aren-t-really-16s-rrna" title="Permalink to this headline">¶</a></h3>
<p>This step performs a human/contaminant screen the &#8220;safe&#8221; way by identifying and excluding sequences that aren&#8217;t 16S rRNA. We are using <tt class="docutils literal"><span class="pre">uclust</span></tt> against a reference data set with a similarity threshold of 0.70. Our tests suggest that this is sufficient to screen out human genomic reads (the human 18S sequence hits bacterial 16S with a similarity of around 60%). These parameters are specified in <tt class="docutils literal"><span class="pre">sra_parameters.txt</span></tt> as:</p>
<div class="highlight-python"><pre># pick_otus parameters
pick_otus:similarity    0.70
pick_otus:enable_rev_strand_match       True</pre>
</div>
<p>The same procedure is applied to each library.</p>
<p>This step can be bypassed by not providing a reference data set to <tt class="docutils literal"><span class="pre">process_sra_submission.py</span></tt>.</p>
</div>
<div class="section" id="make-per-library-files-of-good-ids-to-pass-to-sfffile">
<h3>Make per-library files of &#8220;good&#8221; ids to pass to sfffile<a class="headerlink" href="#make-per-library-files-of-good-ids-to-pass-to-sfffile" title="Permalink to this headline">¶</a></h3>
<p>This step maps the ids of the representative set back onto the ids of the OTUs they came from so that we can get all the members of the OTUs that had a representative that matched a known 16S rRNA.</p>
<p><strong>Output:</strong> This makes a new directory called <tt class="docutils literal"><span class="pre">E86FECS_demultiplex/per_lib_idlists</span></tt>, which contains a separate file with an id list for each library.</p>
<p>The same procedure is applied to each library.</p>
</div>
<div class="section" id="use-sfffile-to-make-per-library-sff-files">
<h3>Use sfffile to make per-library sff files<a class="headerlink" href="#use-sfffile-to-make-per-library-sff-files" title="Permalink to this headline">¶</a></h3>
<p>This step takes the good lists of ids from step 7 and extracts a separate sff file for each of those lists.</p>
</div>
<div class="section" id="use-sfffile-to-quality-trim-the-barcodes-primers-and-linkers">
<h3>Use sfffile to quality-trim the barcodes, primers and linkers<a class="headerlink" href="#use-sfffile-to-quality-trim-the-barcodes-primers-and-linkers" title="Permalink to this headline">¶</a></h3>
<p>The SRA requires that the user reset the left trim in the sff file to eliminate the technical reads (barcode, primer, linker if present). This means figuring out the length of the technical parts of the read, the length of the current read, writing out a text file with the per-id info, and running sfffile to reset the read lengths.</p>
</div>
<div class="section" id="optional-post-processing-modifying-the-second-stage-submission">
<h3>Optional post-processing: modifying the second-stage submission<a class="headerlink" href="#optional-post-processing-modifying-the-second-stage-submission" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference external" href="../scripts/make_sra_submission.html">make_sra_submission.py</a> script has the ability to include per-experiment attributes or links.  The attributes and links should be specified in separate, tab-delimited files. For example, a file named <tt class="docutils literal"><span class="pre">attributes.txt</span></tt> can be created with the following contents:</p>
<div class="highlight-python"><pre>#EXPERIMENT_ALIAS     Attribute       Value
fierer_hand_study_FA6P1OK     library strategy        targeted-locus
fierer_hand_study_FA6P1OK     gene    16S rRNA V1-V2 region
fierer_hand_study_E86FECS     library strategy        targeted-locus
fierer_hand_study_E86FECS     gene    16S rRNA V1-V2 region</pre>
</div>
<p>The following command will then add &#8220;gene&#8221; and &#8220;library strategy&#8221; attributes to both experiments in the resulting XML. (The experiment alias is specified in <tt class="docutils literal"><span class="pre">experiment.txt</span></tt>, under the field &#8216;EXPERIMENT_ALIAS&#8217;.)</p>
<div class="highlight-python"><pre>make_sra_submission.py -u submission.txt -e experiment.txt -s per_run_sff --experiment_attribute_fp=attributes.txt</pre>
</div>
<p>Links may be added to the experiments in a similar manner. After the <a class="reference external" href="../scripts/make_sra_submission.html">make_sra_submission.py</a> script has been run, the resulting XML files are ready to submit to the SRA.</p>
<p>Note: SRA prefers you give the individual files more meaningful names than the defaults, so suggest not just using generic names like experiment etc.</p>
</div>
</div>
<div class="section" id="questions-about-the-submission-process">
<h2>Questions about the Submission Process<a class="headerlink" href="#questions-about-the-submission-process" title="Permalink to this headline">¶</a></h2>
<blockquote>
<p>Q1. Can I submit multiplexed pyrosequnecing runs now?</p>
<p>A1. Yes.</p>
<p>Q2. Can I combine mock and clinical samples on the same 454 plate? (or, more generally, can I combine samples from different studies on the same 454 plate?)</p>
<p>A2. Yes, but you must specify in the <tt class="docutils literal"><span class="pre">library.txt</span></tt> input file which samples go with which study.</p>
<p>Q3. Can I associate the same sample (and thus reads) with more than one study?</p>
<p>A3. No.</p>
<p>Q4. Can I combine samples that use different primers on the same run?</p>
<p>A4. Yes, but you must specify in the <tt class="docutils literal"><span class="pre">library.txt</span></tt> input file which primer was used for each &#8220;member&#8221; of the pooled library.</p>
<p>Q5. Who will submit what?</p>
<p>A5. At this stage, we expect the DACC to submit both the sample/study metadata and the experiment/library metadata and sequence data for the pilot. Later, the Centers will have the capacity to submit their own data. Centers will be credited with their data appropriately regardless of the mechanism by which the submission is actually performed. The submission will be a two-stage process: (1) the creation of study and sample records by the DACC, (2) the submission of sequence data and associated metadata by the DACC and/ or the Centers.</p>
<p>Q6. Can I associate the same sample with more than one barcode and/or primer?</p>
<p>A6. Yes, but you must specify a unique identifier for each &#8220;member&#8221; of the pool that associates the sample, primer and barcode.</p>
<p>Q7. What is the distinction between a STUDY, an EXPERIMENT, and a RUN?</p>
<p>A7. As SRA uses the terms, a STUDY is a collection of EXPERIMENTS. An EXPERIMENT is a LIBRARY (potentially a library of many samples that form a POOL, if multiplexing was used &#8211; each MEMBER of a pool is associated with a sample, a primer, and a barcode) that was sequenced using one or more instrument runs. A RUN is the sequencing of a particular MEMBER of a pooled library on a particular instrument at a particular time. Thus, a single instrument run gives rise to many RUN entries in SRA.</p>
<p>Q8. Is there an intermediate level between STUDY and EXPERIMENT?</p>
<p>A8. Not for practical purposes. SRA will eventually allow a hierarchy of STUDY entries but this is not yet implemented.</p>
<p>Q9. Do I really have to make a separate sff file for every MEMBER of every POOL for every instrument run?</p>
<p>A9. Yes, and you also have to reset the quality trimming to correspond to the primer that was used for that particular member. The SRA will, in future, provide the demultiplexing service, but for now requires that the submissions be demultiplexed in advance. Fortunately, the accompanying scripts assist with this process.</p>
<p>Q10. Is it OK for primers to be different lengths on the same 454 run?</p>
<p>A10. Yes, but not within the same MEMBER of a library (i.e. if you have primers of different lengths, the different lengths are considered different MEMBER entries and should be marked as such in <tt class="docutils literal"><span class="pre">library.txt</span></tt>).</p>
<p>Q11. How should degenerate primers be handled?</p>
<p>A11. All possible sequences that match the degenerate primer should be allowed using the EXPECTED_BASECALL_TABLE mechanism in <tt class="docutils literal"><span class="pre">experiment.xml</span></tt> (see example).</p>
</blockquote>
</div>
<div class="section" id="standard-sra-parameters-txt-file-for-barcoded-16s-community-sequencing-on-454">
<h2>Standard sra_parameters.txt file for barcoded 16S community sequencing on 454<a class="headerlink" href="#standard-sra-parameters-txt-file-for-barcoded-16s-community-sequencing-on-454" title="Permalink to this headline">¶</a></h2>
<p>Currently our standard parameters files looks like the following. You can copy and paste this to a text file, and pass it with <tt class="docutils literal"><span class="pre">-p</span></tt> to <tt class="docutils literal"><span class="pre">process_sra_submission.py</span></tt>. The <tt class="docutils literal"><span class="pre">pick_otus:similarity</span></tt> value has been carefully chosen to exclude human sequences but include bacterial/archaeal 16S sequences, so it&#8217;s not a good idea to change that without exploring the affect it will have.</p>
<div class="highlight-python"><pre># split_libraries parameters
split_libraries:min-qual-score  5
split_libraries:min-seq-length  30
split_libraries:max-seq-length  1000
split_libraries:barcode-type    12
split_libraries:max-homopolymer 1000
split_libraries:max-primer-mismatch     100
split_libraries:max-ambig       1000

# pick_otus parameters
pick_otus:similarity    0.70
pick_otus:enable_rev_strand_match       True</pre>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
            <h3><a href="../index.html">Table Of Contents</a></h3>
            <ul>
<li><a class="reference external" href="">SRA Submission</a><ul>
<li><a class="reference external" href="#introduction">Introduction</a></li>
<li><a class="reference external" href="#philosophy-of-qiime-s-sra-submission-code">Philosophy of QIIME&#8217;s SRA submission code</a></li>
<li><a class="reference external" href="#overview-of-the-submission-process">Overview of the Submission Process</a></li>
<li><a class="reference external" href="#preparing-a-submission-using-the-qiime-workflow-scripts">Preparing a submission using the QIIME workflow scripts</a><ul>
<li><a class="reference external" href="#set-up-your-data-for-the-tutorial-run">Set up your data for the tutorial run</a></li>
<li><a class="reference external" href="#generate-the-experiment-and-run-metadata-submissions">Generate the experiment and run metadata submissions</a></li>
<li><a class="reference external" href="#generate-the-metadata-xml-submission-files-without-processing-the-sequence-data">Generate the metadata xml submission files without processing the sequence data</a></li>
</ul>
</li>
<li><a class="reference external" href="#notes-regarding-individual-steps-of-the-sra-submission-process">Notes regarding individual steps of the SRA submission process</a><ul>
<li><a class="reference external" href="#print-the-commands-to-be-run-by-the-workflow-without-actually-running-them">Print the commands to be run by the workflow without actually running them</a></li>
<li><a class="reference external" href="#get-fasta-and-qual-from-sff-files">Get fasta and qual from sff files</a></li>
<li><a class="reference external" href="#produce-valid-mapping-file-for-library-demultiplexing">Produce valid mapping file for library demultiplexing</a></li>
<li><a class="reference external" href="#demultiplex-libraries">Demultiplex libraries</a></li>
<li><a class="reference external" href="#optional-human-screening-apply-uclust-ref-against-greengenes-to-eliminate-sequences-that-aren-t-really-16s-rrna">Optional human screening: apply <tt class="docutils literal"><span class="pre">uclust_ref</span></tt> against greengenes to eliminate sequences that aren&#8217;t really 16S rRNA</a></li>
<li><a class="reference external" href="#make-per-library-files-of-good-ids-to-pass-to-sfffile">Make per-library files of &#8220;good&#8221; ids to pass to sfffile</a></li>
<li><a class="reference external" href="#use-sfffile-to-make-per-library-sff-files">Use sfffile to make per-library sff files</a></li>
<li><a class="reference external" href="#use-sfffile-to-quality-trim-the-barcodes-primers-and-linkers">Use sfffile to quality-trim the barcodes, primers and linkers</a></li>
<li><a class="reference external" href="#optional-post-processing-modifying-the-second-stage-submission">Optional post-processing: modifying the second-stage submission</a></li>
</ul>
</li>
<li><a class="reference external" href="#questions-about-the-submission-process">Questions about the Submission Process</a></li>
<li><a class="reference external" href="#standard-sra-parameters-txt-file-for-barcoded-16s-community-sequencing-on-454">Standard sra_parameters.txt file for barcoded 16S community sequencing on 454</a></li>
</ul>
</li>
</ul>


	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="../install/index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


          <div id="searchbox" style="display: none">
            <h3>Quick search</h3>
              <form class="search" action="../search.html" method="get">
                <input type="text" name="q" size="18" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
              </form>
              <p class="searchtip" style="font-size: 90%">
              Enter search terms or a module, class or function name.
              </p>
          </div>
          <script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://pycogent.sourceforge.net" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
      &copy; Copyright 2010, QIIME Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 0.6.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>