

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>QIIME scripts &mdash; Homepage</title>
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.3.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="qiime-scripts">
<span id="index"></span><h1>QIIME scripts<a class="headerlink" href="#qiime-scripts" title="Permalink to this headline">Â¶</a></h1>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="add_taxa.html"><em>add_taxa.py</em> &#8211; Add taxa to OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="adjust_seq_orientation.html"><em>adjust_seq_orientation.py</em> &#8211; Get the reverse complement of all sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="align_seqs.html"><em>align_seqs.py</em> &#8211; Align sequences using a variety of alignment methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="alpha_diversity.html"><em>alpha_diversity.py</em> &#8211; Calculate alpha diversity on each sample in an otu table, using a variety of alpha diversity metrics</a></li>
<li class="toctree-l1"><a class="reference internal" href="alpha_diversity_metrics.html"><em>alpha_diversity_metrics</em> &#8211; List of available metrics</a></li>
<li class="toctree-l1"><a class="reference internal" href="alpha_rarefaction.html"><em>alpha_rarefaction.py</em> &#8211; A workflow script for performing alpha rarefaction</a></li>
<li class="toctree-l1"><a class="reference internal" href="ampliconnoise.html"><em>ampliconnoise.py</em> &#8211; Run AmpliconNoise</a></li>
<li class="toctree-l1"><a class="reference internal" href="assign_taxonomy.html"><em>assign_taxonomy.py</em> &#8211; Assign taxonomy to each sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_diversity.html"><em>beta_diversity.py</em> &#8211; Calculate beta diversity (pairwise sample dissimilarity) on one or many otu tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_diversity_metrics.html"><em>beta_diversity_metrics</em> &#8211; List of available metrics</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_diversity_through_3d_plots.html"><em>beta_diversity_through_3d_plots.py</em> &#8211;</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_diversity_through_plots.html"><em>beta_diversity_through_plots.py</em> &#8211; A workflow script for computing beta diversity distance matrices and the corresponding 3D plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="beta_significance.html"><em>beta_significance.py</em> &#8211; This script runs any of a set of common tests to determine if a sample is statistically significantly different from another sample</a></li>
<li class="toctree-l1"><a class="reference internal" href="blast_wrapper.html"><em>blast_wrapper.py</em> &#8211; Blast Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="categorized_dist_scatterplot.html"><em>categorized_dist_scatterplot.py</em> &#8211; makes a figure representing average distances between samples, broken down by categories. I call it a &#8216;categorized distance scatterplot&#8217;</a></li>
<li class="toctree-l1"><a class="reference internal" href="check_id_map.html"><em>check_id_map.py</em> &#8211; Checks user&#8217;s metadata mapping file for required data, valid format</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster_quality.html"><em>cluster_quality.py</em> &#8211; compute the quality of a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="collate_alpha.html"><em>collate_alpha.py</em> &#8211; Collate alpha diversity results</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_3d_plots.html"><em>compare_3d_plots.py</em> &#8211; Plot several PCoA files on the same 3D plot</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_alpha_diversity.html"><em>compare_alpha_diversity.py</em> &#8211; This script compares alpha diversities based on a t_two_sample test</a></li>
<li class="toctree-l1"><a class="reference internal" href="compare_distance_matrices.html"><em>compare_distance_matrices.py</em> &#8211; Script for computing Mantel correlations between as set of distance matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="consensus_tree.html"><em>consensus_tree.py</em> &#8211; This script outputs a majority consensus tree given a collection of input trees.</a></li>
<li class="toctree-l1"><a class="reference internal" href="convert_otu_table_to_unifrac_sample_mapping.html"><em>convert_otu_table_to_unifrac_sample_mapping.py</em> &#8211; Convert a QIIME OTU table to a UniFrac sample mapping file</a></li>
<li class="toctree-l1"><a class="reference internal" href="convert_unifrac_sample_mapping_to_otu_table.html"><em>convert_unifrac_sample_mapping_to_otu_table.py</em> &#8211; Convert a UniFrac sample mapping file to an OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="core_qiime_analyses.html"><em>core_qiime_analyses.py</em> &#8211; A workflow for running a core set of QIIME analyses.</a></li>
<li class="toctree-l1"><a class="reference internal" href="cytoscape_usage.html"><em>cytoscape_usage</em> &#8211; Loading Results with Cytoscape</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoise.html"><em>denoise.py</em> &#8211;</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoise_wrapper.html"><em>denoise_wrapper.py</em> &#8211; Denoise a flowgram file</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoiser.html"><em>denoiser.py</em> &#8211; Remove noise from  454 sequencing data</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoiser_postprocess.html"><em>denoiser_postprocess.py</em> &#8211; Merge denoiser output with OTU picker output</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoiser_preprocess.html"><em>denoiser_preprocess.py</em> &#8211; Run phase of denoiser algorithm: prefix clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="denoiser_worker.html"><em>denoiser_worker.py</em> &#8211; Start a denoiser worker process</a></li>
<li class="toctree-l1"><a class="reference internal" href="dissimilarity_mtx_stats.html"><em>dissimilarity_mtx_stats.py</em> &#8211; Calculate mean, median and standard deviation from a set of distance matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="exclude_seqs_by_blast.html"><em>exclude_seqs_by_blast.py</em> &#8211; Exclude contaminated sequences using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="extract_seqs_by_sample_id.html"><em>extract_seqs_by_sample_id.py</em> &#8211; Extract sequences based on the SampleID</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_alignment.html"><em>filter_alignment.py</em> &#8211; Filter sequence alignment by removing highly variable regions</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_by_metadata.html"><em>filter_by_metadata.py</em> &#8211; Filter OTU table by removal of specified metadata</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_fasta.html"><em>filter_fasta.py</em> &#8211; This script can be applied to remove sequences from a fasta file based on input criteria.</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_otu_table.html"><em>filter_otu_table.py</em> &#8211; Filters OTU table by minimum OTU count and number of samples or by taxonomy</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_otus_by_sample.html"><em>filter_otus_by_sample.py</em> &#8211; Filter OTU mapping file and sequences by SampleIDs</a></li>
<li class="toctree-l1"><a class="reference internal" href="filter_tree.html"><em>filter_tree.py</em> &#8211; This script prunes a tree based on a set of tip names</a></li>
<li class="toctree-l1"><a class="reference internal" href="fix_arb_fasta.html"><em>fix_arb_fasta.py</em> &#8211; Reformat ARB FASTA files</a></li>
<li class="toctree-l1"><a class="reference internal" href="identify_chimeric_seqs.html"><em>identify_chimeric_seqs.py</em> &#8211; Identify chimeric sequences in input FASTA file</a></li>
<li class="toctree-l1"><a class="reference internal" href="identify_missing_files.html"><em>identify_missing_files.py</em> &#8211; This script checks for the existence expected file in parallel runs.</a></li>
<li class="toctree-l1"><a class="reference internal" href="inflate_denoiser_output.html"><em>inflate_denoiser_output.py</em> &#8211; Inflate denoiser results so they can be passed directly to OTU pickers.</a></li>
<li class="toctree-l1"><a class="reference internal" href="jackknifed_beta_diversity.html"><em>jackknifed_beta_diversity.py</em> &#8211; A workflow script for performing jackknifed UPGMA clustering and build jackknifed 2d and 3D PCoA plots.</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_2d_plots.html"><em>make_2d_plots.py</em> &#8211; Make 2D PCoA Plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_3d_plots.html"><em>make_3d_plots.py</em> &#8211; Make 3D PCoA plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_bootstrapped_tree.html"><em>make_bootstrapped_tree.py</em> &#8211; Make bootstrapped tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_distance_histograms.html"><em>make_distance_histograms.py</em> &#8211; Make distance histograms</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_fastq.html"><em>make_fastq.py</em> &#8211; Make fastq file for ERA submission from paired fasta and qual files</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_library_id_lists.html"><em>make_library_id_lists.py</em> &#8211; Make library id lists</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_otu_heatmap.html"><em>make_otu_heatmap.py</em> &#8211; Make heatmap of OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_otu_heatmap_html.html"><em>make_otu_heatmap_html.py</em> &#8211; Make heatmap of OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_otu_network.html"><em>make_otu_network.py</em> &#8211; Make an OTU network and calculate statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_otu_table.html"><em>make_otu_table.py</em> &#8211; Make OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_per_library_sff.html"><em>make_per_library_sff.py</em> &#8211; Make per-library sff files from ID lists</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_phylogeny.html"><em>make_phylogeny.py</em> &#8211; Make Phylogeny</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_pie_charts.html"><em>make_pie_charts.py</em> &#8211;</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_prefs_file.html"><em>make_prefs_file.py</em> &#8211; Generate preferences file</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_qiime_py_file.html"><em>make_qiime_py_file.py</em> &#8211; Create python file</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_qiime_rst_file.html"><em>make_qiime_rst_file.py</em> &#8211; Make Sphinx RST file</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_rarefaction_plots.html"><em>make_rarefaction_plots.py</em> &#8211; Generate Rarefaction Plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="make_tep.html"><em>make_tep.py</em> &#8211; Makes TopiaryExplorer project file</a></li>
<li class="toctree-l1"><a class="reference internal" href="merge_denoiser_output.html"><em>merge_denoiser_output.py</em> &#8211; Merge the output of denoising step back into QIIME</a></li>
<li class="toctree-l1"><a class="reference internal" href="merge_mapping_files.html"><em>merge_mapping_files.py</em> &#8211; Merge mapping files</a></li>
<li class="toctree-l1"><a class="reference internal" href="merge_otu_maps.html"><em>merge_otu_maps.py</em> &#8211; Merge OTU mapping files</a></li>
<li class="toctree-l1"><a class="reference internal" href="merge_otu_tables.html"><em>merge_otu_tables.py</em> &#8211; Merge two or more OTU tables into a single OTU table.</a></li>
<li class="toctree-l1"><a class="reference internal" href="multiple_rarefactions.html"><em>multiple_rarefactions.py</em> &#8211; Perform multiple subsamplings/rarefactions on an otu table</a></li>
<li class="toctree-l1"><a class="reference internal" href="multiple_rarefactions_even_depth.html"><em>multiple_rarefactions_even_depth.py</em> &#8211; Perform multiple rarefactions on a single otu table, at one depth of sequences/sample</a></li>
<li class="toctree-l1"><a class="reference internal" href="neighbor_joining.html"><em>neighbor_joining.py</em> &#8211; Build a neighbor joining tree comparing samples</a></li>
<li class="toctree-l1"><a class="reference internal" href="nmds.html"><em>nmds.py</em> &#8211; Nonmetric Multidimensional Scaling (NMDS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="otu_category_significance.html"><em>otu_category_significance.py</em> &#8211; OTU significance and co-occurence analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_align_seqs_pynast.html"><em>parallel_align_seqs_pynast.py</em> &#8211; Parallel sequence alignment using PyNAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_alpha_diversity.html"><em>parallel_alpha_diversity.py</em> &#8211; Parallel alpha diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_assign_taxonomy_blast.html"><em>parallel_assign_taxonomy_blast.py</em> &#8211; Parallel taxonomy assignment using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_assign_taxonomy_rdp.html"><em>parallel_assign_taxonomy_rdp.py</em> &#8211; Parallel taxonomy assignment using RDP</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_beta_diversity.html"><em>parallel_beta_diversity.py</em> &#8211; Parallel beta diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_blast.html"><em>parallel_blast.py</em> &#8211; Parallel BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_identify_chimeric_seqs.html"><em>parallel_identify_chimeric_seqs.py</em> &#8211; Parallel chimera detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_multiple_rarefactions.html"><em>parallel_multiple_rarefactions.py</em> &#8211; Parallel multiple file rarefaction</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_pick_otus_blast.html"><em>parallel_pick_otus_blast.py</em> &#8211; Parallel pick otus using BLAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="parallel_pick_otus_uclust_ref.html"><em>parallel_pick_otus_uclust_ref.py</em> &#8211; Parallel pick otus using uclust_ref</a></li>
<li class="toctree-l1"><a class="reference internal" href="per_library_stats.html"><em>per_library_stats.py</em> &#8211; Calculate per library statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="pick_otus.html"><em>pick_otus.py</em> &#8211; OTU picking</a></li>
<li class="toctree-l1"><a class="reference internal" href="pick_otus_through_otu_table.html"><em>pick_otus_through_otu_table.py</em> &#8211; A workflow script for picking OTUs through building OTU tables</a></li>
<li class="toctree-l1"><a class="reference internal" href="pick_reference_otus_through_otu_table.html"><em>pick_reference_otus_through_otu_table.py</em> &#8211; Reference OTU picking/Shotgun UniFrac workflow.</a></li>
<li class="toctree-l1"><a class="reference internal" href="pick_rep_set.html"><em>pick_rep_set.py</em> &#8211; Pick representative set of sequences</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_rank_abundance_graph.html"><em>plot_rank_abundance_graph.py</em> &#8211; plot rank-abundance curve</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_semivariogram.html"><em>plot_semivariogram.py</em> &#8211; Fits a model between two distance matrices and plots the result</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_taxa_summary.html"><em>plot_taxa_summary.py</em> &#8211; Make taxaonomy summary charts based on taxonomy assignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="poller.html"><em>poller.py</em> &#8211; Poller for parallel QIIME scripts.</a></li>
<li class="toctree-l1"><a class="reference internal" href="poller_example.html"><em>poller_example.py</em> &#8211; Create python file</a></li>
<li class="toctree-l1"><a class="reference internal" href="pool_by_metadata.html"><em>pool_by_metadata.py</em> &#8211; pool samples in OTU table and mapping file based on sample metadata from mapping file</a></li>
<li class="toctree-l1"><a class="reference internal" href="preferences_file.html"><em>preferences_file</em> &#8211; Example of a prefs file</a></li>
<li class="toctree-l1"><a class="reference internal" href="principal_coordinates.html"><em>principal_coordinates.py</em> &#8211; Principal Coordinates Analysis (PCoA)</a></li>
<li class="toctree-l1"><a class="reference internal" href="print_qiime_config.html"><em>print_qiime_config.py</em> &#8211; Print out the qiime config settings.</a></li>
<li class="toctree-l1"><a class="reference internal" href="process_iseq.html"><em>process_iseq.py</em> &#8211; Given a directory of per-swath qseq files, this script generates a single fastq per lane.</a></li>
<li class="toctree-l1"><a class="reference internal" href="process_qseq.html"><em>process_qseq.py</em> &#8211; Given a directory of per-swath qseq files, this script generates a single fastq per lane.</a></li>
<li class="toctree-l1"><a class="reference internal" href="process_sff.html"><em>process_sff.py</em> &#8211; Convert sff to FASTA and QUAL files</a></li>
<li class="toctree-l1"><a class="reference internal" href="quality_scores_plot.html"><em>quality_scores_plot.py</em> &#8211; Generates histograms of sequence quality scores and number of nucleotides recorded at a particular index</a></li>
<li class="toctree-l1"><a class="reference internal" href="shared_phylotypes.html"><em>shared_phylotypes.py</em> &#8211; Compute shared OTUs between all pairs of samples</a></li>
<li class="toctree-l1"><a class="reference internal" href="simsam.html"><em>simsam.py</em> &#8211; Simulate samples for each sample in an OTU table, using a phylogenetic tree.</a></li>
<li class="toctree-l1"><a class="reference internal" href="single_rarefaction.html"><em>single_rarefaction.py</em> &#8211; Perform rarefaction on an otu table</a></li>
<li class="toctree-l1"><a class="reference internal" href="sort_denoiser_output.html"><em>sort_denoiser_output.py</em> &#8211; Sort denoiser output by cluster size.</a></li>
<li class="toctree-l1"><a class="reference internal" href="sort_otu_table.html"><em>sort_otu_table.py</em> &#8211; Script for sorting the sample IDs in an OTU table based on a specified value in a mapping file.</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_libraries.html"><em>split_libraries.py</em> &#8211; Split libraries according to barcodes specified in mapping file</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_libraries_fastq.html"><em>split_libraries_fastq.py</em> &#8211; This script performs demultiplexing of Fastq sequence data where barcodes and sequences are contained in two separate fastq files (common on Illumina runs).</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_libraries_illumina.html"><em>split_libraries_illumina.py</em> &#8211; Script for processing raw Illumina Genome Analyzer II data.</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_otu_table.html"><em>split_otu_table.py</em> &#8211; Split in a single OTU table into one OTU table per value in a specified field of the mapping file.</a></li>
<li class="toctree-l1"><a class="reference internal" href="split_otu_table_by_taxonomy.html"><em>split_otu_table_by_taxonomy.py</em> &#8211; Script to split a single OTU table into multiple tables based on the taxonomy at some user-specified depth.</a></li>
<li class="toctree-l1"><a class="reference internal" href="start_parallel_jobs.html"><em>start_parallel_jobs.py</em> &#8211; Starts multiple jobs in parallel on multicore or multiprocessor systems.</a></li>
<li class="toctree-l1"><a class="reference internal" href="start_parallel_jobs_torque.html"><em>start_parallel_jobs_torque.py</em> &#8211; Starts multiple jobs in parallel on torque/qsub based multiprocessor systems.</a></li>
<li class="toctree-l1"><a class="reference internal" href="submit_to_mgrast.html"><em>submit_to_mgrast.py</em> &#8211; This script submits a FASTA file to MG-RAST</a></li>
<li class="toctree-l1"><a class="reference internal" href="summarize_otu_by_cat.html"><em>summarize_otu_by_cat.py</em> &#8211; Summarize an OTU table by a single column in the mapping file.</a></li>
<li class="toctree-l1"><a class="reference internal" href="summarize_taxa.html"><em>summarize_taxa.py</em> &#8211; Summarize Taxa</a></li>
<li class="toctree-l1"><a class="reference internal" href="summarize_taxa_through_plots.html"><em>summarize_taxa_through_plots.py</em> &#8211; A workflow script for performing taxonomy summaries and plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="supervised_learning.html"><em>supervised_learning.py</em> &#8211; Run supervised classification using OTUs as predictors and a mapping file category as class labels.</a></li>
<li class="toctree-l1"><a class="reference internal" href="transform_coordinate_matrices.html"><em>transform_coordinate_matrices.py</em> &#8211; Transform 2 coordinate matrices</a></li>
<li class="toctree-l1"><a class="reference internal" href="tree_compare.html"><em>tree_compare.py</em> &#8211; Compare jackknifed/bootstrapped trees</a></li>
<li class="toctree-l1"><a class="reference internal" href="trflp_file_to_otu_table.html"><em>trflp_file_to_otu_table.py</em> &#8211; Convert TRFLP text file to an OTU table</a></li>
<li class="toctree-l1"><a class="reference internal" href="trim_sff_primers.html"><em>trim_sff_primers.py</em> &#8211; Trim sff primers</a></li>
<li class="toctree-l1"><a class="reference internal" href="truncate_fasta_qual_files.html"><em>truncate_fasta_qual_files.py</em> &#8211; Generates filtered fasta and quality score files by truncating at the specified base position.</a></li>
<li class="toctree-l1"><a class="reference internal" href="unweight_fasta.html"><em>unweight_fasta.py</em> &#8211; Transform fasta files with abundance weighting into unweighted</a></li>
<li class="toctree-l1"><a class="reference internal" href="upgma_cluster.html"><em>upgma_cluster.py</em> &#8211; Build a UPGMA tree comparing samples</a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    

	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="../install/index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="../tutorials/index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://pycogent.sourceforge.net" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2011, QIIME Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.7.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>