<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>join_paired_ends.py – Joins paired-end Illumina reads. &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />
    <link rel="up" title="QIIME Scripts" href="index.html" />
    <link rel="next" title="load_remote_mapping_file.py – Downloads and saves a remote mapping file" href="load_remote_mapping_file.html" />
    <link rel="prev" title="jackknifed_beta_diversity.py – A workflow script for performing jackknifed UPGMA clustering and building jackknifed Emperor PCoA plots." href="jackknifed_beta_diversity.html" />
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="load_remote_mapping_file.html" title="load_remote_mapping_file.py – Downloads and saves a remote mapping file"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="jackknifed_beta_diversity.html" title="jackknifed_beta_diversity.py – A workflow script for performing jackknifed UPGMA clustering and building jackknifed Emperor PCoA plots."
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="../documentation/index.html" >QIIME File Formats</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">QIIME Scripts</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="join-paired-ends"></span><div class="section" id="join-paired-ends-py-joins-paired-end-illumina-reads">
<span id="index-0"></span><h1><em>join_paired_ends.py</em> &#8211; Joins paired-end Illumina reads.<a class="headerlink" href="#join-paired-ends-py-joins-paired-end-illumina-reads" title="Permalink to this headline">¶</a></h1>
<p><strong>Description:</strong></p>
<p>This script takes forward and reverse Illumina reads and joins them using the method chosen. Will optionally create an updated index reads file containing index reads for the surviving joined paired end reads. If the option to write an updated index file is chosen, be sure that the order and header format of the index reads is the same as the order and header format of reads in the files that will be joined (this is the default for reads generated on the Illumina instruments).</p>
<p>Currently, there are two methods that can be selected by the user to join paired-end data:</p>
<ol class="arabic simple">
<li>fastq-join - Erik Aronesty, 2011. ea-utils : &#8220;Command-line tools for processing biological sequencing data&#8221; (<a class="reference external" href="http://code.google.com/p/ea-utils">http://code.google.com/p/ea-utils</a>)</li>
<li>SeqPrep - (<a class="reference external" href="https://github.com/jstjohn/SeqPrep">https://github.com/jstjohn/SeqPrep</a>)</li>
</ol>
<p><strong>Usage:</strong> <tt class="file docutils literal"><span class="pre">join_paired_ends.py</span> <span class="pre">[options]</span></tt></p>
<p><strong>Input Arguments:</strong></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><strong>[REQUIRED]</strong></p>
<dl class="docutils">
<dt>-f, <cite>-</cite>-forward_reads_fp</dt>
<dd>Path to input forward reads in FASTQ format.</dd>
<dt>-r, <cite>-</cite>-reverse_reads_fp</dt>
<dd>Path to input reverse reads in FASTQ format.</dd>
<dt>-o, <cite>-</cite>-output_dir</dt>
<dd>Directory to store result files</dd>
</dl>
<p><strong>[OPTIONAL]</strong></p>
<dl class="last docutils">
<dt>-m, <cite>-</cite>-pe_join_method</dt>
<dd>Method to use for joining paired-ends. Valid choices are: fastq-join, SeqPrep [default: fastq-join]</dd>
<dt>-b, <cite>-</cite>-index_reads_fp</dt>
<dd>Path to the barcode / index reads in FASTQ format. Will be filtered based on surviving joined pairs.</dd>
<dt>-j, <cite>-</cite>-min_overlap</dt>
<dd>Applies to both fastq-join and SeqPrep methods. Minimum allowed overlap in base-pairs required to join pairs. If not set, progam defaults will be used. Must be an integer. [default: None]</dd>
<dt>-p, <cite>-</cite>-perc_max_diff</dt>
<dd>Only applies to fastq-join method, otherwise ignored. Maximum allowed % differences within region of overlap. If not set, progam defaults will be used. Must be an integer between 1-100 [default: None]</dd>
<dt>-y, <cite>-</cite>-max_ascii_score</dt>
<dd>Only applies to SeqPrep method, otherwise ignored. Maximum quality score / ascii code allowed to appear within joined pairs output. For more information, please see: <a class="reference external" href="http://en.wikipedia.org/wiki/FASTQ_format">http://en.wikipedia.org/wiki/FASTQ_format</a>. [default: J]</dd>
<dt>-n, <cite>-</cite>-min_frac_match</dt>
<dd>Only applies to SeqPrep method, otherwise ignored. Minimum allowed fraction of matching bases required to join reads. Must be a float between 0-1. If not set, progam defaults will be used. [default: None]</dd>
<dt>-g, <cite>-</cite>-max_good_mismatch</dt>
<dd>Only applies to SeqPrep method, otherwise ignored. Maximum mis-matched high quality bases allowed to join reads. Must be a float between 0-1. If not set, progam defaults will be used. [default: None]</dd>
<dt>-6, <cite>-</cite>-phred_64</dt>
<dd>Only applies to SeqPrep method, otherwise ignored. Set if input reads are in phred+64 format. Output will always be phred+33. [default: False]</dd>
</dl>
</div>
<p><strong>Output:</strong></p>
<p>All paired-end joining software will return a joined / merged / assembled paired-end fastq file. Depending on the method chosen, additional files may be written to the user-specified output directory.</p>
<ol class="arabic simple">
<li>fastq-join will output fastq-formatted files as:<ul>
<li>&#8220;*.join&#8221;: assembled / joined reads output</li>
<li>&#8220;*.un1&#8221;: unassembled / unjoined reads1 output</li>
<li>&#8220;*.un2&#8221;: unassembled / unjoined reads2 output</li>
</ul>
</li>
<li>SeqPrep will output fastq-formatted gzipped files as:<ul>
<li>&#8220;*_assembled.gz&#8221;: unassembled / unjoined reads1 output</li>
<li>&#8220;*_unassembled_R1.gz&#8221;: unassembled / unjoined reads1 output</li>
<li>&#8220;*_unassembled_R2.gz&#8221;: unassembled / unjoined reads2 output</li>
</ul>
</li>
<li>If a barcode / index file is provided via the &#8216;-b&#8217; option, an updated
barcodes file will be output as:<ul>
<li>&#8221;..._barcodes.fastq&#8221;: This barcode / index file must be used in
conjunction with the joined
paired-ends file as input to <a class="reference external" href="./split_libraries_fastq.html">split_libraries_fastq.py</a>. Except for
missing reads that may result from failed merging of paired-ends, the
index-reads and joined-reads must be in the same order.</li>
</ul>
</li>
</ol>
<p><strong>Join paired-ends with &#8216;fastq-join&#8217;:</strong></p>
<p>This is the default method to join paired-end Illumina data:</p>
<div class="highlight-python"><div class="highlight"><pre>join_paired_ends.py -f $PWD/forward_reads.fastq -r $PWD/reverse_reads.fastq -o $PWD/fastq-join_joined
</pre></div>
</div>
<p><strong>Join paired-ends with &#8216;SeqPrep&#8217;:</strong></p>
<p>Produces similar output to the &#8216;fastq-join&#8217; but returns data in gzipped format.</p>
<div class="highlight-python"><div class="highlight"><pre>join_paired_ends.py -m SeqPrep -f $PWD/forward_reads.fastq -r $PWD/reverse_reads.fastq -o $PWD/SeqPrep_joined
</pre></div>
</div>
<p><strong>Update the index / barcode reads file to match the surviving joined pairs.:</strong></p>
<p>This is required if you will be using <a class="reference external" href="./split_libraries_fastq.html">split_libraries_fastq.py</a>.</p>
<div class="highlight-python"><div class="highlight"><pre>join_paired_ends.py -f $PWD/forward_reads.fastq -r $PWD/reverse_reads.fastq -b $PWD/barcodes.fastq -o $PWD/fastq-join_joined
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    

    <h3><a href="../index.html">Site index</a></h3>
    <ul><li><ul>
    <li><a href="http://www.qiime.org">Home</a><br /></li>
    <li><a href="../install/index.html">Install</a><br /></li>
    <li><a href="../tutorials/index.html">Tutorials</a><br /></li>
    <li><a href="index.html">Scripts</a><br /></li>
    <li><a href="http://help.qiime.org">Help</a><br /></li>
    <li><a href="http://qiime.org/home_static/dataFiles.html">Resources</a><br /></li>
    <li><a href="../documentation/index.html">File Formats</a><br /></li>
    <li><a href="http://workshops.qiime.org">Workshops</a><br /></li>
    <li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
    <li><a href="../developer/index.html">Developer</a><br /></li>
    </ul></li></ul>


  <h4>Previous topic</h4>
  <p class="topless"><a href="jackknifed_beta_diversity.html"
                        title="previous chapter"><em>jackknifed_beta_diversity.py</em> &#8211; A workflow script for performing jackknifed UPGMA clustering and building jackknifed Emperor PCoA plots.</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="load_remote_mapping_file.html"
                        title="next chapter"><em>load_remote_mapping_file.py</em> &#8211; Downloads and saves a remote mapping file</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="load_remote_mapping_file.html" title="load_remote_mapping_file.py – Downloads and saves a remote mapping file"
             >next</a> |</li>
        <li class="right" >
          <a href="jackknifed_beta_diversity.html" title="jackknifed_beta_diversity.py – A workflow script for performing jackknifed UPGMA clustering and building jackknifed Emperor PCoA plots."
             >previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="../documentation/index.html" >QIIME File Formats</a> &raquo;</li>
          <li><a href="index.html" >QIIME Scripts</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.scikit-bio.org" target="_blank"><img src="../_static/powered_by_skbio.png" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2010-2014, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>