<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>QIIME installation guide &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.8.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />     
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="doc-install"></span><div class="section" id="qiime-installation-guide">
<span id="index-0"></span><h1>QIIME installation guide<a class="headerlink" href="#qiime-installation-guide" title="Permalink to this headline">¶</a></h1>
<p>QIIME consists of native python code and additionally wraps many external applications. This gives the user flexibility to easily build their own analysis pipelines, making use of popular microbial community analysis tools. QIIME handles the processing of input and output of these applications, so the user can spend time analyzing their data rather than parsing, writing, and converting file formats.</p>
<p>As a consequence of this &#8216;pipeline&#8217; architecture, <strong>QIIME has a lot of dependencies and can be very challenging to install</strong>.</p>
<div class="section" id="how-to-not-install-qiime">
<h2>How to not install QIIME<a class="headerlink" href="#how-to-not-install-qiime" title="Permalink to this headline">¶</a></h2>
<p>Because QIIME is hard to install, we have attempted to shift this burden to the QIIME development group rather than our users by providing virtual machines with QIIME and all of its dependencies pre-installed. We, and third-party developers, have also created several automated installation procedures. These alternatives (<a class="reference external" href="../index.html#downloading-and-installing-qiime">summarized here</a>) allow you to bypass the complex installation procedure and have access to a full, working QIIME installation.</p>
<p><strong>We highly recommend going with one of these solutions if you&#8217;re new to QIIME, or just want to test it out to see if it will do what you want.</strong></p>
</div>
<div class="section" id="how-to-install-qiime">
<h2>How to install QIIME<a class="headerlink" href="#how-to-install-qiime" title="Permalink to this headline">¶</a></h2>
<p>If you want to customize QIIME, work with QIIME in a multi-user environment (e.g., a Linux cluster), are interested in getting involved in QIIME development, or want to use the development version of QIIME, you may need to install QIIME manually (but you also may not - see <a class="reference external" href="http://www.wernerlab.org/software/macqiime">MacQIIME</a> and <a class="reference external" href="https://github.com/qiime/qiime-deploy">qiime-deploy</a>).</p>
<p>Depending on the features of QIIME that you or your users are interested in, a <em>QIIME base install</em> may be sufficient. This is much easier than a <em>QIIME full install</em>. Both of these procedures are covered in the following sections.</p>
<p>To install most of following dependencies, either for the base or full install, you need to have a build environment on your machine. On OS X, this involves installing the <a class="reference external" href="http://developer.apple.com/technologies/xcode.html">developer tools</a>. On Debian-based Linux (e.g., Ubuntu), this involves installing the <tt class="docutils literal"><span class="pre">build-essential</span></tt> package:</p>
<div class="highlight-python"><pre>sudo apt-get install build-essential</pre>
</div>
<div class="section" id="shortcuts-in-this-document">
<h3>Shortcuts in this document<a class="headerlink" href="#shortcuts-in-this-document" title="Permalink to this headline">¶</a></h3>
<p>For simplicity throughout this document, we assume that you have downloaded QIIME in <tt class="docutils literal"><span class="pre">$HOME/</span></tt>. You should consider all occurrences of <tt class="docutils literal"><span class="pre">$HOME/</span></tt> in the remainder of this document as references to the directory which contains the QIIME directory which you&#8217;ll have after downloading and unpacking QIIME.</p>
</div>
<div class="section" id="qiime-base-install-standard-qiime-workflow-with-default-parameters">
<h3>QIIME base install (standard QIIME workflow with default parameters)<a class="headerlink" href="#qiime-base-install-standard-qiime-workflow-with-default-parameters" title="Permalink to this headline">¶</a></h3>
<p>When getting started with QIIME, most users will want to begin with the QIIME base install. This allows users to run through the core QIIME workflows (including, but not limited to <a class="reference external" href="../scripts/validate_mapping_file.html">validate_mapping_file.py</a>, <a class="reference external" href="../scripts/split_libraries.html">split_libraries.py</a>, <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>, <a class="reference external" href="../scripts/pick_open_reference_otus.html">pick_open_reference_otus.py</a>, <a class="reference external" href="../scripts/pick_de_novo_otus.html">pick_de_novo_otus.py</a>, <a class="reference external" href="../scripts/pick_closed_reference_otus.html">pick_closed_reference_otus.py</a>, and <a class="reference external" href="../scripts/core_diversity_analyses.html">core_diversity_analyses.py</a>) while limiting the time and effort spent on installation.</p>
<p>You can install the QIIME base dependencies either via pip or by manually downloading and installing each package.</p>
<div class="section" id="installing-qiime-via-pip">
<h4>Installing QIIME via pip<a class="headerlink" href="#installing-qiime-via-pip" title="Permalink to this headline">¶</a></h4>
<p>The easiest way to install the latest QIIME release and its base dependencies is by using pip:</p>
<div class="highlight-python"><pre>pip install numpy==1.7.1
pip install qiime</pre>
</div>
<p><strong>Note:</strong> You may need to prefix the above commands with <tt class="docutils literal"><span class="pre">sudo</span></tt> if you do not have permission to place files in the default locations. You will also need an active Internet connection. If you do not have pip, the easiest way to install it is by running:</p>
<div class="highlight-python"><pre>easy_install pip</pre>
</div>
<p>That&#8217;s it!</p>
</div>
<div class="section" id="installing-qiime-on-mac-os-x-using-macqiime">
<h4>Installing QIIME on Mac OS X (Using MacQIIME)<a class="headerlink" href="#installing-qiime-on-mac-os-x-using-macqiime" title="Permalink to this headline">¶</a></h4>
<p>Jeff Werner builds and maintain MacQIIME, a package that provides a way to get a semi-full QIIME installation on Mac OS X. For more information see the <a class="reference external" href="http://www.wernerlab.org/software/macqiime">installation instructions</a>.</p>
</div>
<div class="section" id="manually-installing-qiime">
<h4>Manually installing QIIME<a class="headerlink" href="#manually-installing-qiime" title="Permalink to this headline">¶</a></h4>
<p>If you decided not to install QIIME using pip, you can install it (and its dependencies) manually.</p>
<p>The following QIIME base install dependencies are grouped by installation method.</p>
<p>The first are the core scientific python dependencies. The easiest way to install these is by installing <a class="reference external" href="https://www.enthought.com/canopy-express/">Canopy Express</a> (formerly EPD Free), which contains core modules for python scientific computing, including those required by QIIME, but also packages such as IPython and Pandas, which QIIME users may also find useful.</p>
<ul class="simple">
<li>Python 2.7.3 (<a class="reference external" href="http://www.python.org/ftp/python/2.7.3/Python-2.7.3.tgz">src_python</a>) (license: PSF)</li>
<li>Numpy 1.5.1 - 1.7.1 (<a class="reference external" href="http://sourceforge.net/projects/numpy/files/NumPy/1.7.1/numpy-1.7.1.tar.gz/download">src_numpy</a>) (license: BSD)</li>
<li>MatPlotLib 1.1.0 - 1.3.1 (<a class="reference external" href="http://downloads.sourceforge.net/project/matplotlib/matplotlib/matplotlib-1.1.0/matplotlib-1.1.0.tar.gz">src_matplotlib</a>) (license: PFS)</li>
</ul>
<p>The next are python packages not included in Canopy Express. Each of these can be installed either via pip (e.g., <tt class="docutils literal"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">&lt;package-name&gt;</span></tt>) or by downloading the package, unzipping with <tt class="docutils literal"><span class="pre">tar</span> <span class="pre">-xzf</span></tt>, changing to the resulting directory, and running <tt class="docutils literal"><span class="pre">python</span> <span class="pre">setup.py</span> <span class="pre">install</span></tt> (see <a class="reference internal" href="#python-setup"><em>Installing with setup.py</em></a> for some notes that may be useful).</p>
<ul class="simple">
<li>QIIME (see <a class="reference internal" href="#getting-qiime"><em>Getting QIIME</em></a>)</li>
<li>PyCogent 1.5.3 (<a class="reference external" href="https://pypi.python.org/packages/source/c/cogent/cogent-1.5.3.tgz">src_pycogent</a>) (license: GPL)</li>
<li>biom-format 1.3.1 (<a class="reference external" href="https://pypi.python.org/packages/source/b/biom-format/biom-format-1.3.1.tar.gz">src_biom</a>) (license: BSD)</li>
<li>qcli 0.1.0 (<a class="reference external" href="https://pypi.python.org/packages/source/q/qcli/qcli-0.1.0.tar.gz">src_qcli</a>) (license: GPL)</li>
<li>PyNAST 1.2.2 (<a class="reference external" href="https://pypi.python.org/packages/source/p/pynast/pynast-1.2.2.tar.gz">src_pynast</a>) (license: BSD)</li>
<li>Emperor 0.9.3 (<a class="reference external" href="https://pypi.python.org/packages/source/e/emperor/emperor-0.9.3.tar.gz">src_emperor</a>) (license: BSD)</li>
</ul>
<p>Next, there are two non-python dependencies required for the QIIME base package. These should be installed by following their respective install instructions.</p>
<ul class="simple">
<li>uclust 1.2.22q (<a class="reference external" href="http://www.drive5.com/uclust/downloads1_2_22q.html">src_uclust</a>) See <a class="reference internal" href="#uclust-install"><em>uclust install notes</em></a>. (licensed specially for Qiime and PyNAST users)</li>
<li>fasttree 2.1.3 (<a class="reference external" href="http://www.microbesonline.org/fasttree/FastTree-2.1.3.c">src_fasttree</a>) See <a class="reference external" href="http://www.microbesonline.org/fasttree/#Install">FastTree install instructions</a> (license: GPL)</li>
</ul>
</div>
<div class="section" id="data-files-necessary-for-the-qiime-base-installation">
<h4>Data files necessary for the QIIME base installation<a class="headerlink" href="#data-files-necessary-for-the-qiime-base-installation" title="Permalink to this headline">¶</a></h4>
<p>After you&#8217;ve installed the base QIIME dependencies, there are several data files that are likely to be useful in your analyses. These can all be obtained using <tt class="docutils literal"><span class="pre">wget</span></tt> or <tt class="docutils literal"><span class="pre">curl</span></tt>, and unzipping where necessary.</p>
<ul class="simple">
<li>greengenes core set data file (<a class="reference external" href="http://greengenes.lbl.gov/Download/Sequence_Data/Fasta_data_files/core_set_aligned.fasta.imputed">fasta</a>)</li>
<li>greengenes alignment lanemask file (<a class="reference external" href="http://greengenes.lbl.gov/Download/Sequence_Data/lanemask_in_1s_and_0s">txt</a>)</li>
<li>Marker gene reference OTUs, taxonomies, and trees  (follow the <em>Resources</em> link from the <a class="reference external" href="http://www.qiime.org">QIIME homepage</a>)</li>
</ul>
<p>You should next <a class="reference external" href="./qiime_config.html">write your QIIME config file</a>.</p>
</div>
</div>
<div class="section" id="testing-the-qiime-base-installation">
<h3>Testing the QIIME base installation<a class="headerlink" href="#testing-the-qiime-base-installation" title="Permalink to this headline">¶</a></h3>
<p>After installing the QIIME base packages, you can test this for sanity by running:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">print_qiime_config</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">tb</span>
</pre></div>
</div>
<p>You should see output that looks like the following:</p>
<div class="highlight-python"><pre>System information
==================
         Platform:      darwin
   Python version:      2.7.1 (r271:86832, Aug 30 2012, 10:07:33)  [GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)]
Python executable:      $HOME/.virtualenvs/qiime/bin/python

Dependency versions
===================
             PyCogent version:  1.5.3
                NumPy version:  1.5.1
           matplotlib version:  1.1.0
          biom-format version:  1.2.0
                 qcli version:  0.1.0
        QIIME library version:  1.7.0
         QIIME script version:  1.7.0
PyNAST version (if installed):  1.2.1
              Emperor version:  0.9.2

QIIME config values
===================
                     blastmat_dir:      None
                         sc_queue:      all.q
      topiaryexplorer_project_dir:      $HOME/code/TopiaryExplorer-0.9.1/
     pynast_template_alignment_fp:      $HOME/data/greengenes_core_sets/core_set_aligned_imputed.fasta_11_8_07.no_dots
                  cluster_jobs_fp:      start_parallel_jobs.py
pynast_template_alignment_blastdb:      None
assign_taxonomy_reference_seqs_fp:      $HOME/data/gg_13_5_otus/rep_set/97_otus.fasta
                     torque_queue:      friendlyq
   template_alignment_lanemask_fp:      $HOME/data/greengenes_core_sets/lanemask_in_1s_and_0s.txt
                    jobs_to_start:      2
                cloud_environment:      False
                qiime_scripts_dir:      $HOME/code/Qiime/scripts
            denoiser_min_per_core:      50
                      working_dir:      None
                    python_exe_fp:      python
                         temp_dir:      $HOME/temp
                      blastall_fp:      blastall
                 seconds_to_sleep:      1
assign_taxonomy_id_to_taxonomy_fp:      $HOME/data/gg_13_5_otus/taxonomy/97_otu_taxonomy.txt
................
----------------------------------------------------------------------
Ran 16 tests in 0.440s

OK</pre>
</div>
<p>This indicates that you have a complete QIIME base install.</p>
<p>You should next <a class="reference internal" href="#run-test-suite"><em>run QIIME&#8217;s unit tests</em></a>. You will experience some test failures as a result of not having a full QIIME install. If you have questions about these failures, you should post to the <a class="reference external" href="http://forum.qiime.org">QIIME Forum</a>.</p>
</div>
<div class="section" id="qiime-full-install-for-access-to-advanced-features-in-qiime-and-non-default-processing-pipelines">
<h3>QIIME full install (for access to advanced features in QIIME, and non-default processing pipelines)<a class="headerlink" href="#qiime-full-install-for-access-to-advanced-features-in-qiime-and-non-default-processing-pipelines" title="Permalink to this headline">¶</a></h3>
<p>The dependencies described below will support a full QIIME install. These are grouped by the features that each dependency will provide access to. Installation instructions should be followed for each individual package (e.g., from the project&#8217;s website or README/INSTALL file).</p>
<p>Alignment, tree-building, taxonomy assignment, OTU picking, and other data generation steps (required for non-default processing pipelines):</p>
<ul class="simple">
<li>jre1.6.0_05 (<a class="reference external" href="http://java.sun.com/javase/downloads/index.jsp">src_jre</a>) (license: GPL2)</li>
<li>rdp_classifier-2.2 (<a class="reference external" href="http://sourceforge.net/projects/rdp-classifier/files/rdp-classifier/rdp_classifier_2.2.zip/download">src_rdp</a>) See <a class="reference internal" href="#rdp-install"><em>RDP install notes</em></a>. (license: GPL)</li>
<li>tax2tree 1.0.0 (<a class="reference external" href="https://downloads.sourceforge.net/project/tax2tree/tax2tree-v1.0.tar.gz">src_tax2tree</a>)</li>
<li>blast-2.2.22 (legacy BLAST from NCBI, <em>NOT</em> BLAST+) (<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/release/2.2.22/blast-2.2.22-universal-macosx.tar.gz">OS X</a> or <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/release/2.2.22/blast-2.2.22-ia32-linux.tar.gz">linux 32-bit</a>) (license: GNU)</li>
<li>cd-hit 3.1.1 (<a class="reference external" href="http://www.bioinformatics.org/download/cd-hit/cd-hit-2007-0131.tar.gz">src_cdhit</a>) (license: Free access)</li>
<li>ChimeraSlayer (via microbiomeutil_2010-04-29) (<a class="reference external" href="http://sourceforge.net/projects/microbiomeutil/files/">src_chimeraslayer</a>) See <a class="reference internal" href="#chimeraslayer-install"><em>ChimeraSlayer install notes</em></a>.</li>
<li>mothur 1.25.0 (<a class="reference external" href="http://www.mothur.org/w/images/6/6d/Mothur.1.25.0.zip">src_mothur</a>) (license: GPL)</li>
<li>clearcut v1.0.9 (<a class="reference external" href="http://www.mothur.org/w/images/9/91/Clearcut.source.zip">src_clearcut</a>)</li>
<li>raxml 7.3.0 (<a class="reference external" href="ftp://thebeast.colorado.edu/pub/QIIME-v1.5.0-dependencies/stamatak-standard-RAxML-5_7_2012.tgz">src_raxml</a>)</li>
<li>infernal 1.0.2 (<a class="reference external" href="ftp://selab.janelia.org/pub/software/infernal/infernal.tar.gz">src_infernal</a>) (license: GPL)</li>
<li>cdbtools (<a class="reference external" href="ftp://occams.dfci.harvard.edu/pub/bio/tgi/software/cdbfasta/cdbfasta.tar.gz">src_cdbtools</a>)</li>
<li>muscle 3.8.31 (<a class="reference external" href="http://www.drive5.com/muscle/downloads.htm">src_muscle</a>) (Public domain)</li>
<li>rtax 0.984 (<a class="reference external" href="http://static.davidsoergel.com/rtax-0.984.tgz">src_rtax</a>) (license: BSD)</li>
<li>pplacer 1.1 (<a class="reference external" href="http://matsen.fhcrc.org/pplacer/builds/pplacer-v1.1-Linux.tar.gz">src_pplacer</a>) (license: GPL)</li>
<li>ParsInsert 1.04 (<a class="reference external" href="http://downloads.sourceforge.net/project/parsinsert/ParsInsert.1.04.tgz">src_parsinsert</a>) (license: GPL)</li>
<li>usearch v5.2.236 and/or usearch v6.1 (<a class="reference external" href="http://www.drive5.com/usearch/">src_usearch</a>) (license: see <a class="reference external" href="http://www.drive5.com/usearch/nonprofit_form.html">http://www.drive5.com/usearch/nonprofit_form.html</a>) <strong>At this stage two different versions of usearch are supported.</strong> usearch v5.2.236 is referred to as <tt class="docutils literal"><span class="pre">usearch</span></tt> in QIIME, and usearch v6.1 is referred to as <tt class="docutils literal"><span class="pre">usearch61</span></tt>.</li>
</ul>
<p>Processing sff files:</p>
<ul class="simple">
<li>sfffile and sffinfo (optional, QIIME 1.2.0 and later contain built-in tools for processing sff files although they are about 10x slower than the tools from Roche) (license: proprietary - must be obtained from Roche/454)</li>
</ul>
<p>Denoising 454 data:</p>
<ul class="simple">
<li>GNU Science Library (required by AmpliconNoise) (<a class="reference external" href="ftp://ftp.gnu.org/gnu/gsl/gsl-1.9.tar.gz">src_gsl</a>)</li>
<li>AmpliconNoise 1.27 (<a class="reference external" href="http://ampliconnoise.googlecode.com/files/AmpliconNoiseV1.27.tar.gz">src_ampliconnoise</a>) See <a class="reference internal" href="#ampliconnoise-install"><em>AmpliconNoise install notes</em></a>.</li>
<li>ghc 6.8 (required by the QIIME denoiser) (<a class="reference external" href="http://haskell.org/ghc">src_ghc</a>)</li>
</ul>
<p>Visualization and plotting steps:</p>
<ul class="simple">
<li>cytoscape v2.7.0 (<a class="reference external" href="http://www.cytoscape.org/">src_cytoscape</a>) (license: LGPL)</li>
</ul>
<p>Supervised learning (<tt class="docutils literal"><span class="pre">supervised_learning.py</span></tt>) and <tt class="docutils literal"><span class="pre">compare_categories.py</span></tt>:</p>
<ul class="simple">
<li>R 3.0.2 (<a class="reference external" href="http://www.r-project.org/">src_r</a>) See <a class="reference internal" href="#r-install"><em>R install notes</em></a>. (license: GPL2)</li>
</ul>
<p>If you plan to build the QIIME documentation locally:</p>
<ul class="simple">
<li>Sphinx 1.0.4 (<a class="reference external" href="http://pypi.python.org/pypi/Sphinx">src</a>) See <a class="reference internal" href="#build-qiime-docs"><em>Building the QIIME documentation</em></a> (license: BSD)</li>
</ul>
<p>If you plan to use remote mapping files (stored as Google Spreadsheets) with QIIME (see the tutorial <a class="reference external" href="../tutorials/remote_mapping_files.html">here</a>):</p>
<ul class="simple">
<li>gdata 2.0.17 (<a class="reference external" href="http://gdata-python-client.googlecode.com/files/gdata-2.0.17.tar.gz">src</a>) (license: Apache 2.0)</li>
</ul>
<p>If you plan to use SourceTracker with QIIME:</p>
<ul class="simple">
<li>SourceTracker 0.9.5 (<a class="reference external" href="http://downloads.sourceforge.net/project/sourcetracker/sourcetracker-0.9.5.tar.gz">src</a>) (license: GPL)</li>
</ul>
</div>
<div class="section" id="testing-the-qiime-full-installation">
<h3>Testing the QIIME full installation<a class="headerlink" href="#testing-the-qiime-full-installation" title="Permalink to this headline">¶</a></h3>
<p>After installing the QIIME base packages, you can test this for sanity by running:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">print_qiime_config</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">t</span>
</pre></div>
</div>
<p>You should see output that looks like the following:</p>
<div class="highlight-python"><pre>System information
==================
         Platform:      darwin
   Python version:      2.7.1 (r271:86832, Aug 30 2012, 10:07:33)  [GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)]
Python executable:      $HOME/.virtualenvs/qiime/bin/python

Dependency versions
===================
                     PyCogent version:  1.5.3
                        NumPy version:  1.5.1
                   matplotlib version:  1.1.0
                  biom-format version:  1.2.0-dev
                         qcli version:  0.1.0
                QIIME library version:  1.7.0-dev
                 QIIME script version:  1.7.0-dev
        PyNAST version (if installed):  1.2.1
                      Emperor version:  0.9.2-dev
RDP Classifier version (if installed):  rdp_classifier-2.2.jar
          Java version (if installed):  1.6.0_43

QIIME config values
===================
                     blastmat_dir:      /Applications/blast-2.2.22/data/
                         sc_queue:      all.q
      topiaryexplorer_project_dir:      $HOME/code/TopiaryExplorer-0.9.1/
     pynast_template_alignment_fp:      $HOME/data/greengenes_core_sets/core_set_aligned_imputed.fasta_11_8_07.no_dots
                  cluster_jobs_fp:      start_parallel_jobs.py
pynast_template_alignment_blastdb:      None
assign_taxonomy_reference_seqs_fp:      $HOME/data/gg_13_5_otus/rep_set/97_otus.fasta
                     torque_queue:      friendlyq
   template_alignment_lanemask_fp:      $HOME/data/greengenes_core_sets/lanemask_in_1s_and_0s.txt
                    jobs_to_start:      2
                cloud_environment:      False
                qiime_scripts_dir:      $HOME/code/Qiime/scripts
            denoiser_min_per_core:      50
                      working_dir:      None
                    python_exe_fp:      python
                         temp_dir:      $HOME/temp
                      blastall_fp:      blastall
                 seconds_to_sleep:      1
assign_taxonomy_id_to_taxonomy_fp:      $HOME/data/gg_13_5_otus/taxonomy/97_otu_taxonomy.txt
...................................
----------------------------------------------------------------------
Ran 35 tests in 0.641s

OK</pre>
</div>
<p>You should next <a class="reference internal" href="#run-test-suite"><em>run QIIME&#8217;s unit tests</em></a>. All tests should pass if you have a working full QIIME installation. If you have questions about these failures, you should post to the <a class="reference external" href="http://forum.qiime.org">QIIME Forum</a>.</p>
</div>
</div>
</div>
<div class="section" id="qiime-installation-guide-additional-notes">
<h1>QIIME installation guide: Additional notes<a class="headerlink" href="#qiime-installation-guide-additional-notes" title="Permalink to this headline">¶</a></h1>
<p>The following sections are referenced from the installation guide above.</p>
<div class="section" id="getting-qiime">
<span id="id4"></span><h2>Getting QIIME<a class="headerlink" href="#getting-qiime" title="Permalink to this headline">¶</a></h2>
<p>First, change to the directory where you would like to download QIIME:</p>
<div class="highlight-python"><pre>cd $HOME</pre>
</div>
<div class="section" id="stable-release">
<h3>Stable Release<a class="headerlink" href="#stable-release" title="Permalink to this headline">¶</a></h3>
<p>Currently the most stable version of QIIME is our 1.8.0 release, which you can download from <a class="reference external" href="https://pypi.python.org/pypi/qiime">here</a>.</p>
</div>
<div class="section" id="latest-development-version">
<h3>Latest Development Version<a class="headerlink" href="#latest-development-version" title="Permalink to this headline">¶</a></h3>
<p>To get the latest development version of QIIME, you should check it out of our git repository, which is hosted on GitHub. While this code is subject to changes in interface and hasn&#8217;t been as extensively tested as the release version, it will provide access to the latest and greatest QIIME features. The official web documentation is likely to be out-of-date with respect to the development software. You should instead refer to the documentation in <tt class="docutils literal"><span class="pre">Qiime/doc</span></tt>. Check out the latest version of QIIME using git with the command:</p>
<div class="highlight-python"><pre>git clone git://github.com/qiime/qiime.git Qiime</pre>
</div>
<p>If you are using the latest development version of QIIME, you should periodically update your checkout by running the following command (from within your checkout):</p>
<div class="highlight-python"><pre>git pull</pre>
</div>
</div>
<div class="section" id="unpacking-qiime-release-only">
<h3>Unpacking QIIME (release only)<a class="headerlink" href="#unpacking-qiime-release-only" title="Permalink to this headline">¶</a></h3>
<p>After downloading the QIIME release tar file you&#8217;ll need to unpack the code. For simplicity in this document, we will assume that you have downloaded QIIME to the directory <tt class="docutils literal"><span class="pre">$HOME/</span></tt>.</p>
<p>Unpack the release .tar.gz file with the commands:</p>
<div class="highlight-python"><pre>cd $HOME
tar -xvzf qiime-1.8.0.tar.gz
ln -s $HOME/qiime-1.8.0 $HOME/Qiime</pre>
</div>
<p>If you have downloaded the development version from GitHub, QIIME is already unpacked.</p>
</div>
<div class="section" id="installing-qiime">
<h3>Installing QIIME<a class="headerlink" href="#installing-qiime" title="Permalink to this headline">¶</a></h3>
<p>QIIME consists of library code (in <tt class="docutils literal"><span class="pre">Qiime/qiime</span></tt>), test code (in <tt class="docutils literal"><span class="pre">Qiime/tests</span></tt>), example script input and output (in <tt class="docutils literal"><span class="pre">Qiime/qiime_test_data</span></tt>), documentation (in <tt class="docutils literal"><span class="pre">Qiime/doc</span></tt>), and scripts (in <tt class="docutils literal"><span class="pre">Qiime/scripts</span></tt>). Installing QIIME consists installing the library code in a place where python knows where to find it, and installing the scripts in a place where the shell looks for executable files, and running the tests (optional, but highly recommended).</p>
</div>
<div class="section" id="installing-with-setup-py">
<span id="python-setup"></span><h3>Installing with setup.py<a class="headerlink" href="#installing-with-setup-py" title="Permalink to this headline">¶</a></h3>
<p>Using <tt class="docutils literal"><span class="pre">Qiime/setup.py</span></tt> (and thereby python&#8217;s <tt class="docutils literal"><span class="pre">distutils</span></tt> package) is the recommended way of installing the Qiime library code and scripts. You can optionally specify where the library code and scripts should be installed &#8211; depending on your setup, you may want to do this. By default, the QIIME library code will be placed under python&#8217;s <tt class="docutils literal"><span class="pre">site-packages</span></tt>, and the QIIME scripts will be place in <tt class="docutils literal"><span class="pre">/usr/local/bin/</span></tt>. You may need to run <tt class="docutils literal"><span class="pre">setup.py</span></tt> using <tt class="docutils literal"><span class="pre">sudo</span></tt> if you do not have permission to place files in the default locations.</p>
<p>First, ensure that you are in the top-level QIIME directory:</p>
<div class="highlight-python"><pre>cd $HOME/Qiime</pre>
</div>
<p>By default the QIIME scripts will be installed in <tt class="docutils literal"><span class="pre">/usr/local/bin</span></tt>. As there are a lot of QIIME scripts, we highly recommend customizing the script directory to keep your system organized. This can be customized with the <tt class="docutils literal"><span class="pre">--install_scripts</span></tt> option. You also can specify an alternate directory for the library files with <tt class="docutils literal"><span class="pre">--install-purelib</span></tt>. An example command is:</p>
<div class="highlight-python"><pre>python setup.py install --install-scripts=$HOME/bin/ --install-purelib=$HOME/lib/</pre>
</div>
<p>For a complete discussion of customizations related to the setup.py script, <a class="reference external" href="http://docs.python.org/release/2.7.1/install/index.html#alternate-installation">see this page</a>.</p>
<p>If you used default values for <tt class="docutils literal"><span class="pre">--install-scripts</span></tt> and <tt class="docutils literal"><span class="pre">--install-purelib</span></tt> (by not specifying them), your installation should be complete. If you specified an alternate value for <tt class="docutils literal"><span class="pre">--install-scripts</span></tt>, you&#8217;ll need to ensure that the shell knows where to look for the scripts. If you are using the bash shell and the locations specified in the examples above, you can do this with the following command:</p>
<div class="highlight-python"><pre>echo "export PATH=$HOME/bin/:$PATH" &gt;&gt; $HOME/.bashrc</pre>
</div>
<p>If you specified an alternate value for <tt class="docutils literal"><span class="pre">--install-purelib</span></tt>, you&#8217;ll need to be sure that python knows where to look for Qiime. If you are using the bash shell and the locations specified in the examples above, you can do this with the following command:</p>
<div class="highlight-python"><pre>echo "export PYTHONPATH=$HOME/lib/:$PYTHONPATH" &gt;&gt; $HOME/.bashrc</pre>
</div>
<p>The source your <tt class="docutils literal"><span class="pre">.bashrc</span></tt>:</p>
<div class="highlight-python"><pre>source $HOME/.bashrc</pre>
</div>
<p id="set-script-dir">Finally, you&#8217;ll need to create and edit a custom <tt class="docutils literal"><span class="pre">qiime_config</span></tt> file to tell QIIME where to look for the QIIME scripts. Create a custom <tt class="docutils literal"><span class="pre">qiime_config</span></tt> file by copying the default <tt class="docutils literal"><span class="pre">qiime_config</span></tt> packaged with QIIME:</p>
<div class="highlight-python"><pre>cp $HOME/Qiime/qiime/support_files/qiime_config $HOME/.qiime_config</pre>
</div>
<p>Open the new file, <tt class="docutils literal"><span class="pre">$HOME/.qiime_config</span></tt>, in a text editor such as TextEdit (on Mac), gedit (on Linux), vim, or emacs (but not Microsoft Word, which is a <a class="reference external" href="http://en.wikipedia.org/wiki/Word_processor">word processor</a>, not a <a class="reference external" href="http://en.wikipedia.org/wiki/Text_editor">text editor</a>!). Find the line beginning <tt class="docutils literal"><span class="pre">qiime_scripts_dir</span></tt> and add a tab, followed by the QIIME scripts directory. If you&#8217;ve used the default value (i.e., you didn&#8217;t specify <tt class="docutils literal"><span class="pre">--install-scripts</span></tt>) the value you add will be <tt class="docutils literal"><span class="pre">/usr/local/bin/</span></tt>. Otherwise, specify the value that you provided for <tt class="docutils literal"><span class="pre">--install-scripts</span></tt>. In the example above, this would look like:</p>
<div class="highlight-python"><pre>qiime_scripts_dir       $HOME/bin/</pre>
</div>
<p>Note that the delimiter between the key and the value here is a tab, not a space! For additional information on the qiime_config file, <a class="reference external" href="./qiime_config.html">see this document</a>.</p>
</div>
<div class="section" id="running-the-test-suite">
<span id="run-test-suite"></span><h3>Running the test suite<a class="headerlink" href="#running-the-test-suite" title="Permalink to this headline">¶</a></h3>
<p>Next you should run the test suite. Execute the following commands:</p>
<div class="highlight-python"><pre>cd $HOME/Qiime/tests/
python all_tests.py</pre>
</div>
<p>You will see test output on the terminal indicating test successes and failures. Some failures are OK. The <tt class="docutils literal"><span class="pre">all_tests.py</span></tt> command will complete with a summary of test failures. Some tests may fail due to missing external applications &#8211; these will be noted separately from other test failures. If these are related to features of QIIME that you are not using, this is acceptable. Otherwise, you&#8217;ll need to ensure that you have the external applications installed correctly (and the correct versions), and re-run the tests.</p>
</div>
</div>
<div class="section" id="license-information-for-external-dependencies">
<h2>License information for external dependencies<a class="headerlink" href="#license-information-for-external-dependencies" title="Permalink to this headline">¶</a></h2>
<p>We have attempted to provide accurate licensing information for the above dependencies for the convenience of our users. This information is by no means definitive and may contain errors. Any questions about licenses or the legality of specific uses of these software packages should be directed to the authors of the software. Do not rely solely on the license information presented above!</p>
</div>
<div class="section" id="external-application-install-notes">
<h2>External application install notes<a class="headerlink" href="#external-application-install-notes" title="Permalink to this headline">¶</a></h2>
<div class="section" id="path-environment-variable">
<h3>PATH Environment Variable<a class="headerlink" href="#path-environment-variable" title="Permalink to this headline">¶</a></h3>
<p>External applications used by QIIME need to be visible to the shell by existing in executable search path (i.e., listed in the <tt class="docutils literal"><span class="pre">$PATH</span></tt> environment variable). For example, if you plan to use cd-hit, and have the cd-hit executables installed in <tt class="docutils literal"><span class="pre">$HOME/bin</span></tt> you can add this directory to your system path with the commands:</p>
<div class="highlight-python"><pre>echo "export PATH=$HOME/bin/:$PATH" &gt;&gt; $HOME/.bashrc
source $HOME/.bashrc</pre>
</div>
</div>
<div class="section" id="pythonpath-environment-variable">
<h3>PYTHONPATH Environment Variable<a class="headerlink" href="#pythonpath-environment-variable" title="Permalink to this headline">¶</a></h3>
<p>Qiime, PyCogent, and NumPy must be visible to python for all features of QIIME. matplotlib must be visible to python if you plan to use graphics features of QIIME; PyNAST must be visible to python if you plan to use PyNAST for multiple sequence alignment; and Denoiser must be visible to python if you plan to denoise 454 data. With the exception of Denoiser, all of these packages come with setup.py scripts. If you have used these, you should not need to modify your PYTHONPATH to make the library code visible. If you haven&#8217;t used the respective setup.py scripts, or if you specified an alternate value for <tt class="docutils literal"><span class="pre">--install-purelib</span></tt>, you may need to add the locations of these libraries to your PYTHONPATH environment variable.</p>
<p>For example, if you&#8217;ve installed PyNAST in <tt class="docutils literal"><span class="pre">$HOME/PyNAST</span></tt> you can add this to your PYTHONPATH with the commands:</p>
<div class="highlight-python"><pre>echo "export PYTHONPATH=$HOME/PyNAST/:$PYTHONPATH" &gt;&gt; $HOME/.bashrc
source $HOME/.bashrc</pre>
</div>
</div>
<div class="section" id="rdp-jar-path-environment-variable">
<span id="rdp-install"></span><h3>RDP_JAR_PATH Environment Variable<a class="headerlink" href="#rdp-jar-path-environment-variable" title="Permalink to this headline">¶</a></h3>
<p>If you plan to use the RDP classifier for taxonomy assignment you must define an <tt class="docutils literal"><span class="pre">RDP_JAR_PATH</span></tt> environment variable. If you downloaded and unzipped the RDP classifier folder in <tt class="docutils literal"><span class="pre">$HOME/app/</span></tt>, you can do this with the following commands:</p>
<div class="highlight-python"><pre>echo "export RDP_JAR_PATH=$HOME/app/rdp_classifier_2.2/rdp_classifier-2.2.jar" &gt;&gt; $HOME/.bashrc
source $HOME/.bashrc</pre>
</div>
<p>Note that you will need the contents inside <tt class="docutils literal"><span class="pre">rdp_classifier_2.2</span></tt> for the program to function properly.</p>
</div>
<div class="section" id="uclust-install-notes">
<span id="uclust-install"></span><h3>uclust Install Notes<a class="headerlink" href="#uclust-install-notes" title="Permalink to this headline">¶</a></h3>
<p>The uclust binary must be called <tt class="docutils literal"><span class="pre">uclust</span></tt>, which differs from the names of the posted binaries, but is the name of the binary if you build from source. If you&#8217;ve installed the binary <tt class="docutils literal"><span class="pre">uclust1.2.21q_i86linux64</span></tt> as <tt class="docutils literal"><span class="pre">$HOME/bin/uclust1.2.21q_i86linux64</span></tt>, we recommend creating a symbolic link to this file:</p>
<div class="highlight-python"><pre>ln -s $HOME/bin/uclust1.2.21q_i86linux64 $HOME/bin/uclust</pre>
</div>
</div>
<div class="section" id="usearch-install-notes">
<span id="usearch-install"></span><h3>usearch Install Notes<a class="headerlink" href="#usearch-install-notes" title="Permalink to this headline">¶</a></h3>
<p>The usearch binary must be called <tt class="docutils literal"><span class="pre">usearch</span></tt>, which differs from the names of the posted binaries, but is the name of the binary if you build from source. If you&#8217;ve installed the binary <tt class="docutils literal"><span class="pre">usearch5.2.236_i86linux32</span></tt> as <tt class="docutils literal"><span class="pre">$HOME/bin/usearch5.2.236_i86linux32</span></tt>, we recommend creating a symbolic link to this file:</p>
<div class="highlight-python"><pre>ln -s $HOME/bin/usearch5.2.236_i86linux32 $HOME/bin/usearch</pre>
</div>
</div>
<div class="section" id="chimeraslayer-install-notes">
<span id="chimeraslayer-install"></span><h3>ChimeraSlayer Install Notes<a class="headerlink" href="#chimeraslayer-install-notes" title="Permalink to this headline">¶</a></h3>
<p>ChimeraSlayer can only be run from the directory where it was unpacked and built as it depends on several of its dependencies being in specific places relative to the executable (<tt class="docutils literal"><span class="pre">ChimeraSlayer/ChimeraSlayer.pl</span></tt>). Carefully follow the ChimeraSlayer install instructions. Then add the directory containing <tt class="docutils literal"><span class="pre">ChimeraSlayer.pl</span></tt> to your <tt class="docutils literal"><span class="pre">$PATH</span></tt> environment variable. If your <tt class="docutils literal"><span class="pre">ChimeraSlayer</span></tt> folder is in <tt class="docutils literal"><span class="pre">$HOME/app/</span></tt> you can set the <tt class="docutils literal"><span class="pre">$PATH</span></tt> environment variable as follows:</p>
<div class="highlight-python"><pre>echo "export PATH=$HOME/app/ChimeraSlayer:$PATH" &gt;&gt; $HOME/.bashrc
source $HOME/.bashrc</pre>
</div>
<p>If you&#8217;re having trouble getting ChimeraSlayer to work via QIIME, you should first check to see if you can run it directly from a directory other than its install directory. For example, try running <tt class="docutils literal"><span class="pre">ChimeraSlayer.pl</span></tt> from your home directory.</p>
<p>Once you have configured Qiime, you can test your ChimeraSlayer install by running:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">print_qiime_config</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">t</span>
</pre></div>
</div>
<p>This includes a check for obvious problems with your ChimeraSlayer install, and should help you determine if you have it installed correctly.</p>
</div>
<div class="section" id="r-install-notes">
<span id="r-install"></span><h3>R Install Notes<a class="headerlink" href="#r-install-notes" title="Permalink to this headline">¶</a></h3>
<p>To install R visit <a class="reference external" href="http://www.r-project.org/">http://www.r-project.org/</a> and follow the install instructions. Once R is installed, run R and excecute the following commands:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s">&#39;randomForest&#39;</span><span class="p">)</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s">&#39;optparse&#39;</span><span class="p">)</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s">&#39;vegan&#39;</span><span class="p">)</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s">&#39;ape&#39;</span><span class="p">)</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s">&#39;MASS&#39;</span><span class="p">)</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s">&#39;gtools&#39;</span><span class="p">)</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s">&#39;klaR&#39;</span><span class="p">)</span>
<span class="n">install</span><span class="o">.</span><span class="n">packages</span><span class="p">(</span><span class="s">&#39;RColorBrewer&#39;</span><span class="p">)</span>
<span class="n">q</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="ampliconnoise-install-notes">
<span id="ampliconnoise-install"></span><h3>AmpliconNoise Install Notes<a class="headerlink" href="#ampliconnoise-install-notes" title="Permalink to this headline">¶</a></h3>
<p>AmpliconNoise requires that several environment variables are set. After you&#8217;ve installed AmpliconNoise, you can set these with the following commands (assuming your AmpliconNoise install directory is <tt class="docutils literal"><span class="pre">$HOME/AmpliconNoiseV1.27/</span></tt>):</p>
<div class="highlight-python"><pre>echo "export PATH=$HOME/AmpliconNoiseV1.27/Scripts:$HOME/AmpliconNoiseV1.27/bin:$PATH" &gt;&gt; $HOME/.bashrc

echo "export PYRO_LOOKUP_FILE=$HOME/AmpliconNoiseV1.27/Data/LookUp_E123.dat" &gt;&gt; $HOME/.bashrc
echo "export SEQ_LOOKUP_FILE=$HOME/AmpliconNoiseV1.27/Data/Tran.dat" &gt;&gt; $HOME/.bashrc</pre>
</div>
</div>
<div class="section" id="qiime-denoiser-install-notes">
<h3>QIIME Denoiser Install Notes<a class="headerlink" href="#qiime-denoiser-install-notes" title="Permalink to this headline">¶</a></h3>
<p>If you do not install QIIME using <tt class="docutils literal"><span class="pre">setup.py</span></tt> and you plan to use the QIIME Denoiser, you&#8217;ll need to compile the FlowgramAlignment program. To do this you&#8217;ll need to have <tt class="docutils literal"><span class="pre">ghc</span></tt> installed. Then from the <tt class="docutils literal"><span class="pre">Qiime/qiime/support_files/denoiser/FlowgramAlignment/</span></tt> directory, run the following command:</p>
<div class="highlight-python"><pre>make ; make install</pre>
</div>
</div>
</div>
<div class="section" id="building-the-qiime-documentation">
<span id="build-qiime-docs"></span><h2>Building The QIIME Documentation<a class="headerlink" href="#building-the-qiime-documentation" title="Permalink to this headline">¶</a></h2>
<p>If you are using the development version of QIIME, you may want to build the documentation locally for access to the latest version. You can change to the <tt class="docutils literal"><span class="pre">Qiime/doc</span></tt> directory and run:</p>
<div class="highlight-python"><pre>make html</pre>
</div>
<p>We try to update the documentation as we update the code, but development version users may notice some discrepancies. After building the documentation, you can view it in a web browser by opening the file <tt class="docutils literal"><span class="pre">Qiime/doc/_build/html/index.html</span></tt>. You may want to bookmark that page for easy access.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">QIIME installation guide</a><ul>
<li><a class="reference internal" href="#how-to-not-install-qiime">How to not install QIIME</a></li>
<li><a class="reference internal" href="#how-to-install-qiime">How to install QIIME</a><ul>
<li><a class="reference internal" href="#shortcuts-in-this-document">Shortcuts in this document</a></li>
<li><a class="reference internal" href="#qiime-base-install-standard-qiime-workflow-with-default-parameters">QIIME base install (standard QIIME workflow with default parameters)</a><ul>
<li><a class="reference internal" href="#installing-qiime-via-pip">Installing QIIME via pip</a></li>
<li><a class="reference internal" href="#installing-qiime-on-mac-os-x-using-macqiime">Installing QIIME on Mac OS X (Using MacQIIME)</a></li>
<li><a class="reference internal" href="#manually-installing-qiime">Manually installing QIIME</a></li>
<li><a class="reference internal" href="#data-files-necessary-for-the-qiime-base-installation">Data files necessary for the QIIME base installation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#testing-the-qiime-base-installation">Testing the QIIME base installation</a></li>
<li><a class="reference internal" href="#qiime-full-install-for-access-to-advanced-features-in-qiime-and-non-default-processing-pipelines">QIIME full install (for access to advanced features in QIIME, and non-default processing pipelines)</a></li>
<li><a class="reference internal" href="#testing-the-qiime-full-installation">Testing the QIIME full installation</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#qiime-installation-guide-additional-notes">QIIME installation guide: Additional notes</a><ul>
<li><a class="reference internal" href="#getting-qiime">Getting QIIME</a><ul>
<li><a class="reference internal" href="#stable-release">Stable Release</a></li>
<li><a class="reference internal" href="#latest-development-version">Latest Development Version</a></li>
<li><a class="reference internal" href="#unpacking-qiime-release-only">Unpacking QIIME (release only)</a></li>
<li><a class="reference internal" href="#installing-qiime">Installing QIIME</a></li>
<li><a class="reference internal" href="#installing-with-setup-py">Installing with setup.py</a></li>
<li><a class="reference internal" href="#running-the-test-suite">Running the test suite</a></li>
</ul>
</li>
<li><a class="reference internal" href="#license-information-for-external-dependencies">License information for external dependencies</a></li>
<li><a class="reference internal" href="#external-application-install-notes">External application install notes</a><ul>
<li><a class="reference internal" href="#path-environment-variable">PATH Environment Variable</a></li>
<li><a class="reference internal" href="#pythonpath-environment-variable">PYTHONPATH Environment Variable</a></li>
<li><a class="reference internal" href="#rdp-jar-path-environment-variable">RDP_JAR_PATH Environment Variable</a></li>
<li><a class="reference internal" href="#uclust-install-notes">uclust Install Notes</a></li>
<li><a class="reference internal" href="#usearch-install-notes">usearch Install Notes</a></li>
<li><a class="reference internal" href="#chimeraslayer-install-notes">ChimeraSlayer Install Notes</a></li>
<li><a class="reference internal" href="#r-install-notes">R Install Notes</a></li>
<li><a class="reference internal" href="#ampliconnoise-install-notes">AmpliconNoise Install Notes</a></li>
<li><a class="reference internal" href="#qiime-denoiser-install-notes">QIIME Denoiser Install Notes</a></li>
</ul>
</li>
<li><a class="reference internal" href="#building-the-qiime-documentation">Building The QIIME Documentation</a></li>
</ul>
</li>
</ul>


	<h3><a href="../index.html">Site index</a></h3>
	<ul><li><ul>
	<li><a href="../index.html">Home</a><br /></li>
	<li><a href="index.html">Install</a><br /></li>
	<li><a href="../documentation/index.html">Documentation</a><br /></li>
	<li><a href="../tutorials/index.html">Tutorials</a><br /></li>
	<li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
	<li><a href="../developer/index.html">Developer</a><br /></li>
	</ul></li></ul>


<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">Home</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.pycogent.org" target="_blank"><img src="../_static/PoweredByPycogent.jpg" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2010-2014, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>