<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>core_diversity_analyses.py – A workflow for running a core set of QIIME diversity analyses. &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />
    <link rel="up" title="QIIME Scripts" href="index.html" />
    <link rel="next" title="count_seqs.py –" href="count_seqs.html" />
    <link rel="prev" title="convert_fastaqual_fastq.py – From a FASTA file and a matching QUAL file, generates a FASTQ file. From FASTQ file generates FASTA file and matching QUAL file." href="convert_fastaqual_fastq.html" />
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="count_seqs.html" title="count_seqs.py –"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="convert_fastaqual_fastq.html" title="convert_fastaqual_fastq.py – From a FASTA file and a matching QUAL file, generates a FASTQ file. From FASTQ file generates FASTA file and matching QUAL file."
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="../documentation/index.html" >QIIME File Formats</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">QIIME Scripts</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="core-diversity-analyses"></span><div class="section" id="core-diversity-analyses-py-a-workflow-for-running-a-core-set-of-qiime-diversity-analyses">
<span id="index-0"></span><h1><em>core_diversity_analyses.py</em> &#8211; A workflow for running a core set of QIIME diversity analyses.<a class="headerlink" href="#core-diversity-analyses-py-a-workflow-for-running-a-core-set-of-qiime-diversity-analyses" title="Permalink to this headline">¶</a></h1>
<p><strong>Description:</strong></p>
<p>This script plugs several QIIME diversity analyses together to form a basic workflow beginning with a BIOM table, mapping file, and optional phylogenetic tree.</p>
<p>The included scripts are those run by the workflow scripts <a class="reference external" href="./alpha_rarefaction.html">alpha_rarefaction.py</a>, <a class="reference external" href="./beta_diversity_through_plots.html">beta_diversity_through_plots.py</a>, <a class="reference external" href="./summarize_taxa_through_plots.html">summarize_taxa_through_plots.py</a>, plus the (non-workflow) scripts <a class="reference external" href="./make_distance_boxplots.html">make_distance_boxplots.py</a>, <a class="reference external" href="./compare_alpha_diversity.html">compare_alpha_diversity.py</a>, and <a class="reference external" href="./group_significance.html">group_significance.py</a>. To update parameters to the workflow scripts, you should pass the same parameters file that you would pass if calling the workflow script directly.</p>
<p>Additionally, a table summary is generated by running the &#8216;biom summarize-table&#8217; command (part of the biom-format package). To update parameters to this command, your parameters file should use &#8216;biom-summarize-table&#8217; (without quotes) as the script name. See <a class="reference external" href="http://qiime.org/documentation/qiime_parameters_files.html">http://qiime.org/documentation/qiime_parameters_files.html</a> for more details.</p>
<p><strong>Usage:</strong> <tt class="file docutils literal"><span class="pre">core_diversity_analyses.py</span> <span class="pre">[options]</span></tt></p>
<p><strong>Input Arguments:</strong></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><strong>[REQUIRED]</strong></p>
<dl class="docutils">
<dt>-i, <cite>-</cite>-input_biom_fp</dt>
<dd>The input biom file [REQUIRED]</dd>
<dt>-o, <cite>-</cite>-output_dir</dt>
<dd>The output directory [REQUIRED]</dd>
<dt>-m, <cite>-</cite>-mapping_fp</dt>
<dd>The mapping filepath [REQUIRED]</dd>
<dt>-e, <cite>-</cite>-sampling_depth</dt>
<dd>Sequencing depth to use for even sub-sampling and maximum rarefaction depth. You should review the output of the &#8216;biom summarize-table&#8217; command to decide on this value.</dd>
</dl>
<p><strong>[OPTIONAL]</strong></p>
<dl class="last docutils">
<dt>-p, <cite>-</cite>-parameter_fp</dt>
<dd>Path to the parameter file, which specifies changes to the default behavior. For more information, see www.qiime.org/documentation/qiime_parameters_files.html [if omitted, default values will be used]</dd>
<dt>-a, <cite>-</cite>-parallel</dt>
<dd>Run in parallel where available. Specify number of jobs to start with -O or in the parameters file. [default: False]</dd>
<dt><cite>-</cite>-nonphylogenetic_diversity</dt>
<dd>Apply non-phylogenetic alpha (chao1 and observed_otus) and beta (bray_curtis) diversity calculations. This is useful if, for example, you are working with non-amplicon BIOM tables, or if a reliable tree is not available (e.g., if you&#8217;re  working with ITS amplicons) [default: False]</dd>
<dt><cite>-</cite>-suppress_taxa_summary</dt>
<dd>Suppress generation of taxa summary plots. [default: False]</dd>
<dt><cite>-</cite>-suppress_beta_diversity</dt>
<dd>Suppress beta diversity analyses. [default: False]</dd>
<dt><cite>-</cite>-suppress_alpha_diversity</dt>
<dd>Suppress alpha diversity analyses. [default: False]</dd>
<dt><cite>-</cite>-suppress_group_significance</dt>
<dd>Suppress OTU/category significance analysis. [default: False]</dd>
<dt>-t, <cite>-</cite>-tree_fp</dt>
<dd>Path to the tree file if one should be used. Required unless &#8211;nonphylogenetic_diversity is passed. [default: no tree will be used]</dd>
<dt>-c, <cite>-</cite>-categories</dt>
<dd>The metadata category or categories to compare (i.e., column headers in the mapping file) for categorical analyses. These should be passed  as a comma-separated list. [default: None; do not perform categorical analyses]</dd>
<dt>-w, <cite>-</cite>-print_only</dt>
<dd>Print the commands but don&#8217;t call them &#8211; useful for debugging or recovering from failed runs. [default: False]</dd>
<dt><cite>-</cite>-recover_from_failure</dt>
<dd>Don&#8217;t fail if output directory exists, but attempt to recover from the failed run. [default: False]</dd>
<dt>-O, <cite>-</cite>-jobs_to_start</dt>
<dd>Number of jobs to start. NOTE: you must also pass -a to run in parallel, this defines the number of jobs to be started if and only if -a is passed [default: 1]</dd>
</dl>
</div>
<p><strong>Output:</strong></p>
<p>Run diversity analyses at 20 sequences/sample, with categorical analyses focusing on the SampleType and day categories. ALWAYS SPECIFY ABSOLUTE FILE PATHS (absolute path represented here as $PWD, but will generally look something like /home/ubuntu/my_analysis/).</p>
<div class="highlight-python"><div class="highlight"><pre>core_diversity_analyses.py -i $PWD/otu_table.biom -o $PWD/core_output -m $PWD/map.txt -c SampleType,day -t $PWD/rep_set.tre -e 20
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    

    <h3><a href="../index.html">Site index</a></h3>
    <ul><li><ul>
    <li><a href="http://www.qiime.org">Home</a><br /></li>
    <li><a href="../install/index.html">Install</a><br /></li>
    <li><a href="../tutorials/index.html">Tutorials</a><br /></li>
    <li><a href="index.html">Scripts</a><br /></li>
    <li><a href="http://help.qiime.org">Help</a><br /></li>
    <li><a href="http://qiime.org/home_static/dataFiles.html">Resources</a><br /></li>
    <li><a href="../documentation/index.html">File Formats</a><br /></li>
    <li><a href="http://workshops.qiime.org">Workshops</a><br /></li>
    <li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
    <li><a href="../developer/index.html">Developer</a><br /></li>
    </ul></li></ul>


  <h4>Previous topic</h4>
  <p class="topless"><a href="convert_fastaqual_fastq.html"
                        title="previous chapter"><em>convert_fastaqual_fastq.py</em> &#8211; From a FASTA file and a matching QUAL file, generates a FASTQ file. From FASTQ file generates FASTA file and  matching QUAL file.</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="count_seqs.html"
                        title="next chapter"><em>count_seqs.py</em> &#8211;</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="count_seqs.html" title="count_seqs.py –"
             >next</a> |</li>
        <li class="right" >
          <a href="convert_fastaqual_fastq.html" title="convert_fastaqual_fastq.py – From a FASTA file and a matching QUAL file, generates a FASTQ file. From FASTQ file generates FASTA file and matching QUAL file."
             >previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="../documentation/index.html" >QIIME File Formats</a> &raquo;</li>
          <li><a href="index.html" >QIIME Scripts</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.scikit-bio.org" target="_blank"><img src="../_static/powered_by_skbio.png" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2010-2014, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>