<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Data Analysis Steps &mdash; Homepage</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.9',
        COLLAPSE_MODINDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="shortcut icon" href="_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="index.html" />
    <link rel="up" title="QIIME Tutorial" href="tutorial.html" />
    <link rel="next" title="QIIME Documentation" href="documentation.html" />
    <link rel="prev" title="Essential Files" href="tut_essential_files.html" /> 
  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="index.html"><img src="_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;">
	<p style="font-size:12pt;padding: 5px;"><b>News and Announcements &raquo;</b>
<script type="text/javascript" src="http://app.feed.informer.com/digest3/U9OLMKHF9N.js"><noscript><a href="http://app.feed.informer.com/digest3/U9OLMKHF9N.html">Click for &quot;QIIME&quot;.</a> Powered by <a href="http://feed.informer.com/">RSS Feed Informer</a></noscript></script></p>
</div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="documentation.html" title="QIIME Documentation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tut_essential_files.html" title="Essential Files"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">Home</a> &raquo;</li>
          <li><a href="tutorial.html" accesskey="U">QIIME Tutorial</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <ul class="simple">
</ul>
<div class="section" id="id1">
<h1>Data Analysis Steps<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h1>
<p>In this walkthrough, invocation of scripts will be noted with indentation and the sign denoting the start of a command line. You can find full usage information for each script by passing the -h option (help) and/or by reading the full description in the <a class="reference external" href="./documentation.html">Documentation</a>. First, assemble the sequences (.fna), quality scores (.qual), and metadata mapping file into a directory. Execute all tutorial commands from within this directory. To access QIIME python scripts, it may be useful to set an environment variable to the location of the innermost QIIME directory (the one containing <a class="reference external" href="./scripts/check_id_map.html">check_id_map.py</a>, for example)</p>
<div class="highlight-python"><pre>qdir=/path/to/QIIME/scripts/</pre>
</div>
<p>then further commands of the form <tt class="docutils literal"><span class="pre">python</span> <span class="pre">QIIMEscript.py</span> <span class="pre">-o</span> <span class="pre">option</span></tt> can be invoked as <tt class="docutils literal"><span class="pre">python</span> <span class="pre">$qdir/QIIMEscript.py</span> <span class="pre">-o</span> <span class="pre">option</span></tt>.</p>
<div class="section" id="pre-processing-454-data">
<span id="preprocessing454"></span><h2>Pre-processing 454 Data<a class="headerlink" href="#pre-processing-454-data" title="Permalink to this headline">¶</a></h2>
<p>Filter the reads based on quality, and assign multiplexed reads to starting sample by nucleotide barcode.</p>
<div class="section" id="check-mapping-file">
<span id="checkmapping"></span><h3>Check Mapping File<a class="headerlink" href="#check-mapping-file" title="Permalink to this headline">¶</a></h3>
<p>Before beginning the pipeline, you should ensure that your mapping file is formatted correctly with the <a class="reference external" href="./scripts/check_id_map.html">check_id_map.py</a> script.</p>
<div class="highlight-python"><pre>python $qdir/check_id_map.py -m Fasting_Map.txt -o mapping_output/</pre>
</div>
<p>If verbose (-v) is enabled, this utility will print to STDOUT a message indicating whether or not problems were found in the mapping file. Errors and warnings will the output to a log file, which will be present in the specified (-o) output directory. Errors will cause fatal problems with subsequent scripts and must be corrected before moving forward. Warnings will not cause fatal problems, but it is encouraged that you fix these problems as they are often indicative of typos in your mapping file, invalid characters, or other unintended errors that will impact downstream analysis. A <tt class="docutils literal"><span class="pre">corrected_mapping.txt</span></tt> file will also be created in the output directory, which will have a copy of the mapping file with invalid characters replaced by underscores, or a message indicating that no invalid characters were found.</p>
</div>
<div class="section" id="assign-samples-to-mulitplex-reads">
<span id="assignsamples"></span><h3>Assign Samples to Mulitplex Reads<a class="headerlink" href="#assign-samples-to-mulitplex-reads" title="Permalink to this headline">¶</a></h3>
<p>The next task is to assign the mulitplex reads to samples based on their nucleotide barcode. Also, this step performs quality filtering based on the characteristics of each sequence, removing any low quality or ambiguous reads. The script for this step is <a class="reference external" href="./scripts/split_libraries.html">split_libraries.py</a>. A full description of parameters for this script are described in the <a class="reference external" href="./documentation.html">Documentation</a>. For this tutorial, we will use default parameters (minimum quality score = 25, minimum/maximum length = 200/1000, no ambiguous bases allowed and no mismatches allowed in the primer sequence).:</p>
<div class="highlight-python"><pre>python $qdir/split_libraries.py -m Fasting_Map.txt -f Fasting_Example.fna -q Fasting_Example.qual -o Fasting_Split_Lib_Out</pre>
</div>
<p>This invocation will create three files in the new directory &#8220;Fasting_Split_Lib_Out/&#8221;:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">split_library_log.txt</span></tt> : This file contains the summary of splitting, including the number of reads detected for each sample and a brief summary of any reads that were removed due to quality considerations.</li>
<li><tt class="docutils literal"><span class="pre">histograms.txt</span></tt> : This tab delimited file shows the number of reads at regular size intervals before and after splitting the library.</li>
<li><tt class="docutils literal"><span class="pre">seqs.fna</span></tt> : This is a fasta formatted file where each sequence is renamed according to the sample it came from. The header line also contains the name of the read in the input fasta file and information on any barcode errors that were corrected.</li>
</ul>
<p>A few lines from the <tt class="docutils literal"><span class="pre">seqs.fna</span></tt> file are shown below:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>&gt;PC.634_1 FLP3FBN01ELBSX orig_bc=ACAGAGTCGGCT new_bc=ACAGAGTCGGCT bc_diffs=0</li>
<li>CTGGGCCGTGTCTCAGTCCCAATGTGGCCGTTTACCCTCTCAGGCCGGCTACGCATCATCGCC....</li>
<li>&gt;PC.634_2 FLP3FBN01EG8AX orig_bc=ACAGAGTCGGCT new_bc=ACAGAGTCGGCT bc_diffs=0</li>
<li>TTGGACCGTGTCTCAGTTCCAATGTGGGGGCCTTCCTCTCAGAACCCCTATCCATCGAAGGCTT....</li>
<li>&gt;PC.354_3 FLP3FBN01EEWKD orig_bc=AGCACGAGCCTA new_bc=AGCACGAGCCTA bc_diffs=0</li>
<li>TTGGGCCGTGTCTCAGTCCCAATGTGGCCGATCAGTCTCTTAACTCGGCTATGCATCATTGCCTT....</li>
<li>&gt;PC.481_4 FLP3FBN01DEHK3 orig_bc=ACCAGCGACTAG new_bc=ACCAGCGACTAG bc_diffs=0</li>
<li>CTGGGCCGTGTCTCAGTCCCAATGTGGCCGTTCAACCTCTCAGTCCGGCTACTGATCGTCGACT....</li>
</ul>
</div>
</div>
</div>
<div class="section" id="pick-operational-taxonomic-units-otus-and-representative-sequences">
<span id="pickotusandrepseqs"></span><h2>Pick Operational Taxonomic Units (OTUs) and Representative Sequences<a class="headerlink" href="#pick-operational-taxonomic-units-otus-and-representative-sequences" title="Permalink to this headline">¶</a></h2>
<div class="section" id="pick-otus-based-on-sequence-similarity-within-the-reads">
<span id="pickotusseqsim"></span><h3>Pick OTUs based on Sequence Similarity within the Reads<a class="headerlink" href="#pick-otus-based-on-sequence-similarity-within-the-reads" title="Permalink to this headline">¶</a></h3>
<p>At this step, all of the sequences from all of the samples will be clustered into Operational Taxonomic Units (OTUs) based on their sequence similarity. OTUs in QIIME are clusters of sequences, frequently intended to represent some degree of taxonomic relatedness. For example, when sequences are clustered at 97% sequence similarity with cd-hit, each resulting cluster is typically thought of as representing a genus. This model and the current techniques for picking OTUs are known to be flawed, and determining exactly how OTUs should be defined, and what they represent, is an active area of research. Thus, OTU-picking will identify highly similar sequences across the samples and provide a platform for comparisons of community structure. The script <a class="reference external" href="./scripts/pick_otus.html">pick_otus.py</a> takes as input the fasta file output from <a class="reference internal" href="#assignsamples"><em>Assign Samples to Mulitplex Reads</em></a> above, and returns a list of OTUs detected and the fasta header for sequences that belong in that OTU. To invoke the script using cd-hit to cluster and the default setting of 97% similarity determining an OTU, use</p>
<div class="highlight-python"><pre>python $qdir/pick_otus.py -i Fasting_Split_Lib_Out/seqs.fna -m cdhit -o Fasting_picked_otus</pre>
</div>
<p>In the newly created directory &#8220;Fasting_picked_otus/&#8221;, there will be two files. One is <tt class="docutils literal"><span class="pre">seqs.log</span></tt>, which contains information about the invocation of the script. The OTUs will be recorded in the tab-delimited file <tt class="docutils literal"><span class="pre">seqs_otus.txt</span></tt>. The OTUs are arbitrarily named by a number, which is recorded in the first column. The subsequent columns in each line identify the sequence or sequences that belong in that OTU.</p>
</div>
<div class="section" id="pick-representative-sequences-for-each-otu">
<span id="pickrepseqsforotu"></span><h3>Pick Representative Sequences for each OTU<a class="headerlink" href="#pick-representative-sequences-for-each-otu" title="Permalink to this headline">¶</a></h3>
<p>Since each OTU may be made up of many sequences, we will pick a representative sequence for that OTU for downstream analysis. This representative sequence will be used for taxonomic identification of the OTU and phylogenetic alignment. The script <a class="reference external" href="./scripts/pick_rep_set.html">pick_rep_set.py</a> uses the otu file created above and extracts a representative sequence from the fasta file by one of several methods. To use the default method, where the most abundant sequence in the OTU is used as the representative sequence, invoke the script as follows</p>
<div class="highlight-python"><pre>python $qdir/pick_rep_set.py -i Fasting_picked_otus/seqs_otus.txt -f Fasting_Split_Lib_Out/seqs.fna -l Fasting_picked_otus/rep_set.log -o rep_set.fa</pre>
</div>
<p>In the current working directory, the script has created two new files - the log file <tt class="docutils literal"><span class="pre">rep_set.log</span></tt> and the fasta file <tt class="docutils literal"><span class="pre">rep_set.fa</span></tt> containing one representative sequence for each OTU. In this fasta file, the sequence has been renamed by the OTU, and the additional information on the header line reflects the sequence used as the representative:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>&gt;0 PC.636_424</li>
<li>CTGGGCCGTATCTCAGTCCCAATGTGGCCGGTCGACCTCTC....</li>
<li>&gt;1 PC.481_321</li>
<li>TTGGGCCGTGTCTCAGTCCCAATGTGGCCGTCCGCCCTCTC....</li>
</ul>
</div>
</div>
</div>
<div class="section" id="assign-taxonomy-and-perform-otu-table-analyses">
<span id="assigntaxotuanalysis"></span><h2>Assign Taxonomy and Perform OTU Table Analyses<a class="headerlink" href="#assign-taxonomy-and-perform-otu-table-analyses" title="Permalink to this headline">¶</a></h2>
<p>In this section, we will assign the representative sequences for each OTU to a taxonomic identity using reference databases, and make an OTU table. Once the OTU table is generated, we show some ways to visualize the table.</p>
<div class="section" id="assign-taxonomy">
<span id="assigntax"></span><h3>Assign Taxonomy<a class="headerlink" href="#assign-taxonomy" title="Permalink to this headline">¶</a></h3>
<p>A primary goal of the QIIME pipeline is to assign high-throughput sequencing reads to taxonomic identities using established databases. This will give you information on the microbial lineages found in your samples. Using <a class="reference external" href="./scripts/assign_taxonomy.html">assign_taxonomy.py</a>, you can compare your OTUs against a reference database of your choosing. For our example, we will use the RDP classification system.:</p>
<div class="highlight-python"><pre>python $qdir/assign_taxonomy.py -i rep_set.fa -m rdp -o Fasting_rdp_taxonomy</pre>
</div>
<p>In the directory &#8220;Fasting_rdp_taxonomy/&#8221;, there will be a log file and a text file. The text file contains a line for each OTU considered, with the RDP taxonomy assignment and a numerical confidence of that assignment (1 is the highest possible confidence). For some OTUs, the assignment will be as specific as a bacterial species, while others may be assignable to nothing more specific than the bacterial domain. Below are the first few lines of the text file and the user should note that the taxonomic assignment and confidence numbers from their run may not coincide with the output shown below, due to the RDP classification algorithm:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>41 PC.356_347      Root;Bacteria                                                                   0.980</li>
<li>63 PC.635_130      Root;Bacteria;Firmicutes;&#8221;Clostridia&#8221;;Clostridiales;&#8221;Lachnospiraceae&#8221;           0.960</li>
<li>353        PC.634_150      Root;Bacteria;Proteobacteria;Deltaproteobacteria                                0.880</li>
<li>18 PC.355_1011     Root;Bacteria;Bacteroidetes;Bacteroidetes;Bacteroidales;Rikenellaceae;Alistipes 0.990</li>
</ul>
</div>
</div>
<div class="section" id="make-otu-table">
<span id="makeotutable"></span><h3>Make OTU Table<a class="headerlink" href="#make-otu-table" title="Permalink to this headline">¶</a></h3>
<p>Using these assignments and the OTU file created in <a class="reference internal" href="#pickotusseqsim"><em>Pick OTUs based on Sequence Similarity within the Reads</em></a>, we can make a readable matrix of OTU by Sample with meaningful taxonomic identifiers for each OTU. The script is <a class="reference external" href="./scripts/make_otu_table.html">make_otu_table.py</a>.:</p>
<div class="highlight-python"><pre>python $qdir/make_otu_table.py -i Fasting_picked_otus/seqs_otus.txt -t Fasting_rdp_taxonomy/rep_set_tax_assignments.txt -o Fasting_otu_table.txt</pre>
</div>
<p>The first few lines of <tt class="docutils literal"><span class="pre">Fasting_otu_table.txt</span></tt> are shown below (OTUs 2-9), where the first column contains the OTU number, the last column contains the taxonomic assignment for the OTU, and 9 columns between are for each of our 9 samples. The value of each <em>ij</em> entry in the matrix is the number of times OTU <em>i</em> was found in the sequences for sample <em>j</em>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>#Full OTU Counts</li>
<li>#OTUID     PC.354  PC.355  PC.356  PC.481  PC.593  PC.607  PC.634  PC.635  PC.636  Consensus       Lineage</li>
<li>2  0       0       0       0       0       1       0       3       1       Root;Bacteria;Bacteroidetes</li>
<li>3  0       0       0       0       0       0       0       0       1       Root;Bacteria</li>
<li>4  0       0       0       0       2       1       1       2       1       Root;Bacteria</li>
<li>5  0       0       2       0       0       0       0       1       0       Root</li>
<li>6  0       0       0       0       0       0       0       1       0       Root;Bacteria</li>
<li>7  0       1       3       0       9       1       1       1       3       Root;Bacteria;Bacteroidetes</li>
<li>8  0       0       0       0       0       1       0       0       0       Root;Bacteria;Bacteroidetes</li>
<li>9  0       0       0       0       0       0       0       0       1       Root;Bacteria;Bacteroidetes</li>
</ul>
</div>
</div>
<div class="section" id="rarify-otu-table-to-remove-sample-heterogeneity-optional">
<span id="rareotutableremovehetero"></span><h3>Rarify OTU Table to Remove Sample Heterogeneity (Optional)<a class="headerlink" href="#rarify-otu-table-to-remove-sample-heterogeneity-optional" title="Permalink to this headline">¶</a></h3>
<p>To remove sample heterogeneity, we can perform rarefaction on our OTU table. Rarefaction is an ecological approach that allows users to standardize the data obtained from samples with different sequencing efforts, and to compare the OTU richness of the samples using this standardized platform. For instance, if one of your samples yielded 10,000 sequence counts, and another yielded only 1,000 counts, the species diversity within those samples may be much more influenced by sequencing effort than underlying biology. The approach of rarefaction is to randomly sample the same number of OTUs from each sample, and use this data to compare the communities at a given level of sampling effort.</p>
<p>To perform rarefaction, you need to set the boundaries for sampling and the step size between sampling intervals. You can find the number of sequences associated with each sample by looking at the last line of the <tt class="docutils literal"><span class="pre">split_library_log.txt</span></tt> file generated in <a class="reference internal" href="#assignsamples"><em>Assign Samples to Mulitplex Reads</em></a> above. The line from our tutorial is pasted here:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>Sample ct min/max/mean: 146 / 150 / 148.11</li>
</ul>
</div>
<p>Since we are only removing sample heterogeneity from the OTU table, we will use <a class="reference external" href="./scripts/single_rarefaction.html">single_rarefaction.py</a>, which only requires the depth of sampling. Rarefaction is most useful when most samples have the specified number of sequences, so your upper bound of rarefaction should be close to the minimum number of sequences found in a sample. For this case, we will set the depth to 146.</p>
<div class="highlight-python"><pre>python $qdir/single_rarefaction.py -i Fasting_otu_table.txt -d 146 -o Fasting_rarified_otu_table.txt</pre>
</div>
<p>As a result, a newly created file <tt class="docutils literal"><span class="pre">Fasting_rarified_otu_table.txt</span></tt> has been written. For all subsequent analyses, one can use this text file in place of <tt class="docutils literal"><span class="pre">Fasting_otu_table.txt</span></tt>, since this file is also an OTU table, except that it has been rarified.</p>
</div>
<div class="section" id="make-otu-heatmap">
<span id="makeheatmap"></span><h3>Make OTU Heatmap<a class="headerlink" href="#make-otu-heatmap" title="Permalink to this headline">¶</a></h3>
<p>The QIIME pipeline includes a very useful utility to generate images of the OTU table. The script is <a class="reference external" href="./scripts/make_otu_heatmap_html.html">make_otu_heatmap_html.py</a></p>
<div class="highlight-python"><pre>python $qdir/make_otu_heatmap_html.py -i Fasting_otu_table.txt -o Fasting_OTU_Heatmap/</pre>
</div>
<p>An html file is created in the directory &#8220;Fasting_OTU_Heatmap/&#8221;. You can open this file with any web browser, and will be prompted to enter a value for &#8220;Filter by Counts per OTU&#8221;. Only OTUs with total counts at or above this threshold will be displayed. The OTU heatmap displays raw OTU counts per sample, where the counts are colored based on the contribution of each OTU to the total OTU count present in that sample (blue: contributes low percentage of OTUs to sample; red: contributes high percentage of OTUs). Click the &#8220;Sample ID&#8221; button, and a graphic will be generated like the figure below. For each sample, you will see in a heatmap the number of times each OTU was found in that sample. You can mouse over any individual count to get more information on the OTU (including taxonomic assignment). Within the mouseover, there is a link for the terminal lineage assignment, so you can easily search Google for more information about that assignment.</p>
<div align="center" class="align-center"><img alt="_images/heatmap.png" class="align-center" src="_images/heatmap.png" /></div>
<p>Alternatively, you can click on one of the counts in the heatmap and a new pop-up window will appear. The pop-up window uses a Google Visualization API called Magic-Table. Depending on which table count you clicked on, the pop-up window will put the clicked-on count in the middle of the pop-up heatmap as shown below. For the following example, the table count with the red arrow mouseover is the same one being focused on using the Magic-Table.</p>
<div align="center" class="align-center"><img alt="_images/fisheyeheatmap.png" class="align-center" src="_images/fisheyeheatmap.png" /></div>
<p>On the original heatmap webpage, if you select the &#8220;Taxonomy&#8221; button instead, you will generate a heatmap keyed by taxon assignment, which allows you to conveniently look for organisms and lineages of interest in your study. Again, mousing over an individual count will show additional information for that OTU and sample.</p>
<div align="center" class="align-center"><img alt="_images/taxheatmap.png" class="align-center" src="_images/taxheatmap.png" /></div>
</div>
<div class="section" id="make-otu-network">
<span id="makeotunetwork"></span><h3>Make OTU Network<a class="headerlink" href="#make-otu-network" title="Permalink to this headline">¶</a></h3>
<p>An alternative to viewing the OTU table as a heatmap, is to create an OTU network, using the following command.:</p>
<div class="highlight-python"><pre>python $qdir/make_otu_network.py -m Fasting_Map.txt -i Fasting_otu_table.txt -o OTU_Network</pre>
</div>
<p>To visualize the network, we use the <a class="reference external" href="http://www.cytoscape.org/">Cytoscape</a> program (which you can run by calling Cytoscape from the command line, beginning with a capital C, if you have Cytoscape installed), where each red circle represents a sample and each white square represents an otu. The lines represent the OTUs present in a particular sample (blue for controls and green for fasting). For more information about opening the files in <a class="reference external" href="http://www.cytoscape.org/">Cytoscape</a> please refer <a class="reference external" href="./scripts/cytoscape_usage.html">here</a>.</p>
<div align="center" class="align-center"><img alt="_images/network.png" class="align-center" src="_images/network.png" /></div>
<p>You can group OTUs by different taxonomic levels (division, class, family) with the script <a class="reference external" href="./scripts/summarize_taxa.html">summarize_taxa.py</a>. The input is the OTU table created above and the taxonomic level you need to group the OTUs. For the RDP taxonomy, the following taxonomic levels correspond to: 2 = Domain (Bacteria), 3 = Phylum (Actinobacteria), 4 = Class, and so on.</p>
<div class="highlight-python"><pre>python $qdir/summarize_taxa.py -i Fasting_otu_table.txt -o Fasting_otu_table_Level3.txt -L 3 -r 0</pre>
</div>
<p>The script will generate a new OTU table <tt class="docutils literal"><span class="pre">Fasting_otu_table_Level3.txt</span></tt>, where the value of each <em>ij</em> entry in the matrix is the count of the number of times all OTUs belonging to the taxon <em>i</em> (for example, Phylum Actinobacteria) were found in the sequences for sample <em>j</em>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>#Full OTU Counts</li>
<li>Taxon                              PC.354 PC.355   PC.356  PC.481  PC.593  PC.607  PC.634  PC.635  PC.636</li>
<li>Root;Bacteria;Actinobacteria       0.0     0.0     0.0     1.0     0.0     2.0     3.0     1.0     1.0</li>
<li>Root;Bacteria;Bacteroidetes        7.0     38.0    15.0    19.0    30.0    40.0    86.0    54.0    90.0</li>
<li>Root;Bacteria;Deferribacteres      0.0     0.0     0.0     0.0     0.0     3.0     5.0     2.0     7.0</li>
<li>Root;Bacteria;Firmicutes   136.0   102.0   115.0   117.0   65.0    66.0    37.0    63.0    34.0</li>
<li>Root;Bacteria;Other                5.0     6.0     18.0    9.0     49.0    35.0    14.0    27.0    14.0</li>
<li>Root;Bacteria;Proteobacteria       0.0     0.0     0.0     0.0     5.0     3.0     2.0     0.0     1.0</li>
<li>Root;Bacteria;TM7          0.0     0.0     0.0     0.0     0.0     0.0     2.0     0.0     0.0</li>
<li>Root;Bacteria;Verrucomicrobia      0.0     0.0     0.0     0.0     0.0     0.0     1.0     0.0     0.0</li>
<li>Root;Other                 0.0     0.0     2.0     0.0     0.0     0.0     0.0     1.0     0.0</li>
</ul>
</div>
</div>
<div class="section" id="make-pie-charts">
<span id="makepiecharts"></span><h3>Make Pie Charts<a class="headerlink" href="#make-pie-charts" title="Permalink to this headline">¶</a></h3>
<p>To visualize the summarized taxa, you can use the <a class="reference external" href="./scripts/make_pie_charts.html">make_pie_charts.py</a> script, which shows which taxons are present in all samples or within each sample (-s).</p>
<div class="highlight-python"><pre>python $qdir/make_pie_charts.py -i Fasting_otu_table_Level3.txt -l Phylum -o Fasting_Pie_Charts -s</pre>
</div>
<p>To view the resulting pie charts, open the html file located in the &#8220;Fasting_Pie_Charts/&#8221; folder. The following pie chart shows the taxa assignments for all samples.</p>
<div align="center" class="align-center"><img alt="_images/piechart1.png" class="align-center" src="_images/piechart1.png" /></div>
<p>The following pie chart shows the taxa assignments for one of the samples (PC.354).</p>
<div align="center" class="align-center"><img alt="_images/piechart2.png" class="align-center" src="_images/piechart2.png" /></div>
</div>
</div>
<div class="section" id="align-otu-sequences-filter-the-alignment-and-generate-a-phylogenetic-tree">
<span id="alignotuseqsfilteralngentree"></span><h2>Align OTU Sequences, Filter the Alignment and Generate a Phylogenetic Tree<a class="headerlink" href="#align-otu-sequences-filter-the-alignment-and-generate-a-phylogenetic-tree" title="Permalink to this headline">¶</a></h2>
<div class="section" id="align-otu-sequences">
<span id="alignotuseq"></span><h3>Align OTU Sequences<a class="headerlink" href="#align-otu-sequences" title="Permalink to this headline">¶</a></h3>
<p>Alignment of the sequences and phylogeny inference is necessary only if phylogenetic tools such as <a class="reference external" href="http://bmf2.colorado.edu/unifrac/index.psp">UniFrac</a> will be subsequently invoked. Alignments can either be generated de novo using programs such as MUSCLE, or through assignment to an existing alignment with tools like <a class="reference external" href="http://pynast.sourceforge.net/">PyNAST</a>. For small studies such as this tutorial, either method is possible. However, for studies involving many sequences (roughly, more than 1000), the de novo aligners are very slow and assignment with <a class="reference external" href="http://pynast.sourceforge.net/">PyNAST</a> is preferred. Either alignment approach is accomplished with the script <a class="reference external" href="./scripts/align_seqs.html">align_seqs.py</a>. Since this is one of the most computationally intensive bottlenecks in the pipeline, large studies would benefit greatly from parallelization of this task (described in detail in the <a class="reference external" href="./documentation.html">Documentation</a>)</p>
<div class="highlight-python"><pre>python $qdir/align_seqs.py -i rep_set.fa -o Default_Aln -t /path/to/Green_genes_core_set/core_set_aligned.fasta.imputed -e 150</pre>
</div>
<p>A log file and an alignment file are created in the directory &#8220;Default_Aln/&#8221;.</p>
</div>
<div class="section" id="filter-alignment">
<span id="filteraln"></span><h3>Filter Alignment<a class="headerlink" href="#filter-alignment" title="Permalink to this headline">¶</a></h3>
<p>Before building the tree, one must filter the alignment to removed columns comprised of only
gaps. Note that depending on where you obtained the lanemask file from, it will either be named lanemask_in_1s_and_0s.txt or lanemask_in_1s_and_0s</p>
<div class="highlight-python"><pre>python $qdir/filter_alignment.py -m /path/to/Green_genes_core_set/lanemask_in_1s_and_0s.txt -i Default_Aln/rep_set_aligned.fa -o Filtered_Aln</pre>
</div>
</div>
<div class="section" id="make-phylogenetic-tree">
<span id="maketree"></span><h3>Make Phylogenetic Tree<a class="headerlink" href="#make-phylogenetic-tree" title="Permalink to this headline">¶</a></h3>
<p>The filtered alignment file produced in the directory &#8220;Filtered_Aln/&#8221; can be used to build a phylogenetic tree using a tree-building program. Here is an example of how to create a tree with the program FastTree and the script <a class="reference external" href="./scripts/make_phylogeny.html">make_phylogeny.py</a>:</p>
<div class="highlight-python"><pre>python $qdir/make_phylogeny.py -t fasttree -o rep_set_aligned.tre -l Filtered_Aln/rep_set_tree.log -i Filtered_Aln/rep_set_aligned_pfiltered.fasta</pre>
</div>
<p>The Newick format tree file is written to <tt class="docutils literal"><span class="pre">rep_set_aligned.tre</span></tt>. This file can be used in tree visualization software, and is necessary for <a class="reference external" href="http://bmf2.colorado.edu/unifrac/index.psp">UniFrac</a> diversity measurements (described below). For the following example the FigTree program was used to visualize the phylogenetic tree obtained from <tt class="docutils literal"><span class="pre">rep_set_aligned.tre</span></tt>.</p>
<div align="center" class="align-center"><img alt="_images/tree.png" class="align-center" src="_images/tree.png" /></div>
</div>
</div>
<div class="section" id="compute-alpha-diversity-within-the-samples-and-generate-rarefaction-curves">
<span id="compalphadivrarecurves"></span><h2>Compute Alpha Diversity within the Samples, and Generate Rarefaction Curves<a class="headerlink" href="#compute-alpha-diversity-within-the-samples-and-generate-rarefaction-curves" title="Permalink to this headline">¶</a></h2>
<p>Community ecologists typically describe the microbial diversity within their study. This diversity can be assessed within a sample (alpha diversity) or between a collection of samples (beta diversity). Here, we will determine the level of alpha diversity in our samples using a series of scripts from the QIIME pipeline.</p>
<div class="section" id="rarify-otu-table">
<span id="rareotutable"></span><h3>Rarify OTU Table<a class="headerlink" href="#rarify-otu-table" title="Permalink to this headline">¶</a></h3>
<p>For this highly artificial example, all of the samples had sequence counts between 146 and 150, as discussed in <a class="reference internal" href="#rareotutableremovehetero"><em>Rarify OTU Table to Remove Sample Heterogeneity (Optional)</em></a>. In real datasets, the range will generally be much larger. In practice, rarefaction is most useful when most samples have the specified number of sequences, so your upper bound of rarefaction should be close to the minimum number of sequences found in a sample. This is a judgment call, and the rarefaction can be repeated until satisfactory information is attained. We will use an upper bound (rarefaction maximum, option -x) of 141 sequences, a lower bound (minimum, option -m) of 1, and a step size of 10 (option -s). We want enough steps to create smooth curves, but not so many as to create a computational bottleneck. Finally, we will choose to do 5 iterations (option -n); higher numbers of iterations will give you greater confidence that the average result of the random sampling is representative, but will also increase the compute time.</p>
<div class="highlight-python"><pre>python $qdir/multiple_rarefactions.py -i Fasting_otu_table.txt -m 1 -x 141 -s 10 -n 5 -o Fasting_Rarefaction/</pre>
</div>
<p>The newly created directory &#8220;Fasting_Rarefaction/&#8221; will contain many text files named <tt class="docutils literal"><span class="pre">rarefaction_##_#.txt</span></tt>; the first set of numbers represents the number of sequences sampled, and the last number represents the iteration number. If you opened one of these files, you would find an otu table where for each sample the sum of the counts equals the number of samples taken.</p>
</div>
<div class="section" id="compute-alpha-diversity">
<span id="computealphadiv"></span><h3>Compute Alpha Diversity<a class="headerlink" href="#compute-alpha-diversity" title="Permalink to this headline">¶</a></h3>
<p>The rarefaction tables are the basis for calculating diversity metrics, which reflect the diversity within the sample based on taxon counts of phylogeny. The QIIME pipeline allows users to conveniently calculate more than two dozen different diversity metrics. The full list of available metrics is available by passing the option -s to the script <a class="reference external" href="./scripts/alpha_diversity.html">alpha_diversity.py</a>. Every metric has different strengths and limitations - technical discussion of each metric is readily available online and in ecology textbooks, but it is beyond the scope of this document. Here, we will calculate three metrics:</p>
<ol class="arabic simple">
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Simpson_index">The Simpson Index</a> is a very common measure of species abundance based on OTU counts</li>
<li>The Observed Species metric is simply the count of unique OTUs found in the sample</li>
<li>Phylogenetic Distance (PD_whole_tree) is the only phylogenetic metric used in this script and requires a phylogenetic tree as an input.</li>
</ol>
<p>Command:</p>
<div class="highlight-python"><pre>python $qdir/alpha_diversity.py -i Fasting_Rarefaction/ -o Fasting_Alpha_Metrics/ -m simpson,PD_whole_tree,observed_species -t rep_set_aligned.tre</pre>
</div>
</div>
<div class="section" id="collate-rarified-otu-tables">
<span id="collateotutable"></span><h3>Collate Rarified OTU Tables<a class="headerlink" href="#collate-rarified-otu-tables" title="Permalink to this headline">¶</a></h3>
<p>The output directory &#8220;Fasting_Alpha_Metric/&#8221; will contain one text file <tt class="docutils literal"><span class="pre">alpha_rarefaction_##_#</span></tt> for every file input from ″Fasting_Rarefaction/″, where the numbers represent the number of samples and iterations as before. The content of this tab delimited file is the calculated metrics for each sample. To collapse the individual files into a single combined table, use the script <a class="reference external" href="./scripts/collate_alpha.html">collate_alpha.py</a>:</p>
<div class="highlight-python"><pre>python $qdir/collate_alpha.py -i Fasting_Alpha_Metrics/ -o Fasting_Alpha_Collated/</pre>
</div>
<p>In the newly created directory &#8220;Fasting_Alpha_Collated/&#8221;, there will be one matrix for every diversity metric used in the <a class="reference external" href="./scripts/alpha_diversity.html">alpha_diversity.py</a> script. This matrix will contain the metric for every sample, arranged in ascending order from lowest number of sequences per sample to highest. A portion of the <tt class="docutils literal"><span class="pre">observed_species.txt</span></tt> file are shown below:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>Sequences per sample       iteration       PC.354  PC.355  PC.356  PC.481  PC.593</li>
<li>alpha_rarefaction_21_0.txt 21                      0               14.0    16.0    18.0    18.0    13.0</li>
<li>alpha_rarefaction_21_1.txt 21                      1               15.0    17.0    18.0    20.0    12.0</li>
<li>alpha_rarefaction_21_2.txt 21                      2               15.0    16.0    21.0    19.0    13.0</li>
<li>alpha_rarefaction_21_3.txt 21                      3               10.0    19.0    18.0    21.0    13.0</li>
<li>alpha_rarefaction_21_4.txt 21                      4               14.0    18.0    16.0    15.0    12.0</li>
<li>...</li>
</ul>
</div>
</div>
<div class="section" id="generate-rarefaction-curves">
<span id="generaterarecurves"></span><h3>Generate Rarefaction Curves<a class="headerlink" href="#generate-rarefaction-curves" title="Permalink to this headline">¶</a></h3>
<p>The script <a class="reference external" href="./scripts/make_rarefaction_plots.html">make_rarefaction_plots.py</a> takes a mapping file and any number of rarefaction files generated by <a class="reference external" href="./scripts/collate_alpha.html">collate_alpha.py</a> and uses matplotlib to create rarefaction curves. Each curve represents a sample and can be colored by the sample metadata supplied in the mapping file. For this study, we will use the PD_whole_tree results and color by the Treatment Category.</p>
<div class="highlight-python"><pre>python $qdir/make_rarefaction_plots.py -m Fasting_Map.txt -r Fasting_Alpha_Collated/PD_whole_tree.txt -o Rarefaction_Plots -p Treatment -i png</pre>
</div>
<p>To view the rarefaction plot navigate to the bottom-level file within the &#8220;rarefaction_graph##&#8221; folder, where you will find an html file titled <tt class="docutils literal"><span class="pre">rarefaction_plots.html</span></tt>. This plot shows the averages of each value for in the Treatment category with their corresponding error bars.</p>
<div align="center" class="align-center"><img alt="_images/rarecurve.png" class="align-center" src="_images/rarecurve.png" /></div>
</div>
</div>
<div class="section" id="compute-beta-diversity-and-generate-2d-3d-principal-coordinate-analysis-pcoa-plots">
<span id="compbetadivgenpcoa"></span><h2>Compute Beta Diversity and Generate 2D/3D Principal Coordinate Analysis (PCoA) Plots<a class="headerlink" href="#compute-beta-diversity-and-generate-2d-3d-principal-coordinate-analysis-pcoa-plots" title="Permalink to this headline">¶</a></h2>
<div class="section" id="compute-beta-diversity">
<span id="compbetadiv"></span><h3>Compute Beta Diversity<a class="headerlink" href="#compute-beta-diversity" title="Permalink to this headline">¶</a></h3>
<p>Beta-diversity metrics assess the differences between microbial communities. In general, these metrics are calculated to study diversity along an environmental gradient (pH or temperature) or different disease states (lean vs. obese). The basic output of this comparison is a square matrix where a &#8220;distance&#8221; is calculated between every pair of samples reflecting the similarity between the samples. The data in this distance matrix can be visualized with clustering analyses, namely Principal Coordinate Analysis (PCoA) and UPGMA clustering. Like alpha diversity, there are many possible metrics which can be calculated with the QIIME pipeline - the full list of options can be seen by passing option -s to the script <a class="reference external" href="./scripts/beta_diversity.html">beta_diversity.py</a>. For our example, we will calculate the unweighted unifrac metric, which is a phylogenetic measure used extensively in recent microbial community sequencing projects.</p>
<div class="highlight-python"><pre>python $qdir/beta_diversity.py -i Fasting_otu_table.txt -m unweighted_unifrac -o unifrac_dist -t rep_set_aligned.tre</pre>
</div>
<p>The resulting distance matrix (unifrac_dist/unweighted_unifrac_Fasting_otu_table.txt) is the basis for two methods of visualization and sample comparison: PCoA and UPGMA.</p>
</div>
<div class="section" id="generate-distance-histograms">
<span id="gendisthist"></span><h3>Generate Distance Histograms<a class="headerlink" href="#generate-distance-histograms" title="Permalink to this headline">¶</a></h3>
<p>Distance Histograms are a way to compare different categories and see which tend to have larger/smaller distances than others. For example, in the hand study, you may want to compare the distances between hands to the distances between individuals.</p>
<div class="highlight-python"><pre>python $qdir/make_distance_histograms.py -d unifrac_dist/unweighted_unifrac_Fasting_otu_table.txt -m Fasting_Map.txt --fields Treatment -o Distance_Histograms --html_output</pre>
</div>
<p>For each of these groups of distances a histogram is made. The output is a HTML file (<tt class="docutils literal"><span class="pre">QIIME_Distance_Histograms.html</span></tt>) where you can look at all the distance histograms individually, and compare them between each other. Within the webpage, the user can mouseover and/or select the checkboxes in the right panel to turn on/off the different distances within/between categories. For this example, we are comparing the distances between the samples in the Control versus themselves, along with samples from Fasting versus the Control.</p>
<div align="center" class="align-center"><img alt="_images/hist.png" class="align-center" src="_images/hist.png" /></div>
</div>
<div class="section" id="generate-principal-coordinates">
<span id="genpcoa"></span><h3>Generate Principal Coordinates<a class="headerlink" href="#generate-principal-coordinates" title="Permalink to this headline">¶</a></h3>
<p>Principal Coordinate Analysis (PCoA) is a technique that helps to extract and visualize a few highly informative gradients of variation from complex, multidimensional data. This is a complex transformation that maps the distance matrix to a new set of orthogonal axes such that a maximum amount of variation is explained by the first principal coordinate, the second largest amount of variation is explained by the second principal coordinate, etc. The principal coordinates can be plotted in two or three dimensions to provide an intuitive visualization of the data structure and look at differences between the samples, and look for similarities by sample category. The transformation is accomplished with the script <a class="reference external" href="./scripts/principal_coordinates.html">principal_coordinates.py</a>:</p>
<div class="highlight-python"><pre>python $qdir/principal_coordinates.py -i unifrac_dist/unweighted_unifrac_Fasting_otu_table.txt -o Fasting_Unw_Unifrac_coords.txt</pre>
</div>
<p>The file &#8220;Fasting_Unw_Unifrac_coords.txt&#8221; lists every sample in the first column, and the subsequent columns contain the value for the sample against the noted principal coordinate. At the bottom of each Principal Coordinate column, you will find the eigenvalue and percent of variation explained by the coordinate. To determine which axes are useful for your project, you can generate a &#8220;scree plot&#8221; by plotting the eigenvalues of each principal component in descending order.</p>
</div>
<div class="section" id="generate-2d-and-3d-pcoa-plots">
<span id="gen2d3dpcoa"></span><h3>Generate 2D and 3D PCoA Plots<a class="headerlink" href="#generate-2d-and-3d-pcoa-plots" title="Permalink to this headline">¶</a></h3>
<p>To plot the coordinates, you can use the QIIME scripts <a class="reference external" href="./scripts/make_2d_plots.html">make_2d_plots.py</a> and <a class="reference external" href="./scripts/make_3d_plots.html">make_3d_plots.py</a>. The two dimensional plot will be rendered as a html file which can be opened with a standard web browser, while the three dimensional plot will be a kinemage file which requires additional software to render and manipulate. The usage for both scripts use the same convention, detailed in the <strong>Documentation</strong>.</p>
<div class="highlight-python"><pre>python $qdir/make_2d_plots.py -i Fasting_Unw_Unifrac_coords.txt -m Fasting_Map.txt -o Fasting_2d/ -b Treatment

python $qdir/make_3d_plots.py -i Fasting_Unw_Unifrac_coords.txt -m Fasting_Map.txt -o Fasting_3d/ -b "DOB,Treatment"</pre>
</div>
<p>The html file created in directory &#8220;Fasting_2d/&#8221; shows a plot for each combination of the first three principal coordinates. Since we specified Treatment to use for coloring the samples, each sample colored according to the category it corresponds. You can get the name for each sample by holding your mouse over the data point.</p>
<div align="center" class="align-center"><img alt="_images/pcoa1.png" class="align-center" src="_images/pcoa1.png" /></div>
<p>In the invocation of <a class="reference external" href="./scripts/make_3d_plots.html">make_3d_plots.py</a>, we specified that the samples should be colored by the value of the &#8220;Treatment&#8221; and &#8220;DOB&#8221; columns in the mapping file. For the &#8220;Treatment&#8221; column, all samples with the same &#8220;Treatment&#8221; will get the same color. For our tutorial, the five control samples are all blue and the four control samples are all green. This lets you easily visualize &#8220;clustering&#8221; by metadata category. The 3d visualization software allows you to rotate the axes to see the data from different perspectives. By default, the script will plot the first three dimensions in your file. Other combinations can be viewed using the &#8220;Views:Choose viewing axes&#8221; option in the KiNG viewer (may require the installation of kinemage software). The first 10 components can be viewed using &#8220;Views:Paralleled coordinates&#8221; option or typing &#8220;/&#8221;.</p>
<div align="center" class="align-center"><img alt="_images/pcoa2.png" class="align-center" src="_images/pcoa2.png" /></div>
</div>
</div>
<div class="section" id="upgma-clustering-and-jackknifing-support">
<span id="hiarchclustjack"></span><h2>UPGMA Clustering and Jackknifing Support<a class="headerlink" href="#upgma-clustering-and-jackknifing-support" title="Permalink to this headline">¶</a></h2>
<div class="section" id="upgma-clustering">
<span id="hiarchclust"></span><h3>UPGMA Clustering<a class="headerlink" href="#upgma-clustering" title="Permalink to this headline">¶</a></h3>
<p>Unweighted Pair Group Method with Arithmetic mean (UPGMA) is type of UPGMA clustering method using average linkage and can be used to visualize the distance matrix produced by <a class="reference external" href="./scripts/beta_diversity.html">beta_diversity.py</a>. The output is a file that can be opened with tree viewing software, such as FigTree.</p>
<div class="highlight-python"><pre>python $qdir/upgma_cluster.py -i unifrac_dist/unweighted_unifrac_Fasting_otu_table.txt -o Fasting_Unw_Unifrac_upgma.tre</pre>
</div>
<div align="center" class="align-center"><img alt="_images/hiarchclust.png" class="align-center" src="_images/hiarchclust.png" /></div>
<p>This tree shows the relationship among the 9 samples, and reveals that the 4 samples from the guts of fasting mice cluster together (PC.6xx, fasting data is in <tt class="docutils literal"><span class="pre">Fasting_Map.txt</span></tt>).</p>
</div>
<div class="section" id="perform-jackknifing-support">
<span id="jacksupport"></span><h3>Perform Jackknifing Support<a class="headerlink" href="#perform-jackknifing-support" title="Permalink to this headline">¶</a></h3>
<p>To measure the robustness of this result to sequencing effort, we perform a jackknifing analysis, wherein a smaller number of sequences are chosen at random from each sample, and the resulting UPGMA tree from this subset of data is compared with the tree representing the entire available data set. This process is repeated with many random subsets of data, and the tree nodes which prove more consistent across jackknifed datasets are deemed more robust.</p>
<p>First the jackknifed otu tables must be generated, by subsampling the full available data set. In this tutorial, each sample contains between 146 and 150 sequences, as shown in the <tt class="docutils literal"><span class="pre">split_library_log.txt</span></tt> file:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ul class="last simple">
<li>Sample ct min/max/mean: 146 / 150 / 148.11</li>
</ul>
</div>
<p>To ensure that a random subset of sequences is selected from each sample, we chose to select 110 sequences from each sample (75% of the smallest sample, though this value is only a guideline). We use <cite>multiple_rarefactions.py &lt;./scripts/multiple_rarefactions.html</cite>, which has already been described:</p>
<div class="highlight-python"><pre>python $qdir/multiple_rarefactions.py -i Fasting_otu_table.txt -m 110 -x 110 -s 1 -n 20 -o jackknife_otu/ --small_included</pre>
</div>
<p>This generates 20 subsets of the available data by random sampling, simulating a smaller sampling effort (110 sequences in each sample). Because our goal is to compare these jackknifed trees to the tree already generated, we use &#8211;small_included to ensure that every sample present in the original tree is present in the jackknifed ones. (here &#8211;small_included isn&#8217;t actually necessary, as all samples have &gt; 110 sequences).</p>
<p>We then calculate the distance matrix for each jackknifed dataset, using <a class="reference external" href="./scripts/beta_diversity.html">beta_diversity.py</a> as before, but now in batch mode:</p>
<div class="highlight-python"><pre>python $qdir/beta_diversity.py -i jackknife_otu -m dist_unweighted_unifrac -t rep_set_aligned.tre -o jackknife_dist</pre>
</div>
<p>which results in 20 distance matrix files written to the &#8220;jackknife_dist/&#8221; directory. Each of those is then used as the basis for UPGMA clustering, using <a class="reference external" href="./scripts/upgma_cluster.html">upgma_cluster.py</a> in batch mode:</p>
<div class="highlight-python"><pre>python $qdir/upgma_cluster.py -i jackknife_dist -o jackknife_upgma</pre>
</div>
</div>
<div class="section" id="compare-jackknifed-trees-to-cluster-tree">
<span id="compjackclustertree"></span><h3>Compare Jackknifed Trees to Cluster Tree<a class="headerlink" href="#compare-jackknifed-trees-to-cluster-tree" title="Permalink to this headline">¶</a></h3>
<p>UPGMA clustering of the 20 distance matrix files results in 20 UPGMA samples clusters, each based on a random sub-sample of the available sequence data. These are then compared to the UPGMA result using all available data:</p>
<div class="highlight-python"><pre>python $qdir/tree_compare.py -m Fasting_Unw_Unifrac_upgma.tre -s jackknife_upgma -o Fasting_jackknife_support</pre>
</div>
<p>This compares the UPGMA clustering based on all available data (-m) with the jackknifed UPGMA results (-s).  Two files are written to &#8220;Fasting_jackknife_support/&#8221;:</p>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">master_tree.tre</span></tt> is virtually identical to <tt class="docutils literal"><span class="pre">Fasting_Unw_Unifrac_upgma.tre</span></tt>, but each internal node of the UPGMA clustering is assigned a unique name</li>
<li><tt class="docutils literal"><span class="pre">jackknife_support.txt</span></tt> explains how frequently a given internal node had the same set of descendant samples in the jackknifed UPGMA clusters as it does in the UPGMA cluster using the full available data.  A value of 0.5 indicates that half of the jackknifed data sets support that node, while 1.0 indicates perfect support.</li>
</ul>
</div>
<div class="section" id="generate-bootstrapped-tree">
<span id="genboottree"></span><h3>Generate Bootstrapped Tree<a class="headerlink" href="#generate-bootstrapped-tree" title="Permalink to this headline">¶</a></h3>
<p>This data can be visualized using <a class="reference external" href="./scripts/make_bootstrapped_tree.html">make_bootstrapped_tree.py</a>:</p>
<div class="highlight-python"><pre>python $qdir/make_bootstrapped_tree.py -m Fasting_jackknife_support/master_tree.tre -s Fasting_jackknife_support/jackknife_support.txt -o Fasting_jackknife_support/Unw_unifrac_support.pdf</pre>
</div>
<p>The resulting pdf shows the tree with internal nodes colored, red for 75-100% support, yellow for 50-75%, green for 25-50%, and blue for &lt; 25% support. Although UPGMA shows that PC.354 and PC.593 cluster together and PC.481 with PC.6xx cluster together, we can not have high confidence in that result. However, there is excellent jackknife support for all fasted samples (PC.6xx) which are clustering together, separate from the non-fasted (PC.35x) samples.</p>
<div align="center" class="align-center"><img alt="_images/boottree.png" class="align-center" src="_images/boottree.png" /></div>
<div class="section" id="references">
<h4>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h4>
<p>Crawford, P. A., Crowley, J. R., Sambandam, N., Muegge, B. D., Costello, E. K., Hamady, M., et al. (2009). Regulation of myocardial ketone body metabolism by the gut microbiota during nutrient deprivation. Proc Natl Acad Sci U S A, 106(27), 11276-11281.</p>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
            <h3><a href="index.html">Table Of Contents</a></h3>
            <ul>
<li><a class="reference external" href="">Data Analysis Steps</a><ul>
<li><a class="reference external" href="#pre-processing-454-data">Pre-processing 454 Data</a><ul>
<li><a class="reference external" href="#check-mapping-file">Check Mapping File</a></li>
<li><a class="reference external" href="#assign-samples-to-mulitplex-reads">Assign Samples to Mulitplex Reads</a></li>
</ul>
</li>
<li><a class="reference external" href="#pick-operational-taxonomic-units-otus-and-representative-sequences">Pick Operational Taxonomic Units (OTUs) and Representative Sequences</a><ul>
<li><a class="reference external" href="#pick-otus-based-on-sequence-similarity-within-the-reads">Pick OTUs based on Sequence Similarity within the Reads</a></li>
<li><a class="reference external" href="#pick-representative-sequences-for-each-otu">Pick Representative Sequences for each OTU</a></li>
</ul>
</li>
<li><a class="reference external" href="#assign-taxonomy-and-perform-otu-table-analyses">Assign Taxonomy and Perform OTU Table Analyses</a><ul>
<li><a class="reference external" href="#assign-taxonomy">Assign Taxonomy</a></li>
<li><a class="reference external" href="#make-otu-table">Make OTU Table</a></li>
<li><a class="reference external" href="#rarify-otu-table-to-remove-sample-heterogeneity-optional">Rarify OTU Table to Remove Sample Heterogeneity (Optional)</a></li>
<li><a class="reference external" href="#make-otu-heatmap">Make OTU Heatmap</a></li>
<li><a class="reference external" href="#make-otu-network">Make OTU Network</a></li>
<li><a class="reference external" href="#make-pie-charts">Make Pie Charts</a></li>
</ul>
</li>
<li><a class="reference external" href="#align-otu-sequences-filter-the-alignment-and-generate-a-phylogenetic-tree">Align OTU Sequences, Filter the Alignment and Generate a Phylogenetic Tree</a><ul>
<li><a class="reference external" href="#align-otu-sequences">Align OTU Sequences</a></li>
<li><a class="reference external" href="#filter-alignment">Filter Alignment</a></li>
<li><a class="reference external" href="#make-phylogenetic-tree">Make Phylogenetic Tree</a></li>
</ul>
</li>
<li><a class="reference external" href="#compute-alpha-diversity-within-the-samples-and-generate-rarefaction-curves">Compute Alpha Diversity within the Samples, and Generate Rarefaction Curves</a><ul>
<li><a class="reference external" href="#rarify-otu-table">Rarify OTU Table</a></li>
<li><a class="reference external" href="#compute-alpha-diversity">Compute Alpha Diversity</a></li>
<li><a class="reference external" href="#collate-rarified-otu-tables">Collate Rarified OTU Tables</a></li>
<li><a class="reference external" href="#generate-rarefaction-curves">Generate Rarefaction Curves</a></li>
</ul>
</li>
<li><a class="reference external" href="#compute-beta-diversity-and-generate-2d-3d-principal-coordinate-analysis-pcoa-plots">Compute Beta Diversity and Generate 2D/3D Principal Coordinate Analysis (PCoA) Plots</a><ul>
<li><a class="reference external" href="#compute-beta-diversity">Compute Beta Diversity</a></li>
<li><a class="reference external" href="#generate-distance-histograms">Generate Distance Histograms</a></li>
<li><a class="reference external" href="#generate-principal-coordinates">Generate Principal Coordinates</a></li>
<li><a class="reference external" href="#generate-2d-and-3d-pcoa-plots">Generate 2D and 3D PCoA Plots</a></li>
</ul>
</li>
<li><a class="reference external" href="#upgma-clustering-and-jackknifing-support">UPGMA Clustering and Jackknifing Support</a><ul>
<li><a class="reference external" href="#upgma-clustering">UPGMA Clustering</a></li>
<li><a class="reference external" href="#perform-jackknifing-support">Perform Jackknifing Support</a></li>
<li><a class="reference external" href="#compare-jackknifed-trees-to-cluster-tree">Compare Jackknifed Trees to Cluster Tree</a></li>
<li><a class="reference external" href="#generate-bootstrapped-tree">Generate Bootstrapped Tree</a><ul>
<li><a class="reference external" href="#references">References</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>



            <h4>Previous topic</h4>
            <p class="topless"><a href="tut_essential_files.html"
                                  title="previous chapter">Essential Files</a></p>
            <h4>Next topic</h4>
            <p class="topless"><a href="documentation.html"
                                  title="next chapter">QIIME Documentation</a></p>
            <h3>This Page</h3>
            <ul class="this-page-menu">
              <li><a href="_sources/tut_data_analysis_steps.txt"
                     rel="nofollow">Show Source</a></li>
            </ul>
          <div id="searchbox" style="display: none">
            <h3>Quick search</h3>
              <form class="search" action="search.html" method="get">
                <input type="text" name="q" size="18" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
              </form>
              <p class="searchtip" style="font-size: 90%">
              Enter search terms or a module, class or function name.
              </p>
          </div>
          <script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="documentation.html" title="QIIME Documentation"
             >next</a> |</li>
        <li class="right" >
          <a href="tut_essential_files.html" title="Essential Files"
             >previous</a> |</li>
        <li><a href="index.html">Home</a> &raquo;</li>
          <li><a href="tutorial.html" >QIIME Tutorial</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
      &copy; Copyright 2010, QIIME Team.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 0.6.3.
    </div>
  </body>
</html>